{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running SNPGenie on H5N1 data\n",
    "\n",
    "May 10, 2018 and December 8, 2017\n",
    "\n",
    "Popoolation is not a great program for our purposes. I have known this for awhile, but it has always suited our needs well enough that I haven't looked elsewhere. However, now that I would like to do some comparisons of calculating pi at different SNP frequency cutoffs, I need something else. Chase Nelson's program, SNPGenie does this and so I am going to try to get that to work for me. There seem to be a few quirks with it that I will need to work around: \n",
    "\n",
    "1. SNPGenie only works on 1 gene segment at a time. This is a huge drag. It needs to be run 8 separate times for 1 influenza genome. I also figured out that for each genome, you need to provide a vcf that contains only sites from that gene, a gtf file with only coordinates for that cds, and a reference fasta with only that sequence. \n",
    "\n",
    "2. The vcf format options provided are quite close to what I need, but not quite. Varscan output is almost exactly the same as vcf option 4. The only difference is that SNPGenie assumes AD (allele depth) to be the number of reads supporting the reference base, when in varscan, it supports the number of reads for the alternative allele. Additionally, instead of having the reference base reads in 1 column and the alternative allele reads in another, SNPGenie would like them in the same column, separate by a comma. \n",
    "\n",
    "3. SNPGenie will not over-write old folders. So if I want everything in 1 folder, I will need to make new names for them as we proceed.\n",
    "\n",
    "I will need to do several things in order to make SNPGenie work.\n",
    "\n",
    "1. Fix the AD column issue/reformat my vcfs to accomodate that change. \n",
    "2. Automate splitting up vcf files, reference fasta files, and gtf files. \n",
    "3. Take care of automatic renaming of folders. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import necessary modules\n",
    "import sys, subprocess, glob, os, shutil, re, importlib, Bio, csv\n",
    "from subprocess import call\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import seaborn as sns\n",
    "import rpy2\n",
    "%load_ext rpy2.ipython "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Fix AD column issue and split vcf by gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "directory = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a function to combine the AD column for vcf files\n",
    "\n",
    "def format_vcfs_for_snpgenie(vcf_file):\n",
    "    with open(vcf_file, \"r\") as csvfile:\n",
    "        reader = csv.reader(csvfile, delimiter=\"\\t\")\n",
    "        \n",
    "        # get a list of the genes that have SNPs in the original, combined vcf file\n",
    "        genes = []\n",
    "        for row in reader: \n",
    "            if \"#\" not in row[0] and \"##\" not in row[0]:\n",
    "                gene = row[0]\n",
    "                \n",
    "                if gene.split(\"_\")[-1:] == ['MP']:\n",
    "                    genes.append(gene.replace(\"MP\",\"M1\"))\n",
    "                    genes.append(gene.replace(\"MP\",\"M2\"))\n",
    "                if gene.split(\"_\")[-1:] == ['NS']:\n",
    "                    genes.append(gene.replace(\"NS\",\"NS1\"))\n",
    "                    genes.append(gene.replace(\"NS\",\"NEP\"))\n",
    "                else:\n",
    "                    genes.append(gene)\n",
    "        genes = set(genes)\n",
    "        \n",
    "        # for each unique gene, make an empty output file with that gene as its name\n",
    "        for i in genes:\n",
    "            with open(directory + i + \".vcf\", \"w\") as outfile:\n",
    "                outfile.write(\"\")\n",
    "                \n",
    "    # re-open and re-loop through            \n",
    "    with open(vcf_file, \"r\") as infile:\n",
    "        for line in infile:\n",
    "            # if it is a comment line, write it out to all of the output files\n",
    "            if \"#\" in line or \"##\" in line:\n",
    "                for i in genes:\n",
    "                    with open(directory + i + \".vcf\", \"a\") as outfile:\n",
    "                        outfile.write(line)\n",
    "            \n",
    "            # if it's a SNP line, add in the AD annotation and write out to the appropriate file\n",
    "            else: \n",
    "                gene = line.split(\"\\t\")[0]\n",
    "                fix = line.split(\"\\t\")[9]\n",
    "                ref_reads = fix.split(\":\")[4]\n",
    "                alt_reads = fix.split(\":\")[5]\n",
    "                new_depth = ref_reads + \",\" + alt_reads\n",
    "                first_part = fix.split(\":\")[0:5]\n",
    "                first_part = \":\".join(first_part)\n",
    "                second_part = fix.split(\":\")[6:14]\n",
    "                second_part = \":\".join(second_part)\n",
    "                fixed = first_part + \":\" + new_depth + \":\" + second_part\n",
    "                \n",
    "                # combine into a new line\n",
    "                new_line = line.replace(fix, fixed)\n",
    "                \n",
    "                if gene.split(\"_\")[-1:] == ['MP']:\n",
    "                    with open(directory + gene.replace(\"MP\",\"M1\") + \".vcf\", \"a\") as outfile:\n",
    "                        outfile.write(new_line)\n",
    "                    with open(directory + gene.replace(\"MP\",\"M2\") + \".vcf\", \"a\") as outfile:\n",
    "                        outfile.write(new_line)\n",
    "                        \n",
    "                elif gene.split(\"_\")[-1:] == ['NS']:\n",
    "                    with open(directory + gene.replace(\"NS\",\"NS1\") + \".vcf\", \"a\") as outfile:\n",
    "                        outfile.write(new_line)\n",
    "                    with open(directory + gene.replace(\"NS\",\"NEP\") + \".vcf\", \"a\") as outfile:\n",
    "                        outfile.write(new_line)\n",
    "                  \n",
    "                else:\n",
    "                    with open(directory + gene + \".vcf\", \"a\") as outfile:\n",
    "                        outfile.write(new_line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import os\n",
    "\n",
    "def split_vcf_by_gene(vcf_file, output_directory):\n",
    "    # Ensure output directory exists\n",
    "    os.makedirs(output_directory, exist_ok=True)\n",
    "\n",
    "    # Step 1: Identify unique genes\n",
    "    genes = set()\n",
    "    with open(vcf_file, \"r\") as infile:\n",
    "        for line in infile:\n",
    "            if not line.startswith(\"#\"):  # Ignore headers for now\n",
    "                gene = line.split(\"\\t\")[0]\n",
    "                genes.add(gene)\n",
    "\n",
    "    # Step 2: Create empty files for each gene\n",
    "    for gene in genes:\n",
    "        with open(os.path.join(output_directory, f\"{gene}.vcf\"), \"w\") as outfile:\n",
    "            outfile.write(\"\")  # Create empty file\n",
    "\n",
    "    # Step 3: Read the VCF file again and distribute content\n",
    "    with open(vcf_file, \"r\") as infile:\n",
    "        headers = []\n",
    "        for line in infile:\n",
    "            if line.startswith(\"#\"):\n",
    "                headers.append(line)  # Store headers\n",
    "            else:\n",
    "                gene = line.split(\"\\t\")[0]\n",
    "                with open(os.path.join(output_directory, f\"{gene}.vcf\"), \"a\") as outfile:\n",
    "                    outfile.write(line)\n",
    "\n",
    "        # Step 4: Write headers to all gene-specific files\n",
    "        for gene in genes:\n",
    "            with open(os.path.join(output_directory, f\"{gene}.vcf\"), \"r+\") as outfile:\n",
    "                content = outfile.read()\n",
    "                outfile.seek(0, 0)  # Move cursor to the beginning\n",
    "                outfile.writelines(headers + [content])  # Write headers first, then content\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# go through directory and make a vcf file list of all the files that need to be reformatted\n",
    "\n",
    "vcfs = []\n",
    "fastas = []\n",
    "gtfs = []\n",
    "\n",
    "for f in glob.glob(directory + \"*.vcf\"):\n",
    "    vcfs.append(f)\n",
    "\n",
    "for f in glob.glob(directory + \"*.fasta\"):\n",
    "    fastas.append(f)\n",
    "\n",
    "gtf_directory = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/\"\n",
    "for f in glob.glob(gtf_directory + \"*/reference.gtf\"):\n",
    "    gtfs.append(f)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF files found: []\n"
     ]
    }
   ],
   "source": [
    "\n",
    "vcfs = glob.glob(directory + \"*.vcf\")\n",
    "print(\"VCF files found:\", vcfs)  # <-- This should list all .vcf files\n",
    "\n",
    "output_subdir = os.path.join(output_directory, os.path.basename(vcf_file).replace(\".vcf\", \"\"))\n",
    "# split_vcf_by_gene(vcf_file, output_subdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # run the vcf format function\n",
    "# output_directory = '/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf'\n",
    "# output_subdir = os.path.join(output_directory, os.path.basename(vcf_file).replace(\".vcf\", \"\"))\n",
    "\n",
    "# for vcf in vcfs: \n",
    "#     vcf_file = vcf\n",
    "#     split_vcf_by_gene(vcf_file, output_subdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found VCFs: ['/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-1.vcf', '/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-2.vcf', '/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w2_replicate-1.vcf']\n",
      "Processing /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-1.vcf → Output in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf/be_w1_replicate-1\n",
      "✅ Finished /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-1.vcf → Files saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf/be_w1_replicate-1\n",
      "Processing /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-2.vcf → Output in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf/be_w1_replicate-2\n",
      "✅ Finished /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-2.vcf → Files saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf/be_w1_replicate-2\n",
      "Processing /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w2_replicate-1.vcf → Output in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf/be_w2_replicate-1\n",
      "✅ Finished /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w2_replicate-1.vcf → Files saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf/be_w2_replicate-1\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import glob\n",
    "\n",
    "directory = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/\"\n",
    "output_directory = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf/\"\n",
    "\n",
    "# Find all VCF files\n",
    "vcfs = glob.glob(os.path.join(directory, \"*.vcf\"))\n",
    "\n",
    "print(\"Found VCFs:\", vcfs)  # Debugging\n",
    "\n",
    "if not vcfs:\n",
    "    print(\"⚠️ No VCF files found! Check your directory path.\")\n",
    "else:\n",
    "    for vcf_file in vcfs:\n",
    "        # Extract the sample name from the filename\n",
    "        sample_name = os.path.basename(vcf_file).replace(\".vcf\", \"\")\n",
    "        \n",
    "        # Create a subdirectory for each sample\n",
    "        sample_output_dir = os.path.join(output_directory, sample_name)\n",
    "        os.makedirs(sample_output_dir, exist_ok=True)  # Ensure directory exists\n",
    "        \n",
    "        print(f\"Processing {vcf_file} → Output in {sample_output_dir}\")\n",
    "        \n",
    "        # Run your function, directing output to the sample-specific folder\n",
    "        split_vcf_by_gene(vcf_file, sample_output_dir)\n",
    "        \n",
    "        print(f\"✅ Finished {vcf_file} → Files saved in {sample_output_dir}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split gtfs and fastas by gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split up fasta files\n",
    "\n",
    "for fasta in fastas:\n",
    "    with open(fasta, \"r\") as infile: \n",
    "        for seq in SeqIO.parse(infile, \"fasta\"):\n",
    "            name = seq.description\n",
    "            sequence = str(seq.seq)\n",
    "        \n",
    "            if name.endswith(\"MP\"):\n",
    "                with open(directory + name.replace(\"MP\",\"M1\") + \".fasta\", \"w\") as outfile:\n",
    "                    outfile.write(\">\" + name.replace(\"MP\",\"M1\") + \"\\n\")\n",
    "                    outfile.write(sequence) \n",
    "                with open(directory + name.replace(\"MP\",\"M2\") + \".fasta\", \"w\") as outfile:\n",
    "                    outfile.write(\">\" + name.replace(\"MP\",\"M2\") + \"\\n\")\n",
    "                    outfile.write(sequence)              \n",
    "                    \n",
    "            elif name.endswith(\"NS\"):\n",
    "                with open(directory + name.replace(\"NS\",\"NS1\") + \".fasta\", \"w\") as outfile:\n",
    "                    outfile.write(\">\" + name.replace(\"NS\",\"NS1\") + \"\\n\")\n",
    "                    outfile.write(sequence) \n",
    "                with open(directory + name.replace(\"NS\",\"NEP\") + \".fasta\", \"w\") as outfile:\n",
    "                    outfile.write(\">\" + name.replace(\"NS\",\"NEP\") + \"\\n\")\n",
    "                    outfile.write(sequence) \n",
    "                    \n",
    "            else:\n",
    "                with open(directory + name + \".fasta\", \"w\") as outfile:\n",
    "                    outfile.write(\">\" + name + \"\\n\")\n",
    "                    outfile.write(sequence) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split up gtf files \n",
    "\n",
    "for gtf in gtfs: \n",
    "    with open(gtf, \"r\") as infile:\n",
    "        for line in infile:\n",
    "            gene = line.split(\"\\t\")[0]\n",
    "        \n",
    "            if gene.endswith(\"MP\") and \"\\\"M1\\\"\" in line:\n",
    "                with open(directory + gene.replace(\"MP\",\"M1\") + \".gtf\", \"a\") as outfile:\n",
    "                        outfile.write(line)\n",
    "            \n",
    "            elif gene.endswith(\"MP\") and \"\\\"M2\\\"\" in line:\n",
    "                with open(directory + gene.replace(\"MP\",\"M2\") + \".gtf\", \"a\") as outfile:\n",
    "                    outfile.write(line)\n",
    "                    \n",
    "            elif gene.endswith(\"NS\") and \"\\\"NS1\\\"\" in line:\n",
    "                with open(directory + gene.replace(\"NS\",\"NS1\") + \".gtf\", \"a\") as outfile:\n",
    "                    outfile.write(line)\n",
    "            \n",
    "            elif gene.endswith(\"NS\") and \"\\\"NEP\\\"\" in line:\n",
    "                with open(directory + gene.replace(\"NS\",\"NEP\") + \".gtf\", \"a\") as outfile:\n",
    "                    outfile.write(line)\n",
    "                    \n",
    "            else:\n",
    "                with open(directory + gene + \".gtf\", \"a\") as outfile:\n",
    "                    outfile.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ns: reference.gb → ns.gb\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ns: reference.gtf → ns.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ns: reference.fasta → ns.fasta\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/na: reference.gb → na.gb\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/na: reference.gtf → na.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/na: reference.fasta → na.fasta\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb2: reference.gb → pb2.gb\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb2: reference.gtf → pb2.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb2: reference.fasta → pb2.fasta\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pa: reference.gb → pa.gb\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pa: reference.gtf → pa.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pa: reference.fasta → pa.fasta\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ha: reference.gtf → ha.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ha: reference.fasta → ha.fasta\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/np: reference.gb → np.gb\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/np: reference.gtf → np.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/np: reference.fasta → np.fasta\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb1: reference.gb → pb1.gb\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb1: reference.gtf → pb1.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb1: reference.fasta → pb1.fasta\n"
     ]
    }
   ],
   "source": [
    "##renaming reference gtfs and fastas\n",
    "\n",
    "import os\n",
    "\n",
    "base_directory = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference\"  # Change this to your main directory\n",
    "\n",
    "for root, _, files in os.walk(base_directory):\n",
    "    subdir_name = os.path.basename(root)  # Get the subdirectory name\n",
    "\n",
    "    for filename in files:\n",
    "        if filename.startswith(\".\"):  # Ignore hidden/system files\n",
    "            continue\n",
    "\n",
    "        file_extension = os.path.splitext(filename)[-1]  # Extract file extension\n",
    "        \n",
    "        if \"reference\" in filename:\n",
    "            new_name = filename.replace(\"reference\", subdir_name)  # Replace 'reference' with subdirectory name\n",
    "            old_path = os.path.join(root, filename)\n",
    "            new_path = os.path.join(root, new_name)\n",
    "\n",
    "            os.rename(old_path, new_path)\n",
    "            print(f\"Renamed in {root}: {filename} → {new_name}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "📂 Processing: be_w1_replicate-2\n",
      "✅ Renamed: np.vcf → be_w1_replicate-2_np.vcf\n",
      "✅ Renamed: ns.vcf → be_w1_replicate-2_ns.vcf\n",
      "✅ Renamed: na.vcf → be_w1_replicate-2_na.vcf\n",
      "✅ Renamed: pa.vcf → be_w1_replicate-2_pa.vcf\n",
      "✅ Renamed: pb2.vcf → be_w1_replicate-2_pb2.vcf\n",
      "✅ Renamed: ha.vcf → be_w1_replicate-2_ha.vcf\n",
      "✅ Renamed: pb1.vcf → be_w1_replicate-2_pb1.vcf\n",
      "📂 Processing: gtf\n",
      "✅ Renamed: pa.gtf → gtf_pa.gtf\n",
      "✅ Renamed: pb1.gtf → gtf_pb1.gtf\n",
      "✅ Renamed: ha.gtf → gtf_ha.gtf\n",
      "✅ Renamed: pb2.gtf → gtf_pb2.gtf\n",
      "✅ Renamed: ns.gtf → gtf_ns.gtf\n",
      "✅ Renamed: np.gtf → gtf_np.gtf\n",
      "✅ Renamed: na.gtf → gtf_na.gtf\n",
      "📂 Processing: fasta\n",
      "✅ Renamed: pa.fasta → fasta_pa.fasta\n",
      "✅ Renamed: ha.fasta → fasta_ha.fasta\n",
      "✅ Renamed: na.fasta → fasta_na.fasta\n",
      "✅ Renamed: pb2.fasta → fasta_pb2.fasta\n",
      "✅ Renamed: np.fasta → fasta_np.fasta\n",
      "✅ Renamed: ns.fasta → fasta_ns.fasta\n",
      "✅ Renamed: pb1.fasta → fasta_pb1.fasta\n",
      "📂 Processing: be_w1_replicate-1\n",
      "✅ Renamed: np.vcf → be_w1_replicate-1_np.vcf\n",
      "✅ Renamed: ns.vcf → be_w1_replicate-1_ns.vcf\n",
      "✅ Renamed: na.vcf → be_w1_replicate-1_na.vcf\n",
      "✅ Renamed: pa.vcf → be_w1_replicate-1_pa.vcf\n",
      "✅ Renamed: pb2.vcf → be_w1_replicate-1_pb2.vcf\n",
      "✅ Renamed: ha.vcf → be_w1_replicate-1_ha.vcf\n",
      "✅ Renamed: pb1.vcf → be_w1_replicate-1_pb1.vcf\n",
      "📂 Processing: be_w2_replicate-1\n",
      "✅ Renamed: np.vcf → be_w2_replicate-1_np.vcf\n",
      "✅ Renamed: ns.vcf → be_w2_replicate-1_ns.vcf\n",
      "✅ Renamed: na.vcf → be_w2_replicate-1_na.vcf\n",
      "✅ Renamed: pa.vcf → be_w2_replicate-1_pa.vcf\n",
      "✅ Renamed: pb2.vcf → be_w2_replicate-1_pb2.vcf\n",
      "✅ Renamed: ha.vcf → be_w2_replicate-1_ha.vcf\n",
      "✅ Renamed: pb1.vcf → be_w2_replicate-1_pb1.vcf\n",
      "🎉 Done! All files renamed.\n"
     ]
    }
   ],
   "source": [
    "##renaming parsed vcfs to have the sample name so they can be all in one directory\n",
    "\n",
    "# Set the parent directory containing sample subdirectories\n",
    "parent_directory = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf\"  # Change this!\n",
    "\n",
    "# Loop through each subdirectory\n",
    "for sample in os.listdir(parent_directory):\n",
    "    sample_path = os.path.join(parent_directory, sample)\n",
    "    \n",
    "    # Ensure it's a directory\n",
    "    if os.path.isdir(sample_path):\n",
    "        print(f\"📂 Processing: {sample}\")\n",
    "\n",
    "        # Loop through each file in the sample's directory\n",
    "        for filename in os.listdir(sample_path):\n",
    "            old_file_path = os.path.join(sample_path, filename)\n",
    "            \n",
    "            # Construct the new filename\n",
    "            new_filename = f\"{sample}_{filename}\"\n",
    "            new_file_path = os.path.join(sample_path, new_filename)\n",
    "\n",
    "            # Rename the file\n",
    "            os.rename(old_file_path, new_file_path)\n",
    "            print(f\"✅ Renamed: {filename} → {new_filename}\")\n",
    "\n",
    "print(\"🎉 Done! All files renamed.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Consolidate files together into folders by sample name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a list of file names, which are just the gene names that the gtf file, fasta file and vcf files share\n",
    "file_list = []\n",
    "for f in glob.glob(output_directory + \"*.vcf\"):\n",
    "    f = f.replace(\".vcf\",\"\")\n",
    "    file_list.append(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf/'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['be_w1_replicate-2_pa', 'be_w1_replicate-2_pb1', 'be_w1_replicate-2_ha', 'be_w1_replicate-2_pb2', 'be_w1_replicate-2_np', 'be_w1_replicate-2_ns', 'be_w1_replicate-2_na', 'be_w1_replicate-1_ha', 'be_w1_replicate-1_pa', 'be_w1_replicate-1_na', 'be_w1_replicate-1_np', 'be_w1_replicate-1_ns', 'be_w1_replicate-1_pb2', 'be_w1_replicate-1_pb1', 'be_w2_replicate-1_pa', 'be_w2_replicate-1_ha', 'be_w2_replicate-1_ns', 'be_w2_replicate-1_np', 'be_w2_replicate-1_na', 'be_w2_replicate-1_pb1', 'be_w2_replicate-1_pb2']\n"
     ]
    }
   ],
   "source": [
    "import glob\n",
    "import os\n",
    "\n",
    "file_list = []\n",
    "\n",
    "# Search for all VCF files within subdirectories of output_directory\n",
    "for f in glob.glob(os.path.join(output_directory, \"**\", \"*.vcf\"), recursive=True):\n",
    "    # Extract the filename without extension\n",
    "    sample_name = os.path.basename(f).replace(\".vcf\", \"\")\n",
    "    file_list.append(sample_name)\n",
    "\n",
    "print(file_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running SNPGenie for be_w1_replicate-2_pa...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "Results for be_w1_replicate-2_pa saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-2_pa_results/\n",
      "Running SNPGenie for be_w1_replicate-2_pb1...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene PA-X in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING:\n",
      "## The SNP Report /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-2_pb1_results/be_w1_replicate-2_pb1_snpgenie.vcf has no header.\n",
      "## TERMINATED.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Results for be_w1_replicate-2_pb1 saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-2_pb1_results/\n",
      "Running SNPGenie for be_w1_replicate-2_ha...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Results for be_w1_replicate-2_ha saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-2_ha_results/\n",
      "Running SNPGenie for be_w1_replicate-2_pb2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING:\n",
      "## The SNP Report /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-2_ha_results/be_w1_replicate-2_ha_snpgenie.vcf has no header.\n",
      "## TERMINATED.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING:\n",
      "## The SNP Report /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-2_pb2_results/be_w1_replicate-2_pb2_snpgenie.vcf has no header.\n",
      "## TERMINATED.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Results for be_w1_replicate-2_pb2 saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-2_pb2_results/\n",
      "Running SNPGenie for be_w1_replicate-2_np...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "Results for be_w1_replicate-2_np saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-2_np_results/\n",
      "Running SNPGenie for be_w1_replicate-2_ns...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene NP in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene NEP in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "Results for be_w1_replicate-2_ns saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-2_ns_results/\n",
      "Running SNPGenie for be_w1_replicate-2_na...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Results for be_w1_replicate-2_na saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-2_na_results/\n",
      "Running SNPGenie for be_w1_replicate-1_ha...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING:\n",
      "## The SNP Report /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-2_na_results/be_w1_replicate-2_na_snpgenie.vcf has no header.\n",
      "## TERMINATED.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING:\n",
      "## The SNP Report /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-1_ha_results/be_w1_replicate-1_ha_snpgenie.vcf has no header.\n",
      "## TERMINATED.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Results for be_w1_replicate-1_ha saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-1_ha_results/\n",
      "Running SNPGenie for be_w1_replicate-1_pa...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "Results for be_w1_replicate-1_pa saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-1_pa_results/\n",
      "Running SNPGenie for be_w1_replicate-1_na...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene PA-X in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING:\n",
      "## The SNP Report /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-1_na_results/be_w1_replicate-1_na_snpgenie.vcf has no header.\n",
      "## TERMINATED.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Results for be_w1_replicate-1_na saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-1_na_results/\n",
      "Running SNPGenie for be_w1_replicate-1_np...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "Results for be_w1_replicate-1_np saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-1_np_results/\n",
      "Running SNPGenie for be_w1_replicate-1_ns...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene NP in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene NEP in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "Results for be_w1_replicate-1_ns saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-1_ns_results/\n",
      "Running SNPGenie for be_w1_replicate-1_pb2...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Results for be_w1_replicate-1_pb2 saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-1_pb2_results/\n",
      "Running SNPGenie for be_w1_replicate-1_pb1...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING:\n",
      "## The SNP Report /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-1_pb2_results/be_w1_replicate-1_pb2_snpgenie.vcf has no header.\n",
      "## TERMINATED.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING:\n",
      "## The SNP Report /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-1_pb1_results/be_w1_replicate-1_pb1_snpgenie.vcf has no header.\n",
      "## TERMINATED.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene PA-X in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Results for be_w1_replicate-1_pb1 saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w1_replicate-1_pb1_results/\n",
      "Running SNPGenie for be_w2_replicate-1_pa...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "Results for be_w2_replicate-1_pa saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w2_replicate-1_pa_results/\n",
      "Running SNPGenie for be_w2_replicate-1_ha...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING:\n",
      "## The SNP Report /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w2_replicate-1_ha_results/be_w2_replicate-1_ha_snpgenie.vcf has no header.\n",
      "## TERMINATED.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene NEP in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results for be_w2_replicate-1_ha saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w2_replicate-1_ha_results/\n",
      "Running SNPGenie for be_w2_replicate-1_ns...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "Results for be_w2_replicate-1_ns saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w2_replicate-1_ns_results/\n",
      "Running SNPGenie for be_w2_replicate-1_np...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "Results for be_w2_replicate-1_np saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w2_replicate-1_np_results/\n",
      "Running SNPGenie for be_w2_replicate-1_na...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Results for be_w2_replicate-1_na saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w2_replicate-1_na_results/\n",
      "Running SNPGenie for be_w2_replicate-1_pb1...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene NP in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING:\n",
      "## The SNP Report /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w2_replicate-1_na_results/be_w2_replicate-1_na_snpgenie.vcf has no header.\n",
      "## TERMINATED.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Results for be_w2_replicate-1_pb1 saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w2_replicate-1_pb1_results/\n",
      "Running SNPGenie for be_w2_replicate-1_pb2...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "You have not selected a MIN. MINOR ALLELE FREQ. All variants in the SNP report(s) will be included.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Results for be_w2_replicate-1_pb2 saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w2_replicate-1_pb2_results/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING:\n",
      "## The SNP Report /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w2_replicate-1_pb1_results/be_w2_replicate-1_pb1_snpgenie.vcf has no header.\n",
      "## TERMINATED.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING:\n",
      "## The SNP Report /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/be_w2_replicate-1_pb2_results/be_w2_replicate-1_pb2_snpgenie.vcf has no header.\n",
      "## TERMINATED.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from subprocess import call\n",
    "\n",
    "# Define the specific output directory where results will go\n",
    "output_directory = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/\"\n",
    "os.makedirs(output_directory, exist_ok=True)\n",
    "\n",
    "# Define the base path for the reference FASTA and GTF files\n",
    "reference_base_path = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf/reference/\"\n",
    "\n",
    "# Define the mapping between sample suffixes and reference file names\n",
    "reference_mapping = {\n",
    "    'pa': 'pa',  # Reference files for 'pa' samples\n",
    "    'pb1': 'pb1',  # Reference files for 'pb1' samples\n",
    "    'pb2': 'pb2',  # Reference files for 'pb2' samples\n",
    "    'ha': 'ha',  # Reference files for 'ha' samples\n",
    "    'np': 'np',  # Reference files for 'np' samples\n",
    "    'ns': 'ns',  # Reference files for 'ns' samples\n",
    "    'na': 'na',  # Reference files for 'na' samples\n",
    "}\n",
    "\n",
    "# Run SNPGenie for each sample\n",
    "for sample in file_list:\n",
    "    # Extract the sample suffix (e.g., 'pa', 'pb1', 'ha' etc. from sample name)\n",
    "    sample_suffix = sample.split('_')[-1]  # Assuming the format ends with 'pa', 'pb1', etc.\n",
    "    \n",
    "    # Check if the sample_suffix has a valid mapping\n",
    "    if sample_suffix in reference_mapping:\n",
    "        reference_name = reference_mapping[sample_suffix]\n",
    "        \n",
    "        # Construct the paths to the reference FASTA and GTF files\n",
    "        reference_fasta = os.path.join(reference_base_path, f\"{reference_name}.fasta\")\n",
    "        reference_gtf = os.path.join(reference_base_path, f\"{reference_name}.gtf\")\n",
    "\n",
    "        # Check if the reference files exist (optional, for error handling)\n",
    "        if not os.path.exists(reference_fasta) or not os.path.exists(reference_gtf):\n",
    "            print(f\"Warning: Missing reference files for {sample_suffix}. Skipping {sample}.\")\n",
    "            continue\n",
    "        \n",
    "        # Construct the VCF file path (assuming the VCF file is named after the sample)\n",
    "        vcf_file = os.path.join('/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf/', f\"{sample}.vcf\")  # Adjust path if needed\n",
    "        \n",
    "        # Check if the VCF file exists\n",
    "        if not os.path.exists(vcf_file):\n",
    "            print(f\"VCF file missing for sample {sample}. Skipping.\")\n",
    "            continue\n",
    "\n",
    "        # Define the SNPGenie output directory for the sample\n",
    "        snpgenie_output_dir = os.path.join(output_directory, f\"{sample}_results/\")\n",
    "        os.makedirs(snpgenie_output_dir, exist_ok=True)\n",
    "\n",
    "        # Run SNPGenie using the corresponding reference and the sample's VCF file\n",
    "        snpgenie_command = f\"perl /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/snpgenie.pl --vcfformat=4 --gtffile={reference_gtf} --fastafile={reference_fasta} --snpreport={snpgenie_output_dir}{sample}_snpgenie.vcf\"\n",
    "        \n",
    "        # Execute the SNPGenie command\n",
    "        print(f\"Running SNPGenie for {sample}...\")\n",
    "        call(snpgenie_command, shell=True)\n",
    "\n",
    "        # Move SNPGenie results to the specific output folder (if applicable)\n",
    "        # Assuming SNPGenie generates a directory named SNPGenie_Results by default\n",
    "        call(f\"mv SNPGenie_Results {snpgenie_output_dir}\", shell=True)\n",
    "        \n",
    "        print(f\"Results for {sample} saved in {snpgenie_output_dir}\")\n",
    "    else:\n",
    "        print(f\"No reference mapping found for sample suffix {sample_suffix}. Skipping.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# consolidate fastas, vcfs, and gtfs all into 1 folder together \n",
    "for f in file_list:\n",
    "    call(\"mkdir {f}\".format(f=f), shell=True)\n",
    "    for n in glob.glob(directory + \"*\"):\n",
    "        if f in n:\n",
    "            call(\"mv {n} {f}\".format(n=n,f=f), shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the specific path where you want the new directories to be created\n",
    "new_dir_path = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/\"\n",
    "\n",
    "# Consolidate fastas, vcfs, and gtfs all into the specific folder\n",
    "for f in file_list:\n",
    "    # Create the directory in the specific path\n",
    "    call(\"mkdir -p {}/{}\".format(new_dir_path, f), shell=True)\n",
    "\n",
    "    # Move files to the new directory\n",
    "    for n in glob.glob(directory + \"*\"):\n",
    "        if f in n:\n",
    "            call(\"mv {n} {}/{}/\".format(new_dir_path, f, n), shell=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Note: \n",
    "\n",
    "At this point, there will be gtf and fasta files for genes that are not in folders. That is ok. That means that there were no SNPs called for that gene in that sample. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['be_w1_replicate-2_pa',\n",
       " 'be_w1_replicate-2_pb1',\n",
       " 'be_w1_replicate-2_ha',\n",
       " 'be_w1_replicate-2_pb2',\n",
       " 'be_w1_replicate-2_np',\n",
       " 'be_w1_replicate-2_ns',\n",
       " 'be_w1_replicate-2_na',\n",
       " 'be_w1_replicate-1_ha',\n",
       " 'be_w1_replicate-1_pa',\n",
       " 'be_w1_replicate-1_na',\n",
       " 'be_w1_replicate-1_np',\n",
       " 'be_w1_replicate-1_ns',\n",
       " 'be_w1_replicate-1_pb2',\n",
       " 'be_w1_replicate-1_pb1',\n",
       " 'be_w2_replicate-1_pa',\n",
       " 'be_w2_replicate-1_ha',\n",
       " 'be_w2_replicate-1_ns',\n",
       " 'be_w2_replicate-1_np',\n",
       " 'be_w2_replicate-1_na',\n",
       " 'be_w2_replicate-1_pb1',\n",
       " 'be_w2_replicate-1_pb2']"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from subprocess import call\n",
    "\n",
    "# Define the specific output directory where results will go\n",
    "output_directory = \"/desired/output/path/\"\n",
    "os.makedirs(output_directory, exist_ok=True)\n",
    "\n",
    "# Define paths to the reference FASTA and GTF files (they are gene-specific, not sample-specific)\n",
    "reference_fasta = \"/path/to/reference/fastas/reference.fasta\"\n",
    "reference_gtf = \"/path/to/reference/gtfs/reference.gtf\"\n",
    "\n",
    "# Run SNPGenie for each sample\n",
    "for f in file_list:\n",
    "    # Extract the sample name (assuming 'f' is the path to the sample directory)\n",
    "    sample_name = os.path.basename(f)\n",
    "\n",
    "    # Path to the sample's VCF file\n",
    "    vcf_file = os.path.join(f, \"vcfs\", f\"{sample_name}.vcf\")  # Adjust as needed to match exact file names\n",
    "\n",
    "    # Ensure the VCF file exists\n",
    "    if os.path.exists(vcf_file):\n",
    "        # Run SNPGenie using the reference files for each sample\n",
    "        snpgenie_command = f\"/Users/asjaeger/Desktop/project_source/hpai_PA/snpgenie.pl --vcfformat=4 --gtffile={reference_gtf} --fastafile={reference_fasta} --snpreport={output_directory}{sample_name}_results.vcf\"\n",
    "        \n",
    "        # Execute the SNPGenie command\n",
    "        call(snpgenie_command, shell=True)\n",
    "\n",
    "        # Move the SNPGenie results to the output directory (if they aren't already there)\n",
    "        call(f\"mv SNPGenie_Results {output_directory}{sample_name}_results/\", shell=True)\n",
    "    else:\n",
    "        print(f\"VCF file missing for sample {sample_name}. Skipping SNPGenie.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/bin/sh: /Users/asjaeger/Desktop/project_source/hpai_PA/snpgenie.pl: No such file or directory\n",
      "mv: rename SNPGenie_Results to /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-1/SNPGenie_Results: No such file or directory\n",
      "/bin/sh: /Users/asjaeger/Desktop/project_source/hpai_PA/snpgenie.pl: No such file or directory\n",
      "mv: rename SNPGenie_Results to /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-2/SNPGenie_Results: No such file or directory\n",
      "/bin/sh: /Users/asjaeger/Desktop/project_source/hpai_PA/snpgenie.pl: No such file or directory\n",
      "mv: rename SNPGenie_Results to /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w2_replicate-1/SNPGenie_Results: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "# run SNP Genie on all of these\n",
    "for f in file_list:\n",
    "    f1 = f.replace(output_directory, \"\")\n",
    "    call(\"/Users/asjaeger/Desktop/project_source/hpai_PA/snpgenie.pl --vcfformat=4 --gtffile={f}/{f1}.gtf --fastafile={f}/{f1}.fasta --snpreport={f}/{f1}.vcf\".format(f=f,f1=f1),shell=True)\n",
    "    call(\"mv SNPGenie_Results {f}/\".format(f=f,f1=f1), shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 352,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compile all of the output files from the various genes\n",
    "call(\"grep -i '' */SNPGenie_Results/product_results.txt >> combined.txt\".format(directory=directory),shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 583,
   "metadata": {},
   "outputs": [],
   "source": [
    "# take out all of the unwanted lines from the combined file\n",
    "def cleanup_snpgenie_output(directory, infile):\n",
    "    with open(directory + infile, \"r\") as infile: \n",
    "        linenumber = 0\n",
    "        for line in infile:\n",
    "            linenumber += 1\n",
    "            if linenumber == 1:\n",
    "                new_line = line.split(\":\")[1]\n",
    "                with open(directory + \"combined_pi_formatted.txt\", \"a\") as outfile:\n",
    "                    outfile.write(new_line)\n",
    "        \n",
    "            elif \"N_diffs\" not in line:\n",
    "                new_line = line.replace(\"/SNPGenie_Results/product_results.txt:temp_vcf4_Sample1.vcf\",\"\")\n",
    "            \n",
    "                genes = [\"_PB2\",\"_PB1\",\"_PA\",\"_H5\",\"_NP\",\"_N1\",\"_M1\",\"_M2\",\"_NS1\",\"_NEP\"]\n",
    "                for g in genes: \n",
    "                    if g + \"\\t\" in new_line:\n",
    "                        new_line2 = new_line.replace(g, \"\")\n",
    "                with open(directory + \"combined_pi_formatted.txt\", \"a\") as outfile:\n",
    "                    outfile.write(new_line2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 584,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleanup_snpgenie_output(directory, \"combined.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "source": [
    "## Results:\n",
    "\n",
    "This actually all worked really well! Except for the grep, which for some reason works totally fine in shell but isn't working for me when run from within this notebook. Otherwise though, things look good! I will now go through and see which genes didn't work/have empty values. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data wrangling and plotting:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 585,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>product</th>\n",
       "      <th>N_diffs</th>\n",
       "      <th>S_diffs</th>\n",
       "      <th>N_diffs_vs_ref</th>\n",
       "      <th>S_diffs_vs_ref</th>\n",
       "      <th>N_sites</th>\n",
       "      <th>S_sites</th>\n",
       "      <th>piN</th>\n",
       "      <th>piS</th>\n",
       "      <th>mean_dN_vs_ref</th>\n",
       "      <th>mean_dS_vs_ref</th>\n",
       "      <th>mean_gdiv_polymorphic</th>\n",
       "      <th>mean_N_gdiv</th>\n",
       "      <th>mean_S_gdiv</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AH7E5L724F516_A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>HA</td>\n",
       "      <td>0.118954</td>\n",
       "      <td>0.403580</td>\n",
       "      <td>0.063241</td>\n",
       "      <td>0.278997</td>\n",
       "      <td>1328.854414</td>\n",
       "      <td>372.145586</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.001084</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.000750</td>\n",
       "      <td>0.260399283550513</td>\n",
       "      <td>0.118483338280554</td>\n",
       "      <td>0.402315228820471</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AH7E5L724F516_A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>NP</td>\n",
       "      <td>0.235276</td>\n",
       "      <td>0.445634</td>\n",
       "      <td>0.122255</td>\n",
       "      <td>0.235606</td>\n",
       "      <td>1145.707418</td>\n",
       "      <td>348.292582</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.001279</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>0.000676</td>\n",
       "      <td>0.0847834938232775</td>\n",
       "      <td>0.0781347069643767</td>\n",
       "      <td>0.0887727659386179</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AH7E5L724F516_A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.150964</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.081633</td>\n",
       "      <td>1751.833333</td>\n",
       "      <td>525.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000287</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000155</td>\n",
       "      <td>0.14993752603082</td>\n",
       "      <td>*</td>\n",
       "      <td>0.14993752603082</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AJ4MBL718F513_A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>0.549191</td>\n",
       "      <td>0.280857</td>\n",
       "      <td>0.312999</td>\n",
       "      <td>0.151515</td>\n",
       "      <td>1332.785015</td>\n",
       "      <td>371.214985</td>\n",
       "      <td>0.000412</td>\n",
       "      <td>0.000757</td>\n",
       "      <td>0.000235</td>\n",
       "      <td>0.000408</td>\n",
       "      <td>0.137771377959253</td>\n",
       "      <td>0.136868269877355</td>\n",
       "      <td>0.139577594123049</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AJ4MBL718F513_A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>576.666667</td>\n",
       "      <td>179.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          file product   N_diffs   S_diffs  \\\n",
       "0  AH7E5L724F516_A_duck_Cambodia_Y0224304_2014      HA  0.118954  0.403580   \n",
       "1  AH7E5L724F516_A_duck_Cambodia_Y0224304_2014      NP  0.235276  0.445634   \n",
       "2  AH7E5L724F516_A_duck_Cambodia_Y0224304_2014     PB2  0.000000  0.150964   \n",
       "3       AJ4MBL718F513_A_CAMBODIA_V0417301_2011      HA  0.549191  0.280857   \n",
       "4       AJ4MBL718F513_A_CAMBODIA_V0417301_2011      M1  0.000000  0.000000   \n",
       "\n",
       "   N_diffs_vs_ref  S_diffs_vs_ref      N_sites     S_sites       piN  \\\n",
       "0        0.063241        0.278997  1328.854414  372.145586  0.000090   \n",
       "1        0.122255        0.235606  1145.707418  348.292582  0.000205   \n",
       "2        0.000000        0.081633  1751.833333  525.166667  0.000000   \n",
       "3        0.312999        0.151515  1332.785015  371.214985  0.000412   \n",
       "4        0.000000        0.000000   576.666667  179.333333  0.000000   \n",
       "\n",
       "        piS  mean_dN_vs_ref  mean_dS_vs_ref mean_gdiv_polymorphic  \\\n",
       "0  0.001084        0.000048        0.000750     0.260399283550513   \n",
       "1  0.001279        0.000107        0.000676    0.0847834938232775   \n",
       "2  0.000287        0.000000        0.000155      0.14993752603082   \n",
       "3  0.000757        0.000235        0.000408     0.137771377959253   \n",
       "4  0.000000        0.000000        0.000000                     *   \n",
       "\n",
       "          mean_N_gdiv         mean_S_gdiv species  \n",
       "0   0.118483338280554   0.402315228820471    duck  \n",
       "1  0.0781347069643767  0.0887727659386179    duck  \n",
       "2                   *    0.14993752603082    duck  \n",
       "3   0.136868269877355   0.139577594123049   human  \n",
       "4                   *                   *   human  "
      ]
     },
     "execution_count": 585,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in the SNPGenie pi results as a dataframe\n",
    "df = pd.read_table(directory + \"combined_pi_formatted.txt\", sep=\"\\t\")\n",
    "df['product'] = df['product'].fillna('neuram')                                     # replace NA with neuram\n",
    "df['species'] = np.where((df['file'].str.contains(\"duck\")), \"duck\",\"human\")        # add in a species column\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 586,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>measure</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HA</th>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>piN</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1</th>\n",
       "      <td>0.000115</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>piN</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M2</th>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000144</td>\n",
       "      <td>piN</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NEP</th>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>piN</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP</th>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>piN</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         mean        sd measure species\n",
       "HA   0.000045  0.000051     piN    duck\n",
       "M1   0.000115  0.000131     piN    duck\n",
       "M2   0.000065  0.000144     piN    duck\n",
       "NEP  0.000025  0.000057     piN    duck\n",
       "NP   0.000099  0.000108     piN    duck"
      ]
     },
     "execution_count": 586,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cast the dataframe and calculate some mean values for piN \n",
    "pN_h = df[df['species'] == 'human']\n",
    "pN_h = pN_h[['file','product','piN']]\n",
    "pN_h = pN_h.pivot(index='file', columns='product', values='piN')\n",
    "pN_h.index.name = None\n",
    "pN_h.columns.name = None\n",
    "pN_h = pN_h.fillna(0)\n",
    "pN_h.loc['mean'] = pN_h.mean()\n",
    "pN_h.loc['sd'] = pN_h.std()\n",
    "pN_means_h = pN_h.loc[['mean','sd']]\n",
    "pN_means_h = pN_means_h.transpose()\n",
    "pN_means_h[\"measure\"] = \"piN\"\n",
    "pN_means_h[\"species\"] = \"human\"\n",
    "\n",
    "pN_d = df[df['species'] == 'duck']\n",
    "pN_d = pN_d[['file','product','piN']]\n",
    "pN_d = pN_d.pivot(index='file', columns='product', values='piN')\n",
    "pN_d.index.name = None\n",
    "pN_d.columns.name = None\n",
    "pN_d = pN_d.fillna(0)\n",
    "pN_d.loc['mean'] = pN_d.mean()\n",
    "pN_d.loc['sd'] = pN_d.std()\n",
    "pN_means_d = pN_d.loc[['mean','sd']]\n",
    "pN_means_d = pN_means_d.transpose()\n",
    "pN_means_d[\"measure\"] = \"piN\"\n",
    "pN_means_d[\"species\"] = \"duck\"\n",
    "pN_means_d.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 587,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>measure</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HA</th>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000397</td>\n",
       "      <td>piS</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>piS</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>piS</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NEP</th>\n",
       "      <td>0.000262</td>\n",
       "      <td>0.000380</td>\n",
       "      <td>piS</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP</th>\n",
       "      <td>0.000237</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>piS</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         mean        sd measure species\n",
       "HA   0.000207  0.000397     piS    duck\n",
       "M1   0.000000  0.000000     piS    duck\n",
       "M2   0.000000  0.000000     piS    duck\n",
       "NEP  0.000262  0.000380     piS    duck\n",
       "NP   0.000237  0.000469     piS    duck"
      ]
     },
     "execution_count": 587,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cast the dataframe and calculate some mean values for piN \n",
    "pS_h = df[df['species'] == 'human']\n",
    "pS_h = pS_h[['file','product','piS']]\n",
    "pS_h = pS_h.pivot(index='file', columns='product', values='piS')\n",
    "pS_h.index.name = None\n",
    "pS_h.columns.name = None\n",
    "pS_h = pS_h.fillna(0)\n",
    "pS_h.loc['mean'] = pS_h.mean()\n",
    "pS_h.loc['sd'] = pS_h.std()\n",
    "pS_means_h = pS_h.loc[['mean','sd']]\n",
    "pS_means_h = pS_means_h.transpose()\n",
    "pS_means_h[\"measure\"] = \"piS\"\n",
    "pS_means_h[\"species\"] = \"human\"\n",
    "\n",
    "pS_d = df[df['species'] == 'duck']\n",
    "pS_d = pS_d[['file','product','piS']]\n",
    "pS_d = pS_d.pivot(index='file', columns='product', values='piS')\n",
    "pS_d.index.name = None\n",
    "pS_d.columns.name = None\n",
    "pS_d = pS_d.fillna(0)\n",
    "pS_d.loc['mean'] = pS_d.mean()\n",
    "pS_d.loc['sd'] = pS_d.std()\n",
    "pS_means_d = pS_d.loc[['mean','sd']]\n",
    "pS_means_d = pS_means_d.transpose()\n",
    "pS_means_d[\"measure\"] = \"piS\"\n",
    "pS_means_d[\"species\"] = \"duck\"\n",
    "pS_means_d.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 588,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>measure</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>piN</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>piN</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M2</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>piN</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NEP</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>piN</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>piN</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  index      mean        sd measure species\n",
       "0    HA  0.000253  0.000260     piN   human\n",
       "1    M1  0.000033  0.000051     piN   human\n",
       "2    M2  0.000123  0.000175     piN   human\n",
       "3   NEP  0.000051  0.000113     piN   human\n",
       "4    NP  0.000036  0.000055     piN   human"
      ]
     },
     "execution_count": 588,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine the mean and stdev values for piN and piS into 1 dataframe \n",
    "df1 = pd.concat([pN_means_h, pN_means_d, pS_means_h, pS_means_d])\n",
    "df1.reset_index(inplace=True)\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/rpy2/rinterface/__init__.py:145: RRuntimeWarning: Warning:\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/rpy2/rinterface/__init__.py:145: RRuntimeWarning:  Ignoring unknown parameters: stat\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "%%R -w 800 -h 500 -u px -i df1  # this sets the size of the plot...otherwise, it will go off the page\n",
    "require(ggplot2)\n",
    "library(ggplot2)\n",
    "\n",
    "df1$index = gsub(\"neuram\", \"NA\", df1$index)\n",
    "df1$gene = factor(df1$index, levels=c(\"PB2\",\"PB1\",\"PA\",\"HA\",\"NP\",\"NA\",\"M1\",\"M2\",\"NS1\",\"NEP\"))\n",
    "\n",
    "p <- ggplot(data=df1, aes(x=gene, y=mean, fill=species)) + \n",
    "    geom_col(position=\"dodge\", stat=\"identity\")+\n",
    "    geom_errorbar(data=df1, aes(x=gene, ymin=mean-sd, ymax=mean+sd, color=species),position=position_dodge(0.9), stat=\"identity\")+\n",
    "    facet_wrap(~measure, scales=\"free\")+\n",
    "    scale_fill_manual(values=c(\"#99bfaa\",\"#5c3d46\"))+\n",
    "    scale_color_manual(values=c(\"#99bfaa\",\"#5c3d46\"))+\n",
    "    labs(x=\"gene\",y=\"diversity\")+\n",
    "    #ggtitle(\"number of shared sites in random simulations\") + \n",
    "    scale_y_continuous(limits=c(-0.0006,0.002))+\n",
    "    #scale_x_continuous(limits=c(0,500))+\n",
    "    theme(plot.title = element_text(size=20, hjust=0.5))+\n",
    "    theme(panel.grid.major.y=element_line(colour=NA))+\n",
    "    theme(panel.grid.minor=element_line(colour=NA,size=NA))+    \n",
    "    theme(strip.background = element_rect(colour=NA, fill=NA))+\n",
    "    theme(axis.line.x=element_line(colour=\"black\"))+\n",
    "    theme(axis.line.y=element_line(colour=\"black\"))+theme(strip.text.x=element_text(size=11))+\n",
    "    theme(axis.title.y=element_text(size=16, vjust=8))+\n",
    "    theme(axis.title.x=element_text(size=16, vjust=-8))+\n",
    "    theme(axis.text=element_text(size=16, colour=\"black\"))+\n",
    "    theme(axis.text.x=element_text(size=16, angle=90))+\n",
    "    theme(legend.text=element_text(size=16))+theme(legend.title=element_text(size=16, face=\"plain\"))+\n",
    "    theme(panel.margin=unit(1, \"lines\"))+theme(plot.margin=unit(c(1,1,1,1),\"cm\"))+\n",
    "    theme(legend.key.size=unit(0.7, \"cm\"))+\n",
    "    theme(panel.background=element_rect(fill=NA))+\n",
    "    theme(legend.key=element_rect(fill=NA))\n",
    "\n",
    "ggsave(\"piN_and_piS_means-2018-05-10.pdf\", p, width = 10, height = 5, device=pdf, path=\"/Users/lmoncla/Documents/H5N1_Cambodian_outbreak_study/paper-and-figure-drafts/figures-2018-05-08\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot number of SNPs/site\n",
    "\n",
    "I am starting to think that maybe using pi for this dataset isn't so great. I think that I am going to try out and see how it looks to just plot the overall number of nonsynonymous and synonymous SNPs per nonsynonymous or synonymou site per gene per species. I can read in the number of SNPs from the combined SNPs output file. For the number of synonymous and nonsynonymous sites, I need to count the samples in df (those that had a snp) but also those that did not. I think that the easiest way to do that will be to just run SNPGenie on a set of dummy data so that I can get the number of synonymous and nonsynonymous sites per gene per sample.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 1: Calculate the average number of synonymous and nonsynonymous sites per gene \n",
    "\n",
    "This will involve using the ones I already have plus generating fake vcf files so that I can run SNPGenie for the samples/genes that didn't have any SNPs. \n",
    "\n",
    "\n",
    "#### Generate dummy vcfs to get the number of nonsynonymous and synonymous sites for genes that don't have snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 589,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_vcfs_directory = \"/Users/lmoncla/Documents/H5N1_Cambodian_outbreak_study/SNPGenie_diversity_estimates/1_percent_SNPs-2018-05-10/samples_and_genes_with_no_snps/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 590,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = []\n",
    "for f in glob.glob(missing_vcfs_directory + \"*.gtf\"):\n",
    "    files.append(f)\n",
    "    sample_name = f.replace(missing_vcfs_directory, \"\")\n",
    "    sample_name = sample_name.replace(\".gtf\",\"\")\n",
    "    samples.append(sample_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 591,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make fake vcf files for each sample \n",
    "for s in samples: \n",
    "    vcfname = missing_vcfs_directory + s + \".vcf\"\n",
    "    with open(vcfname, \"w\") as vcf: \n",
    "        vcf.write(\"#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tSample1\\n\")\n",
    "        vcf.write(s+\"\\t28\\t.\\tC\\tT\\t.\\tPASS\\tADP=147;WT=1000;HET=1;HOM=0;NC=0\\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\\t0/1:37:147:147:135:135,12:8.16%:1.9326E-4:38:37:61:74:5:7\\n\")\n",
    "        vcf.write(s+\"\\t100\\t.\\tC\\tT\\t.\\tPASS\\tADP=147;WT=1000;HET=1;HOM=0;NC=0\\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\\t0/1:37:147:147:135:135,12:8.16%:1.9326E-4:38:37:61:74:5:7\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 593,
   "metadata": {},
   "outputs": [],
   "source": [
    "# consolidate fastas, vcfs, and gtfs all into 1 folder together \n",
    "for s in samples:\n",
    "    call(\"mkdir {missing_vcfs_directory}{s}\".format(missing_vcfs_directory=missing_vcfs_directory,s=s), shell=True)\n",
    "    for n in glob.glob(missing_vcfs_directory + \"*\"):\n",
    "        if s in n:\n",
    "            call(\"mv {n} {missing_vcfs_directory}{s}\".format(s=s, n=n,missing_vcfs_directory=missing_vcfs_directory), shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 594,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run SNPgenie\n",
    "for s in samples:\n",
    "    call(\"snpgenie.pl --vcfformat=4 --gtffile={missing_vcfs_directory}{s}/{s}.gtf --fastafile={missing_vcfs_directory}{s}/{s}.fasta --snpreport={missing_vcfs_directory}{s}/{s}.vcf\".format(missing_vcfs_directory=missing_vcfs_directory,s=s),shell=True)\n",
    "    call(\"mv SNPGenie_Results {missing_vcfs_directory}{s}/\".format(missing_vcfs_directory=missing_vcfs_directory,s=s), shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compile all of the output files from the various genes\n",
    "call(\"grep -i '' */SNPGenie_Results/product_results.txt >> combined.txt\".format(directory=directory),shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 595,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clean up outfile\n",
    "cleanup_snpgenie_output(missing_vcfs_directory, \"combined.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 605,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42\n"
     ]
    }
   ],
   "source": [
    "# read in the SNPGenie pi results as a dataframe\n",
    "df_dummy_values = pd.read_table(missing_vcfs_directory + \"combined_pi_formatted.txt\", sep=\"\\t\")\n",
    "df_dummy_values['product'] = df['product'].fillna('neuram')                                     # replace NA with neuram\n",
    "df_dummy_values['species'] = np.where((df_dummy_values['file'].str.contains(\"duck\")), \"duck\",\"human\")        # add in a species column\n",
    "df_dummy_values = df_dummy_values[['file','product','N_sites','S_sites','species']]\n",
    "df_dummy_values = df_dummy_values.rename(index=str, columns={\"file\":'sample','product':'gene'})\n",
    "df_dummy_values.head()\n",
    "print(len(df_dummy_values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 606,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "78\n"
     ]
    }
   ],
   "source": [
    "# read in the values from the genes that did have SNPs\n",
    "sites = df[['file','product','N_sites','S_sites','species']]\n",
    "sites = sites.rename(index=str, columns={'file':'sample','product':'gene'})\n",
    "sites.head()\n",
    "print(len(sites))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 607,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>N_sites</th>\n",
       "      <th>S_sites</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA</td>\n",
       "      <td>1328.854414</td>\n",
       "      <td>372.145586</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP</td>\n",
       "      <td>1145.707418</td>\n",
       "      <td>348.292582</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PB2</td>\n",
       "      <td>1751.833333</td>\n",
       "      <td>525.166667</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HA</td>\n",
       "      <td>1332.785015</td>\n",
       "      <td>371.214985</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1</td>\n",
       "      <td>576.666667</td>\n",
       "      <td>179.333333</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene      N_sites     S_sites species\n",
       "0   HA  1328.854414  372.145586    duck\n",
       "1   NP  1145.707418  348.292582    duck\n",
       "2  PB2  1751.833333  525.166667    duck\n",
       "3   HA  1332.785015  371.214985   human\n",
       "4   M1   576.666667  179.333333   human"
      ]
     },
     "execution_count": 607,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine dummy and real data into 1 dataframe\n",
    "sites2 = pd.concat([sites,df_dummy_values])\n",
    "sites2 = sites2[['gene','N_sites','S_sites','species']]\n",
    "sites2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 608,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  # Remove the CWD from sys.path while we load stuff.\n",
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  # This is added back by InteractiveShellApp.init_path()\n",
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  if sys.path[0] == '':\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>N_sites</th>\n",
       "      <th>S_sites</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NEP</td>\n",
       "      <td>286.101039</td>\n",
       "      <td>76.898961</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NEP</td>\n",
       "      <td>286.101039</td>\n",
       "      <td>76.898961</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NEP</td>\n",
       "      <td>285.992593</td>\n",
       "      <td>77.007407</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NS1</td>\n",
       "      <td>520.724740</td>\n",
       "      <td>154.275260</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NS1</td>\n",
       "      <td>521.000000</td>\n",
       "      <td>154.000000</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene     N_sites     S_sites species\n",
       "mean  NEP  286.101039   76.898961   human\n",
       "mean  NEP  286.101039   76.898961   human\n",
       "mean  NEP  285.992593   77.007407    duck\n",
       "mean  NS1  520.724740  154.275260   human\n",
       "mean  NS1  521.000000  154.000000    duck"
      ]
     },
     "execution_count": 608,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate the mean number of nonsynonymous and synonymous sites per gene per species and output that into a dataframe\n",
    "genes = ['PB2','PB1','PA','HA','NP','neuram','M1','M2','NS1','NEP']\n",
    "species = ['duck','human']\n",
    "count = 0\n",
    "\n",
    "for g in genes: \n",
    "    for s in species:\n",
    "        count += 1\n",
    "        subsetdf = sites2.loc[(sites2['gene'] == g) & (sites2['species'] == s)]\n",
    "        subsetdf.loc['mean'] =  subsetdf.mean()\n",
    "        subsetdf['gene'] = g\n",
    "        subsetdf['species'] = s\n",
    "        subsetdf = subsetdf.loc[['mean']]\n",
    "        \n",
    "        if count == 1:\n",
    "            mean_sites_df = subsetdf\n",
    "        else:\n",
    "            mean_sites_df = pd.concat([subsetdf, means_df])\n",
    "\n",
    "mean_sites_df['gene'] = mean_sites_df['gene'].str.replace(\"neuram\",\"NA\")\n",
    "mean_sites_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 609,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'duck': {'M1': {'synonymous': 0, 'nonsynonymous': 4}, 'M2': {'synonymous': 0, 'nonsynonymous': 1}, 'NA': {'synonymous': 2, 'nonsynonymous': 2}, 'NP': {'synonymous': 6, 'nonsynonymous': 6}, 'NS1': {'synonymous': 1, 'nonsynonymous': 3}, 'NEP': {'synonymous': 2, 'nonsynonymous': 1}, 'PA': {'synonymous': 6, 'nonsynonymous': 2}, 'PB2': {'synonymous': 8, 'nonsynonymous': 0}, 'HA': {'synonymous': 2, 'nonsynonymous': 4}, 'PB1': {'synonymous': 3, 'nonsynonymous': 1}}, 'human': {'HA': {'synonymous': 9, 'nonsynonymous': 11}, 'M2': {'synonymous': 1, 'nonsynonymous': 3}, 'NA': {'synonymous': 6, 'nonsynonymous': 4}, 'NP': {'synonymous': 8, 'nonsynonymous': 5}, 'PA': {'synonymous': 17, 'nonsynonymous': 9}, 'PB2': {'synonymous': 6, 'nonsynonymous': 10}, 'PB1': {'synonymous': 14, 'nonsynonymous': 10}, 'M1': {'synonymous': 5, 'nonsynonymous': 2}, 'NS1': {'synonymous': 3, 'nonsynonymous': 0}, 'NEP': {'synonymous': 0, 'nonsynonymous': 1}}}\n"
     ]
    }
   ],
   "source": [
    "# read in snps csv file and count the number of nonsynonymous and synonymous per gene per species\n",
    "\n",
    "with open(snps_path, \"r\") as infile: \n",
    "    snps_dict = {}\n",
    "    for line in infile: \n",
    "        if \"reference_position\" not in line:\n",
    "            sample = line.split(\"\\t\")[0]\n",
    "            sample = \"_\".join(sample.split(\"_\")[:-1])\n",
    "        \n",
    "            if \"duck\" in sample:\n",
    "                species = \"duck\"\n",
    "            else:\n",
    "                species = \"human\"\n",
    "        \n",
    "            gene = line.split(\"\\t\")[1]\n",
    "            syn_nonsyn = line.split(\"\\t\")[6]\n",
    "\n",
    "            if species not in snps_dict:\n",
    "                snps_dict[species] = {}\n",
    "            if gene not in snps_dict[species]:\n",
    "                snps_dict[species][gene] = {\"synonymous\":0, \"nonsynonymous\":0}\n",
    "            if syn_nonsyn == \"nonsynonymous\":\n",
    "                snps_dict[species][gene][\"nonsynonymous\"] += 1\n",
    "            elif syn_nonsyn == \"synonymous\":\n",
    "                snps_dict[species][gene][\"synonymous\"] += 1\n",
    "                \n",
    "print(snps_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 610,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>synonymous</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HA</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1</th>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NEP</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     synonymous  nonsynonymous species\n",
       "HA            2              4    duck\n",
       "M1            0              4    duck\n",
       "M2            0              1    duck\n",
       "NA            2              2    duck\n",
       "NEP           2              1    duck"
      ]
     },
     "execution_count": 610,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make into a dataframe; this has the total number of SNPs detected per gene per species, not an average\n",
    "hsnps_df = pd.DataFrame.from_dict(snps_dict[\"human\"], orient=\"index\")\n",
    "hsnps_df['species'] = \"human\"\n",
    "dsnps_df = pd.DataFrame.from_dict(snps_dict[\"duck\"], orient=\"index\")\n",
    "dsnps_df['species'] = \"duck\"\n",
    "snps_df = pd.concat([dsnps_df, hsnps_df])\n",
    "snps_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 611,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>species</th>\n",
       "      <th>mean_synonymous</th>\n",
       "      <th>mean_nonsynonymous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NA</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NEP</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene  synonymous  nonsynonymous species  mean_synonymous  mean_nonsynonymous\n",
       "0   HA           2              4    duck         0.333333            0.666667\n",
       "1   M1           0              4    duck         0.000000            0.666667\n",
       "2   M2           0              1    duck         0.000000            0.166667\n",
       "3   NA           2              2    duck         0.333333            0.333333\n",
       "4  NEP           2              1    duck         0.333333            0.166667"
      ]
     },
     "execution_count": 611,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# take the mean number \n",
    "snps_df['mean_synonymous'] = snps_df['synonymous']/6\n",
    "snps_df['mean_nonsynonymous'] = snps_df['nonsynonymous']/6\n",
    "snps_df = snps_df.reset_index()\n",
    "snps_df = snps_df.rename(index=str, columns={'index':'gene'})\n",
    "snps_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 612,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>N_sites</th>\n",
       "      <th>S_sites</th>\n",
       "      <th>species</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>mean_synonymous</th>\n",
       "      <th>mean_nonsynonymous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NEP</td>\n",
       "      <td>286.101039</td>\n",
       "      <td>76.898961</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NEP</td>\n",
       "      <td>286.101039</td>\n",
       "      <td>76.898961</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NEP</td>\n",
       "      <td>285.992593</td>\n",
       "      <td>77.007407</td>\n",
       "      <td>duck</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NS1</td>\n",
       "      <td>520.724740</td>\n",
       "      <td>154.275260</td>\n",
       "      <td>human</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NS1</td>\n",
       "      <td>521.000000</td>\n",
       "      <td>154.000000</td>\n",
       "      <td>duck</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene     N_sites     S_sites species  synonymous  nonsynonymous  \\\n",
       "0  NEP  286.101039   76.898961   human           0              1   \n",
       "1  NEP  286.101039   76.898961   human           0              1   \n",
       "2  NEP  285.992593   77.007407    duck           2              1   \n",
       "3  NS1  520.724740  154.275260   human           3              0   \n",
       "4  NS1  521.000000  154.000000    duck           1              3   \n",
       "\n",
       "   mean_synonymous  mean_nonsynonymous  \n",
       "0         0.000000            0.166667  \n",
       "1         0.000000            0.166667  \n",
       "2         0.333333            0.166667  \n",
       "3         0.500000            0.000000  \n",
       "4         0.166667            0.500000  "
      ]
     },
     "execution_count": 612,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = pd.merge(mean_sites_df, snps_df, on=['gene','species'], how=\"outer\")\n",
    "df2 = df2.dropna(axis=0, how='any')\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 613,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>N_sites</th>\n",
       "      <th>S_sites</th>\n",
       "      <th>species</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>mean_synonymous</th>\n",
       "      <th>mean_nonsynonymous</th>\n",
       "      <th>s_snps/s_site</th>\n",
       "      <th>ns_snps/ns_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NEP</td>\n",
       "      <td>286.101039</td>\n",
       "      <td>76.898961</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NEP</td>\n",
       "      <td>286.101039</td>\n",
       "      <td>76.898961</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NEP</td>\n",
       "      <td>285.992593</td>\n",
       "      <td>77.007407</td>\n",
       "      <td>duck</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.004329</td>\n",
       "      <td>0.000583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NS1</td>\n",
       "      <td>520.724740</td>\n",
       "      <td>154.275260</td>\n",
       "      <td>human</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003241</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NS1</td>\n",
       "      <td>521.000000</td>\n",
       "      <td>154.000000</td>\n",
       "      <td>duck</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.001082</td>\n",
       "      <td>0.000960</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene     N_sites     S_sites species  synonymous  nonsynonymous  \\\n",
       "0  NEP  286.101039   76.898961   human           0              1   \n",
       "1  NEP  286.101039   76.898961   human           0              1   \n",
       "2  NEP  285.992593   77.007407    duck           2              1   \n",
       "3  NS1  520.724740  154.275260   human           3              0   \n",
       "4  NS1  521.000000  154.000000    duck           1              3   \n",
       "\n",
       "   mean_synonymous  mean_nonsynonymous  s_snps/s_site  ns_snps/ns_site  \n",
       "0         0.000000            0.166667       0.000000         0.000583  \n",
       "1         0.000000            0.166667       0.000000         0.000583  \n",
       "2         0.333333            0.166667       0.004329         0.000583  \n",
       "3         0.500000            0.000000       0.003241         0.000000  \n",
       "4         0.166667            0.500000       0.001082         0.000960  "
      ]
     },
     "execution_count": 613,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate the mean number of nonsynonymous snps per nonsynonymous site and synonymous snps per synonymous site\n",
    "df2['s_snps/s_site'] = df2['mean_synonymous']/df2['S_sites']\n",
    "df2['ns_snps/ns_site'] = df2['mean_nonsynonymous']/df2['N_sites']\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 614,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>species</th>\n",
       "      <th>variable</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NEP</td>\n",
       "      <td>human</td>\n",
       "      <td>s_snps/s_site</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NEP</td>\n",
       "      <td>human</td>\n",
       "      <td>s_snps/s_site</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NEP</td>\n",
       "      <td>duck</td>\n",
       "      <td>s_snps/s_site</td>\n",
       "      <td>0.004329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NS1</td>\n",
       "      <td>human</td>\n",
       "      <td>s_snps/s_site</td>\n",
       "      <td>0.003241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NS1</td>\n",
       "      <td>duck</td>\n",
       "      <td>s_snps/s_site</td>\n",
       "      <td>0.001082</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene species       variable     value\n",
       "0  NEP   human  s_snps/s_site  0.000000\n",
       "1  NEP   human  s_snps/s_site  0.000000\n",
       "2  NEP    duck  s_snps/s_site  0.004329\n",
       "3  NS1   human  s_snps/s_site  0.003241\n",
       "4  NS1    duck  s_snps/s_site  0.001082"
      ]
     },
     "execution_count": 614,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# melt to plot \n",
    "df2 = pd.melt(df2, id_vars = [\"gene\",\"species\"], value_vars = [\"s_snps/s_site\",\"ns_snps/ns_site\"])\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 615,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/rpy2/rinterface/__init__.py:145: RRuntimeWarning: Warning:\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/rpy2/rinterface/__init__.py:145: RRuntimeWarning:  Ignoring unknown parameters: stat\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "%%R -w 800 -h 500 -u px -i df2  # this sets the size of the plot...otherwise, it will go off the page\n",
    "require(ggplot2)\n",
    "library(ggplot2)\n",
    "\n",
    "df2$gene = factor(df2$gene, levels=c(\"PB2\",\"PB1\",\"PA\",\"HA\",\"NP\",\"NA\",\"M1\",\"M2\",\"NS1\",\"NEP\"))\n",
    "\n",
    "p <- ggplot(data=df2, aes(x=gene, y=value, fill=species)) + \n",
    "    geom_col(position=\"dodge\", stat=\"identity\")+\n",
    "    #geom_errorbar(data=df1, aes(x=gene, ymin=mean-sd, ymax=mean+sd, color=species),position=position_dodge(0.9), stat=\"identity\")+\n",
    "    facet_wrap(~variable, scales=\"free\")+\n",
    "    scale_fill_manual(values=c(\"#99bfaa\",\"#5c3d46\"))+\n",
    "    scale_color_manual(values=c(\"#99bfaa\",\"#5c3d46\"))+\n",
    "    labs(x=\"gene\",y=\"SNPs/site\")+\n",
    "    #ggtitle(\"number of shared sites in random simulations\") + \n",
    "    scale_y_continuous(limits=c(0,0.008))+\n",
    "    #scale_x_continuous(limits=c(0,500))+\n",
    "    theme(plot.title = element_text(size=20, hjust=0.5))+\n",
    "    theme(panel.grid.major.y=element_line(colour=NA))+\n",
    "    theme(panel.grid.minor=element_line(colour=NA,size=NA))+    \n",
    "    theme(strip.background = element_rect(colour=NA, fill=NA))+\n",
    "    theme(axis.line.x=element_line(colour=\"black\"))+\n",
    "    theme(axis.line.y=element_line(colour=\"black\"))+theme(strip.text.x=element_text(size=11))+\n",
    "    theme(axis.title.y=element_text(size=16, vjust=8))+\n",
    "    theme(axis.title.x=element_text(size=16, vjust=-8))+\n",
    "    theme(axis.text=element_text(size=16, colour=\"black\"))+\n",
    "    theme(axis.text.x=element_text(size=16, angle=90))+\n",
    "    theme(legend.text=element_text(size=16))+theme(legend.title=element_text(size=16, face=\"plain\"))+\n",
    "    theme(panel.margin=unit(1, \"lines\"))+theme(plot.margin=unit(c(1,1,1,1),\"cm\"))+\n",
    "    theme(legend.key.size=unit(0.7, \"cm\"))+\n",
    "    theme(panel.background=element_rect(fill=NA))+\n",
    "    theme(legend.key=element_rect(fill=NA))\n",
    "\n",
    "ggsave(\"snps_per_site-2018-05-11.pdf\", p, width = 12, height = 4, device=pdf, path=\"/Users/lmoncla/Documents/H5N1_Cambodian_outbreak_study/paper-and-figure-drafts/figures-2018-05-08\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
