{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running SNPGenie on H5N1 data\n",
    "\n",
    "May 10, 2018 and December 8, 2017\n",
    "\n",
    "Popoolation is not a great program for our purposes. I have known this for awhile, but it has always suited our needs well enough that I haven't looked elsewhere. However, now that I would like to do some comparisons of calculating pi at different SNP frequency cutoffs, I need something else. Chase Nelson's program, SNPGenie does this and so I am going to try to get that to work for me. There seem to be a few quirks with it that I will need to work around: \n",
    "\n",
    "1. SNPGenie only works on 1 gene segment at a time. This is a huge drag. It needs to be run 8 separate times for 1 influenza genome. I also figured out that for each genome, you need to provide a vcf that contains only sites from that gene, a gtf file with only coordinates for that cds, and a reference fasta with only that sequence. \n",
    "\n",
    "2. The vcf format options provided are quite close to what I need, but not quite. Varscan output is almost exactly the same as vcf option 4. The only difference is that SNPGenie assumes AD (allele depth) to be the number of reads supporting the reference base, when in varscan, it supports the number of reads for the alternative allele. Additionally, instead of having the reference base reads in 1 column and the alternative allele reads in another, SNPGenie would like them in the same column, separate by a comma. \n",
    "\n",
    "3. SNPGenie will not over-write old folders. So if I want everything in 1 folder, I will need to make new names for them as we proceed.\n",
    "\n",
    "I will need to do several things in order to make SNPGenie work.\n",
    "\n",
    "1. Fix the AD column issue/reformat my vcfs to accomodate that change. \n",
    "2. Automate splitting up vcf files, reference fasta files, and gtf files. \n",
    "3. Take care of automatic renaming of folders. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    }
   ],
   "source": [
    "# import necessary modules\n",
    "import sys, subprocess, glob, os, shutil, re, importlib, Bio, csv\n",
    "from subprocess import call\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import seaborn as sns\n",
    "import rpy2\n",
    "%load_ext rpy2.ipython "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Fix AD column issue and split vcf by gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "directory = \"seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# function to format VCF files for SNPgenie\n",
    "def format_vcfs_for_snpgenie(vcf_file, directory):\n",
    "    with open(vcf_file, \"r\") as csvfile:\n",
    "        reader = csv.reader(csvfile, delimiter=\"\\t\")\n",
    "        \n",
    "        # get a list of the genes that have SNPs in the original, combined vcf file\n",
    "        genes = []\n",
    "        sample_name = os.path.basename(vcf_file).replace(\".vcf\", \"\")  # Extract the sample name\n",
    "        \n",
    "        for row in reader: \n",
    "            if \"#\" not in row[0] and \"##\" not in row[0]:\n",
    "                gene = row[0]\n",
    "                \n",
    "                if gene.split(\"_\")[-1:] == ['MP']:\n",
    "                    genes.append(gene.replace(\"MP\",\"M1\"))\n",
    "                    genes.append(gene.replace(\"MP\",\"M2\"))\n",
    "                elif gene.split(\"_\")[-1:] == ['NS']:\n",
    "                    genes.append(gene.replace(\"NS\",\"NS1\"))\n",
    "                    genes.append(gene.replace(\"NS\",\"NEP\"))\n",
    "                else:\n",
    "                    genes.append(gene)\n",
    "        genes = set(genes)\n",
    "        \n",
    "        # Create empty output files for each gene and sample name\n",
    "        for gene in genes:\n",
    "            output_filename = f\"{sample_name}_{gene}.vcf\"\n",
    "            with open(os.path.join(directory, output_filename), \"w\") as outfile:\n",
    "                outfile.write(\"\")\n",
    "                \n",
    "    # Re-open and loop through the VCF again to write the data\n",
    "    with open(vcf_file, \"r\") as infile:\n",
    "        for line in infile:\n",
    "            # If it's a comment line, write it out to all of the output files\n",
    "            if \"#\" in line or \"##\" in line:\n",
    "                for gene in genes:\n",
    "                    output_filename = f\"{sample_name}_{gene}.vcf\"\n",
    "                    with open(os.path.join(directory, output_filename), \"a\") as outfile:\n",
    "                        outfile.write(line)\n",
    "            \n",
    "            # If it's a SNP line, add in the AD annotation and write out to the appropriate file\n",
    "            else: \n",
    "                gene = line.split(\"\\t\")[0]\n",
    "                fix = line.split(\"\\t\")[9]\n",
    "                ref_reads = fix.split(\":\")[4]\n",
    "                alt_reads = fix.split(\":\")[5]\n",
    "                new_depth = ref_reads + \",\" + alt_reads\n",
    "                first_part = fix.split(\":\")[0:5]\n",
    "                first_part = \":\".join(first_part)\n",
    "                second_part = fix.split(\":\")[6:14]\n",
    "                second_part = \":\".join(second_part)\n",
    "                fixed = first_part + \":\" + new_depth + \":\" + second_part\n",
    "                \n",
    "                # Combine into a new line\n",
    "                new_line = line.replace(fix, fixed)\n",
    "                \n",
    "                if gene.split(\"_\")[-1:] == ['MP']:\n",
    "                    with open(os.path.join(directory, f\"{sample_name}_{gene.replace('MP', 'M1')}.vcf\"), \"a\") as outfile:\n",
    "                        outfile.write(new_line)\n",
    "                    with open(os.path.join(directory, f\"{sample_name}_{gene.replace('MP', 'M2')}.vcf\"), \"a\") as outfile:\n",
    "                        outfile.write(new_line)\n",
    "                        \n",
    "                elif gene.split(\"_\")[-1:] == ['NS']:\n",
    "                    with open(os.path.join(directory, f\"{sample_name}_{gene.replace('NS', 'NS1')}.vcf\"), \"a\") as outfile:\n",
    "                        outfile.write(new_line)\n",
    "                    with open(os.path.join(directory, f\"{sample_name}_{gene.replace('NS', 'NEP')}.vcf\"), \"a\") as outfile:\n",
    "                        outfile.write(new_line)\n",
    "                  \n",
    "                else:\n",
    "                    with open(os.path.join(directory, f\"{sample_name}_{gene}.vcf\"), \"a\") as outfile:\n",
    "                        outfile.write(new_line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of all VCF files in the directory\n",
    "vcfs = []\n",
    "for f in glob.glob(directory + \"*.vcf\"):\n",
    "    vcfs.append(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/ms_w2_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com6_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/cg_w2_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/cb_com2_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w6_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/g_com2_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/rth_w3_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/gf_w3_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/be_w1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/wws_w1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/g_com1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/cb_com1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com8_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com2_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w8_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w2_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/cg_w1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com5_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/ms_w1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com4_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w3_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com3_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/gf_w1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bo_w1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/be_w3_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w4_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/rth_w1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/blvu_chester_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/gf_w2_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/rth_w2_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w7_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/cb_com3_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/f_w1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/cb_com4_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/tv_w1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w7_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/cb_com3_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/rth_w2_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/gf_w2_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/tv_w1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/cb_com4_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/f_w1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w3_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com4_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/rth_w1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/blvu_chester_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w4_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bo_w1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/be_w3_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com3_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/gf_w1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com2_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com8_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/cb_com1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/g_com1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com5_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/ms_w1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/cg_w1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w2_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w8_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/cg_w2_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/ms_w2_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com6_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/wws_w1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/be_w1_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/gf_w3_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/kc_com1_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/g_com2_replicate-2_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/rth_w3_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/cb_com2_replicate-1_shared.vcf',\n",
       " 'seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/bv_w6_replicate-2_shared.vcf']"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory where you want to store the output files (e.g., parsed VCFs)\n",
    "output_directory = '/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/sample_vcfs/shared_vars_vcf/parsed/'\n",
    "\n",
    "# Run the VCF formatting function on each VCF file\n",
    "for vcf in vcfs: \n",
    "    format_vcfs_for_snpgenie(vcf, output_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['wws_w1_replicate-2_shared_pa', 'blvu_chester_replicate-2_shared_pb2', 'cb_com1_replicate-1_shared_ns', 'ms_w1_replicate-1_shared_pb2', 'tv_w1_replicate-2_shared_pb2', 'f_w1_replicate-2_shared_ha', 'gf_w1_replicate-2_shared_np', 'gf_w3_replicate-1_shared_ha', 'cg_w1_replicate-1_shared_ns', 'cb_com4_replicate-1_shared_pb1', 'f_w1_replicate-2_shared_pb2', 'bv_w7_replicate-1_shared_pb2', 'gf_w1_replicate-2_shared_pb1', 'cb_com2_replicate-2_shared_pb2', 'bv_w6_replicate-2_shared_ha', 'bv_w4_replicate-1_shared_np', 'wws_w1_replicate-1_shared_ha', 'blvu_chester_replicate-2_shared_pb1', 'bv_w3_replicate-1_shared_na', 'kc_com6_replicate-1_shared_pa', 'bv_w1_replicate-2_shared_pa', 'cb_com1_replicate-1_shared_np', 'bv_w3_replicate-2_shared_ns', 'cb_com3_replicate-2_shared_ha', 'ms_w1_replicate-1_shared_pb1', 'gf_w1_replicate-1_shared_na', 'bv_w1_replicate-2_shared_pb2', 'rth_w2_replicate-1_shared_ha', 'bv_w4_replicate-2_shared_na', 'kc_com1_replicate-2_shared_pa', 'kc_com3_replicate-2_shared_ns', 'gf_w1_replicate-2_shared_pb2', 'cb_com4_replicate-1_shared_ha', 'cb_com2_replicate-2_shared_pb1', 'kc_com3_replicate-1_shared_na', 'bv_w4_replicate-1_shared_ns', 'cb_com2_replicate-1_shared_m', 'bv_w6_replicate-1_shared_pa', 'cb_com3_replicate-1_shared_ha', 'bv_w1_replicate-1_shared_pa', 'bv_w3_replicate-1_shared_ns', 'cb_com1_replicate-2_shared_np', 'kc_com6_replicate-2_shared_pa', 'bv_w3_replicate-2_shared_na', 'wws_w1_replicate-2_shared_ha', 'rth_w1_replicate-1_shared_pb1', 'gf_w1_replicate-2_shared_na', 'bv_w6_replicate-2_shared_pb2', 'cb_com3_replicate-1_shared_pb2', 'kc_com6_replicate-1_shared_pb1', 'rth_w2_replicate-2_shared_ha', 'bv_w4_replicate-2_shared_ns', 'bv_w6_replicate-2_shared_pa', 'kc_com3_replicate-2_shared_na', 'cb_com4_replicate-2_shared_ha', 'kc_com1_replicate-1_shared_pa', 'cg_w2_replicate-1_shared_pb1', 'kc_com3_replicate-1_shared_ns', 'bv_w4_replicate-1_shared_na', 'cb_com1_replicate-2_shared_ns', 'wws_w1_replicate-1_shared_pa', 'rth_w1_replicate-1_shared_pb2', 'bv_w6_replicate-2_shared_pb1', 'kc_com6_replicate-1_shared_pb2', 'cb_com3_replicate-1_shared_pb1', 'gf_w3_replicate-2_shared_ha', 'gf_w1_replicate-1_shared_np', 'f_w1_replicate-1_shared_ha', 'cg_w1_replicate-2_shared_ns', 'bv_w4_replicate-2_shared_np', 'bv_w6_replicate-1_shared_ha', 'cg_w2_replicate-1_shared_pb2', 'cb_com1_replicate-1_shared_pa', 'wws_w1_replicate-1_shared_na', 'bv_w7_replicate-2_shared_pb2', 'bv_w3_replicate-1_shared_ha', 'f_w1_replicate-1_shared_pb2', 'cb_com2_replicate-1_shared_pb2', 'gf_w1_replicate-1_shared_pb1', 'wws_w1_replicate-2_shared_ns', 'gf_w1_replicate-1_shared_ha', 'cb_com4_replicate-2_shared_pb1', 'tv_w1_replicate-1_shared_pb2', 'rth_w2_replicate-1_shared_na', 'bv_w4_replicate-2_shared_ha', 'bv_w6_replicate-1_shared_np', 'blvu_chester_replicate-1_shared_pb2', 'ms_w1_replicate-2_shared_pb2', 'bv_w1_replicate-2_shared_ns', 'bv_w3_replicate-2_shared_pa', 'kc_com6_replicate-2_shared_na', 'cb_com1_replicate-2_shared_ha', 'cb_com2_replicate-1_shared_pb1', 'gf_w1_replicate-1_shared_pb2', 'kc_com4_replicate-1_shared_pa', 'kc_com6_replicate-1_shared_ns', 'wws_w1_replicate-2_shared_np', 'bv_w1_replicate-1_shared_pb2', 'bv_w4_replicate-1_shared_pa', 'bv_w6_replicate-1_shared_ns', 'blvu_chester_replicate-1_shared_pb1', 'kc_com3_replicate-2_shared_pa', 'bv_w6_replicate-2_shared_na', 'g_com2_replicate-1_shared_pa', 'ms_w1_replicate-2_shared_pb1', 'wws_w1_replicate-1_shared_np', 'kc_com4_replicate-2_shared_pa', 'kc_com6_replicate-2_shared_ns', 'cb_com1_replicate-1_shared_ha', 'kc_com6_replicate-1_shared_na', 'bv_w1_replicate-1_shared_ns', 'bv_w3_replicate-1_shared_pa', 'cg_w2_replicate-2_shared_pb1', 'rth_w1_replicate-2_shared_pb1', 'bv_w6_replicate-1_shared_pb2', 'kc_com6_replicate-2_shared_pb1', 'cb_com3_replicate-2_shared_pb2', 'g_com2_replicate-2_shared_pa', 'bv_w6_replicate-1_shared_na', 'kc_com3_replicate-1_shared_pa', 'bv_w4_replicate-2_shared_pa', 'bv_w6_replicate-2_shared_ns', 'wws_w1_replicate-1_shared_ns', 'bv_w3_replicate-2_shared_ha', 'wws_w1_replicate-2_shared_na', 'cg_w2_replicate-2_shared_pb2', 'cb_com1_replicate-2_shared_pa', 'gf_w1_replicate-2_shared_ha', 'blvu_chester_replicate-2_shared_m', 'rth_w1_replicate-2_shared_pb2', 'rth_w2_replicate-2_shared_na', 'bv_w6_replicate-1_shared_pb1', 'cb_com3_replicate-2_shared_pb1', 'kc_com6_replicate-2_shared_pb2', 'bv_w6_replicate-2_shared_np', 'bv_w4_replicate-1_shared_ha', 'g_com2_replicate-1_shared_pb2', 'rth_w3_replicate-2_shared_na', 'bv_w8_replicate-1_shared_np', 'rth_w3_replicate-1_shared_ns', 'cg_w1_replicate-1_shared_pb1', 'ms_w2_replicate-1_shared_np', 'cb_com2_replicate-1_shared_na', 'gf_w3_replicate-1_shared_pb1', 'cb_com2_replicate-2_shared_ns', 'kc_com5_replicate-1_shared_pb1', 'bv_w2_replicate-2_shared_ha', 'gf_w2_replicate-1_shared_np', 'be_w3_replicate-1_shared_pa', 'kc_com3_replicate-2_shared_pb2', 'blvu_chester_replicate-2_shared_ha', 'bv_w8_replicate-1_shared_ns', 'bv_w8_replicate-2_shared_na', 'rth_w3_replicate-1_shared_np', 'rth_w1_replicate-2_shared_ha', 'cg_w1_replicate-1_shared_pb2', 'kc_com2_replicate-1_shared_pa', 'bv_w7_replicate-1_shared_na', 'ms_w2_replicate-1_shared_ns', 'rth_w2_replicate-1_shared_pb2', 'kc_com5_replicate-2_shared_pa', 'cb_com2_replicate-2_shared_np', 'gf_w3_replicate-1_shared_pb2', 'bv_w2_replicate-1_shared_pa', 'kc_com5_replicate-1_shared_pb2', 'bv_w3_replicate-1_shared_pb2', 'be_w3_replicate-2_shared_ha', 'be_w1_replicate-1_shared_np', 'blvu_chester_replicate-1_shared_pa', 'kc_com3_replicate-2_shared_pb1', 'gf_w2_replicate-2_shared_na', 'rth_w1_replicate-1_shared_ha', 'bv_w4_replicate-1_shared_pb2', 'rth_w3_replicate-2_shared_np', 'bv_w8_replicate-1_shared_na', 'kc_com8_replicate-2_shared_pb1', 'cb_com1_replicate-2_shared_pb2', 'kc_com4_replicate-2_shared_pb1', 'gf_w2_replicate-2_shared_pb1', 'bv_w8_replicate-1_shared_pb2', 'bv_w8_replicate-2_shared_ns', 'ms_w2_replicate-2_shared_ns', 'wws_w1_replicate-2_shared_pb1', 'bv_w7_replicate-2_shared_na', 'kc_com2_replicate-1_shared_pb2', 'kc_com2_replicate-2_shared_pa', 'be_w3_replicate-1_shared_pb2', 'cb_com2_replicate-1_shared_np', 'bv_w2_replicate-2_shared_pa', 'kc_com5_replicate-1_shared_pa', 'gf_w2_replicate-1_shared_na', 'blvu_chester_replicate-2_shared_pa', 'be_w1_replicate-2_shared_np', 'ms_w2_replicate-1_shared_pb2', 'be_w3_replicate-1_shared_ha', 'rth_w3_replicate-2_shared_pb2', 'bv_w4_replicate-1_shared_pb1', 'rth_w3_replicate-2_shared_ns', 'cb_com1_replicate-2_shared_pb1', 'bv_w8_replicate-2_shared_np', 'bv_w8_replicate-1_shared_pb1', 'rth_w3_replicate-1_shared_na', 'gf_w2_replicate-2_shared_pb2', 'ms_w2_replicate-2_shared_np', 'wws_w1_replicate-2_shared_pb2', 'bv_w2_replicate-1_shared_ha', 'be_w3_replicate-1_shared_pb1', 'cb_com2_replicate-1_shared_ns', 'cb_com2_replicate-2_shared_na', 'blvu_chester_replicate-1_shared_m', 'blvu_chester_replicate-1_shared_ha', 'be_w3_replicate-2_shared_pa', 'gf_w2_replicate-2_shared_np', 'cg_w2_replicate-2_shared_pa', 'rth_w1_replicate-1_shared_ns', 'bv_w8_replicate-2_shared_ha', 'kc_com3_replicate-1_shared_pb2', 'rth_w1_replicate-2_shared_na', 'kc_com5_replicate-2_shared_pb1', 'kc_com2_replicate-1_shared_np', 'gf_w3_replicate-2_shared_pb1', 'bv_w2_replicate-1_shared_np', 'kc_com5_replicate-2_shared_np', 'g_com2_replicate-2_shared_pb2', 'be_w3_replicate-2_shared_na', 'cg_w1_replicate-2_shared_pb1', 'be_w1_replicate-1_shared_pa', 'be_w3_replicate-1_shared_ns', 'gf_w2_replicate-2_shared_ha', 'bv_w3_replicate-2_shared_pb2', 'cb_com2_replicate-2_shared_m', 'bv_w8_replicate-1_shared_pa', 'kc_com3_replicate-1_shared_pb1', 'rth_w2_replicate-2_shared_pb2', 'ms_w2_replicate-1_shared_pa', 'bv_w7_replicate-2_shared_pa', 'kc_com5_replicate-2_shared_pb2', 'gf_w3_replicate-2_shared_pb2', 'g_com1_replicate-2_shared_pa', 'kc_com5_replicate-1_shared_na', 'cb_com2_replicate-1_shared_ha', 'bv_w2_replicate-2_shared_na', 'kc_com5_replicate-2_shared_ns', 'blvu_chester_replicate-1_shared_ns', 'be_w1_replicate-2_shared_ha', 'cg_w1_replicate-2_shared_pb2', 'be_w3_replicate-1_shared_np', 'blvu_chester_replicate-2_shared_na', 'bv_w8_replicate-2_shared_pa', 'ms_w2_replicate-2_shared_pb2', 'g_com1_replicate-1_shared_pa', 'be_w3_replicate-2_shared_pb2', 'bv_w7_replicate-1_shared_pa', 'ms_w2_replicate-2_shared_pa', 'kc_com5_replicate-1_shared_ns', 'bv_w2_replicate-1_shared_na', 'wws_w1_replicate-1_shared_pb1', 'cb_com2_replicate-2_shared_ha', 'kc_com2_replicate-2_shared_pb2', 'kc_com5_replicate-2_shared_na', 'blvu_chester_replicate-1_shared_na', 'be_w3_replicate-2_shared_np', 'kc_com8_replicate-1_shared_pb1', 'be_w1_replicate-1_shared_ha', 'bv_w4_replicate-2_shared_pb2', 'bv_w8_replicate-2_shared_pb2', 'gf_w2_replicate-1_shared_pb1', 'kc_com4_replicate-1_shared_pb1', 'cb_com1_replicate-1_shared_pb2', 'blvu_chester_replicate-2_shared_ns', 'rth_w1_replicate-1_shared_na', 'bv_w8_replicate-1_shared_ha', 'cg_w2_replicate-1_shared_pa', 'rth_w1_replicate-2_shared_ns', 'be_w3_replicate-2_shared_pb1', 'kc_com2_replicate-2_shared_np', 'kc_com5_replicate-1_shared_np', 'wws_w1_replicate-1_shared_pb2', 'bv_w2_replicate-2_shared_np', 'be_w1_replicate-2_shared_pa', 'rth_w3_replicate-1_shared_pb2', 'gf_w2_replicate-1_shared_ha', 'be_w3_replicate-2_shared_ns', 'bv_w4_replicate-2_shared_pb1', 'gf_w2_replicate-1_shared_pb2', 'bv_w8_replicate-2_shared_pb1', 'cb_com1_replicate-1_shared_pb1', 'be_w3_replicate-1_shared_na']\n"
     ]
    }
   ],
   "source": [
    "##get file list for everytthing\n",
    "\n",
    "file_list = []\n",
    "\n",
    "# Search for all VCF files within subdirectories of output_directory\n",
    "for f in glob.glob(os.path.join(output_directory, \"**\", \"*.vcf\"), recursive=True):\n",
    "    # Extract the filename without extension\n",
    "    sample_name = os.path.basename(f).replace(\".vcf\", \"\")\n",
    "    file_list.append(sample_name)\n",
    "\n",
    "print(file_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ns: reference.gb → ns.gb\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ns: reference.gtf → ns.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ns: reference.fasta → ns.fasta\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/na: reference.gb → na.gb\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/na: reference.gtf → na.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/na: reference.fasta → na.fasta\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb2: reference.gb → pb2.gb\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb2: reference.gtf → pb2.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb2: reference.fasta → pb2.fasta\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pa: reference.gb → pa.gb\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pa: reference.gtf → pa.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pa: reference.fasta → pa.fasta\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ha: reference.gtf → ha.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ha: reference.fasta → ha.fasta\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/np: reference.gb → np.gb\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/np: reference.gtf → np.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/np: reference.fasta → np.fasta\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb1: reference.gb → pb1.gb\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb1: reference.gtf → pb1.gtf\n",
      "Renamed in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb1: reference.fasta → pb1.fasta\n"
     ]
    }
   ],
   "source": [
    "##format gtfs and fastas\n",
    "\n",
    "##renaming reference gtfs and fastas\n",
    "base_directory = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference\"  # Change this to your main directory\n",
    "\n",
    "for root, _, files in os.walk(base_directory):\n",
    "    subdir_name = os.path.basename(root)  # Get the subdirectory name\n",
    "\n",
    "    for filename in files:\n",
    "        if filename.startswith(\".\"):  # Ignore hidden/system files\n",
    "            continue\n",
    "\n",
    "        file_extension = os.path.splitext(filename)[-1]  # Extract file extension\n",
    "        \n",
    "        if \"reference\" in filename:\n",
    "            new_name = filename.replace(\"reference\", subdir_name)  # Replace 'reference' with subdirectory name\n",
    "            old_path = os.path.join(root, filename)\n",
    "            new_path = os.path.join(root, new_name)\n",
    "\n",
    "            os.rename(old_path, new_path)\n",
    "            print(f\"Renamed in {root}: {filename} → {new_name}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ns/ns.fasta → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ns.fasta\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/na/na.fasta → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/na.fasta\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb2/pb2.fasta → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb2.fasta\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pa/pa.fasta → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pa.fasta\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ha/ha.fasta → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ha.fasta\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/np/np.fasta → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/np.fasta\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb1/pb1.fasta → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb1.fasta\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ns/ns.gtf → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ns.gtf\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/na/na.gtf → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/na.gtf\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb2/pb2.gtf → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb2.gtf\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pa/pa.gtf → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pa.gtf\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ha/ha.gtf → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/ha.gtf\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/np/np.gtf → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/np.gtf\n",
      "Moved /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb1/pb1.gtf → /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/pb1.gtf\n",
      "All .fasta and .gtf files have been moved to the base directory.\n"
     ]
    }
   ],
   "source": [
    "# Define the base directory where subdirectories are located\n",
    "base_directory = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/reference/\"\n",
    "\n",
    "# Find all .fasta and .gtf files within subdirectories\n",
    "fasta_files = glob.glob(os.path.join(base_directory, \"*\", \"*.fasta\"))\n",
    "gtf_files = glob.glob(os.path.join(base_directory, \"*\", \"*.gtf\"))\n",
    "\n",
    "# Move files to the base directory\n",
    "for file_path in fasta_files + gtf_files:\n",
    "    filename = os.path.basename(file_path)  # Get the file name\n",
    "    destination = os.path.join(base_directory, filename)  # Define new path\n",
    "\n",
    "    # Move file to base directory\n",
    "    shutil.move(file_path, destination)\n",
    "    print(f\"Moved {file_path} → {destination}\")\n",
    "\n",
    "print(\"All .fasta and .gtf files have been moved to the base directory.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running SNPGenie for be_w2_replicate-1_pa...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "Results for be_w2_replicate-1_pa saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w2_replicate-1_pa_results/\n",
      "Running SNPGenie for be_w1_replicate-1_ha...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene PA-X in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Creating temp_vcf4_Sample1.vcf...\n",
      "\n",
      "\n",
      "###########################  CURRENTLY PROCESSING:   ###########################\n",
      "temp_vcf4_Sample1.vcf... VCF format detected\n",
      "\n",
      "Converting temp_vcf4_Sample1.vcf to SNPGenie format...\n",
      "\n",
      "In file temp_vcf4_Sample1.vcf, the newline type is: Unix (LF, \\n)\n",
      "\n",
      "Calculating and storing protein-coding genome and variant data (that dogma stuff)... COMPLETED.\n",
      "\n",
      "Ligating gene segments (computationally, of course!)... COMPLETED.\n",
      "\n",
      "Calculating and storing non-protein-coding genome and variant data (\"junk\" gets a bad rap!)... COMPLETED.\n",
      "\n",
      "Processing all individual sites (nucleotides take time and memory; use a high-memory machine for eukaryotic chromosomes!)... COMPLETED.\n",
      "\n",
      "Processing population genetic estimates codon-by-codon (beware stochasticity!)... \n",
      "## WARNING: In temp_vcf4_Sample1.vcf, HA, site 151,\n",
      "## the reference nucleotide in the SNP Report does not match the FASTA file.\n",
      "## Results at this site are unreliable: proceed with caution, and consider\n",
      "## re-calling SNPs (or fixing the SNP Reports) to address the issue.\n",
      "\n",
      "## WARNING: In temp_vcf4_Sample1.vcf, HA, site 1127,\n",
      "## the reference nucleotide in the SNP Report does not match the FASTA file.\n",
      "## Results at this site are unreliable: proceed with caution, and consider\n",
      "## re-calling SNPs (or fixing the SNP Reports) to address the issue.\n",
      "COMPLETED.\n",
      "\n",
      "Performing final calculations, noncoding overlap analysis, and output... temp_vcf4_Sample1.vcf COMPLETED.\n",
      "\n",
      "################################################################################\n",
      "##                      SNPGenie completed successfully.                      ##\n",
      "##             Please find results in the SNPGenie_Results folder.            ##\n",
      "################################################################################\n",
      "\n",
      "\n",
      "Results for be_w1_replicate-1_ha saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-1_ha_results/\n",
      "Running SNPGenie for be_w2_replicate-1_ha...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Creating temp_vcf4_Sample1.vcf...\n",
      "\n",
      "\n",
      "###########################  CURRENTLY PROCESSING:   ###########################\n",
      "temp_vcf4_Sample1.vcf... VCF format detected\n",
      "\n",
      "Converting temp_vcf4_Sample1.vcf to SNPGenie format...\n",
      "\n",
      "In file temp_vcf4_Sample1.vcf, the newline type is: Unix (LF, \\n)\n",
      "\n",
      "Calculating and storing protein-coding genome and variant data (that dogma stuff)... COMPLETED.\n",
      "\n",
      "Ligating gene segments (computationally, of course!)... COMPLETED.\n",
      "\n",
      "Calculating and storing non-protein-coding genome and variant data (\"junk\" gets a bad rap!)... COMPLETED.\n",
      "\n",
      "Processing all individual sites (nucleotides take time and memory; use a high-memory machine for eukaryotic chromosomes!)... COMPLETED.\n",
      "\n",
      "Processing population genetic estimates codon-by-codon (beware stochasticity!)... \n",
      "## WARNING: In temp_vcf4_Sample1.vcf, HA, site 905,\n",
      "## the reference nucleotide in the SNP Report does not match the FASTA file.\n",
      "## Results at this site are unreliable: proceed with caution, and consider\n",
      "## re-calling SNPs (or fixing the SNP Reports) to address the issue.\n",
      "COMPLETED.\n",
      "\n",
      "Performing final calculations, noncoding overlap analysis, and output... temp_vcf4_Sample1.vcf COMPLETED.\n",
      "\n",
      "################################################################################\n",
      "##                      SNPGenie completed successfully.                      ##\n",
      "##             Please find results in the SNPGenie_Results folder.            ##\n",
      "################################################################################\n",
      "\n",
      "\n",
      "Results for be_w2_replicate-1_ha saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w2_replicate-1_ha_results/\n",
      "Running SNPGenie for be_w1_replicate-1_pa...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "Results for be_w1_replicate-1_pa saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-1_pa_results/\n",
      "Running SNPGenie for be_w2_replicate-1_ns...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene PA-X in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene NEP in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "Results for be_w2_replicate-1_ns saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w2_replicate-1_ns_results/\n",
      "Running SNPGenie for be_w1_replicate-1_na...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Creating temp_vcf4_Sample1.vcf...\n",
      "\n",
      "\n",
      "###########################  CURRENTLY PROCESSING:   ###########################\n",
      "temp_vcf4_Sample1.vcf... VCF format detected\n",
      "\n",
      "Converting temp_vcf4_Sample1.vcf to SNPGenie format...\n",
      "\n",
      "In file temp_vcf4_Sample1.vcf, the newline type is: Unix (LF, \\n)\n",
      "\n",
      "Calculating and storing protein-coding genome and variant data (that dogma stuff)... COMPLETED.\n",
      "\n",
      "Ligating gene segments (computationally, of course!)... COMPLETED.\n",
      "\n",
      "Calculating and storing non-protein-coding genome and variant data (\"junk\" gets a bad rap!)... COMPLETED.\n",
      "\n",
      "Processing all individual sites (nucleotides take time and memory; use a high-memory machine for eukaryotic chromosomes!)... COMPLETED.\n",
      "\n",
      "Processing population genetic estimates codon-by-codon (beware stochasticity!)... COMPLETED.\n",
      "\n",
      "Performing final calculations, noncoding overlap analysis, and output... temp_vcf4_Sample1.vcf COMPLETED.\n",
      "\n",
      "################################################################################\n",
      "##                      SNPGenie completed successfully.                      ##\n",
      "##             Please find results in the SNPGenie_Results folder.            ##\n",
      "################################################################################\n",
      "\n",
      "\n",
      "Results for be_w1_replicate-1_na saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-1_na_results/\n",
      "Running SNPGenie for be_w2_replicate-1_np...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "Results for be_w2_replicate-1_np saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w2_replicate-1_np_results/\n",
      "Running SNPGenie for be_w1_replicate-1_np...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene NP in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene NP in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "Results for be_w1_replicate-1_np saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-1_np_results/\n",
      "Running SNPGenie for be_w2_replicate-1_na...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Creating temp_vcf4_Sample1.vcf...\n",
      "\n",
      "\n",
      "###########################  CURRENTLY PROCESSING:   ###########################\n",
      "temp_vcf4_Sample1.vcf... VCF format detected\n",
      "\n",
      "Converting temp_vcf4_Sample1.vcf to SNPGenie format...\n",
      "\n",
      "In file temp_vcf4_Sample1.vcf, the newline type is: Unix (LF, \\n)\n",
      "\n",
      "Calculating and storing protein-coding genome and variant data (that dogma stuff)... COMPLETED.\n",
      "\n",
      "Ligating gene segments (computationally, of course!)... COMPLETED.\n",
      "\n",
      "Calculating and storing non-protein-coding genome and variant data (\"junk\" gets a bad rap!)... COMPLETED.\n",
      "\n",
      "Processing all individual sites (nucleotides take time and memory; use a high-memory machine for eukaryotic chromosomes!)... COMPLETED.\n",
      "\n",
      "Processing population genetic estimates codon-by-codon (beware stochasticity!)... COMPLETED.\n",
      "\n",
      "Performing final calculations, noncoding overlap analysis, and output... temp_vcf4_Sample1.vcf COMPLETED.\n",
      "\n",
      "################################################################################\n",
      "##                      SNPGenie completed successfully.                      ##\n",
      "##             Please find results in the SNPGenie_Results folder.            ##\n",
      "################################################################################\n",
      "\n",
      "\n",
      "Results for be_w2_replicate-1_na saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w2_replicate-1_na_results/\n",
      "Running SNPGenie for be_w1_replicate-1_ns...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "Results for be_w1_replicate-1_ns saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-1_ns_results/\n",
      "Running SNPGenie for be_w1_replicate-2_pa...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene NEP in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene PA-X in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "Results for be_w1_replicate-2_pa saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-2_pa_results/\n",
      "Running SNPGenie for be_w1_replicate-2_pb1...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Creating temp_vcf4_Sample1.vcf...\n",
      "\n",
      "\n",
      "###########################  CURRENTLY PROCESSING:   ###########################\n",
      "temp_vcf4_Sample1.vcf... VCF format detected\n",
      "\n",
      "Converting temp_vcf4_Sample1.vcf to SNPGenie format...\n",
      "\n",
      "In file temp_vcf4_Sample1.vcf, the newline type is: Unix (LF, \\n)\n",
      "\n",
      "Calculating and storing protein-coding genome and variant data (that dogma stuff)... COMPLETED.\n",
      "\n",
      "Ligating gene segments (computationally, of course!)... COMPLETED.\n",
      "\n",
      "Calculating and storing non-protein-coding genome and variant data (\"junk\" gets a bad rap!)... COMPLETED.\n",
      "\n",
      "Processing all individual sites (nucleotides take time and memory; use a high-memory machine for eukaryotic chromosomes!)... COMPLETED.\n",
      "\n",
      "Processing population genetic estimates codon-by-codon (beware stochasticity!)... COMPLETED.\n",
      "\n",
      "Performing final calculations, noncoding overlap analysis, and output... temp_vcf4_Sample1.vcf COMPLETED.\n",
      "\n",
      "################################################################################\n",
      "##                      SNPGenie completed successfully.                      ##\n",
      "##             Please find results in the SNPGenie_Results folder.            ##\n",
      "################################################################################\n",
      "\n",
      "\n",
      "Results for be_w1_replicate-2_pb1 saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-2_pb1_results/\n",
      "Running SNPGenie for be_w1_replicate-2_ha...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Creating temp_vcf4_Sample1.vcf...\n",
      "\n",
      "\n",
      "###########################  CURRENTLY PROCESSING:   ###########################\n",
      "temp_vcf4_Sample1.vcf... VCF format detected\n",
      "\n",
      "Converting temp_vcf4_Sample1.vcf to SNPGenie format...\n",
      "\n",
      "In file temp_vcf4_Sample1.vcf, the newline type is: Unix (LF, \\n)\n",
      "\n",
      "Calculating and storing protein-coding genome and variant data (that dogma stuff)... COMPLETED.\n",
      "\n",
      "Ligating gene segments (computationally, of course!)... COMPLETED.\n",
      "\n",
      "Calculating and storing non-protein-coding genome and variant data (\"junk\" gets a bad rap!)... COMPLETED.\n",
      "\n",
      "Processing all individual sites (nucleotides take time and memory; use a high-memory machine for eukaryotic chromosomes!)... COMPLETED.\n",
      "\n",
      "Processing population genetic estimates codon-by-codon (beware stochasticity!)... COMPLETED.\n",
      "\n",
      "Performing final calculations, noncoding overlap analysis, and output... temp_vcf4_Sample1.vcf COMPLETED.\n",
      "\n",
      "################################################################################\n",
      "##                      SNPGenie completed successfully.                      ##\n",
      "##             Please find results in the SNPGenie_Results folder.            ##\n",
      "################################################################################\n",
      "\n",
      "\n",
      "Results for be_w1_replicate-2_ha saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-2_ha_results/\n",
      "Running SNPGenie for be_w1_replicate-2_pb2...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Creating temp_vcf4_Sample1.vcf...\n",
      "\n",
      "\n",
      "###########################  CURRENTLY PROCESSING:   ###########################\n",
      "temp_vcf4_Sample1.vcf... VCF format detected\n",
      "\n",
      "Converting temp_vcf4_Sample1.vcf to SNPGenie format...\n",
      "\n",
      "In file temp_vcf4_Sample1.vcf, the newline type is: Unix (LF, \\n)\n",
      "\n",
      "Calculating and storing protein-coding genome and variant data (that dogma stuff)... COMPLETED.\n",
      "\n",
      "Ligating gene segments (computationally, of course!)... COMPLETED.\n",
      "\n",
      "Calculating and storing non-protein-coding genome and variant data (\"junk\" gets a bad rap!)... COMPLETED.\n",
      "\n",
      "Processing all individual sites (nucleotides take time and memory; use a high-memory machine for eukaryotic chromosomes!)... COMPLETED.\n",
      "\n",
      "Processing population genetic estimates codon-by-codon (beware stochasticity!)... \n",
      "## WARNING: In temp_vcf4_Sample1.vcf, PB2, site 924,\n",
      "## the reference nucleotide in the SNP Report does not match the FASTA file.\n",
      "## Results at this site are unreliable: proceed with caution, and consider\n",
      "## re-calling SNPs (or fixing the SNP Reports) to address the issue.\n",
      "COMPLETED.\n",
      "\n",
      "Performing final calculations, noncoding overlap analysis, and output... temp_vcf4_Sample1.vcf COMPLETED.\n",
      "\n",
      "################################################################################\n",
      "##                      SNPGenie completed successfully.                      ##\n",
      "##             Please find results in the SNPGenie_Results folder.            ##\n",
      "################################################################################\n",
      "\n",
      "\n",
      "Results for be_w1_replicate-2_pb2 saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-2_pb2_results/\n",
      "Running SNPGenie for be_w1_replicate-2_np...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "Results for be_w1_replicate-2_np saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-2_np_results/\n",
      "Running SNPGenie for be_w1_replicate-2_ns...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene NP in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n",
      "\n",
      "\n",
      "## WARNING: The CDS coordinates for gene NEP in the gtf file do not yield a set of complete codons,\n",
      "## or are absent from the file. The number of nucleotides must be a multiple of 3.\n",
      "## SNPGenie terminated.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "Results for be_w1_replicate-2_ns saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-2_ns_results/\n",
      "Running SNPGenie for be_w1_replicate-2_na...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Creating temp_vcf4_Sample1.vcf...\n",
      "\n",
      "\n",
      "###########################  CURRENTLY PROCESSING:   ###########################\n",
      "temp_vcf4_Sample1.vcf... VCF format detected\n",
      "\n",
      "Converting temp_vcf4_Sample1.vcf to SNPGenie format...\n",
      "\n",
      "In file temp_vcf4_Sample1.vcf, the newline type is: Unix (LF, \\n)\n",
      "\n",
      "Calculating and storing protein-coding genome and variant data (that dogma stuff)... COMPLETED.\n",
      "\n",
      "Ligating gene segments (computationally, of course!)... COMPLETED.\n",
      "\n",
      "Calculating and storing non-protein-coding genome and variant data (\"junk\" gets a bad rap!)... COMPLETED.\n",
      "\n",
      "Processing all individual sites (nucleotides take time and memory; use a high-memory machine for eukaryotic chromosomes!)... COMPLETED.\n",
      "\n",
      "Processing population genetic estimates codon-by-codon (beware stochasticity!)... COMPLETED.\n",
      "\n",
      "Performing final calculations, noncoding overlap analysis, and output... temp_vcf4_Sample1.vcf COMPLETED.\n",
      "\n",
      "################################################################################\n",
      "##                      SNPGenie completed successfully.                      ##\n",
      "##             Please find results in the SNPGenie_Results folder.            ##\n",
      "################################################################################\n",
      "\n",
      "\n",
      "Results for be_w1_replicate-2_na saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-2_na_results/\n",
      "Running SNPGenie for be_w2_replicate-1_pb1...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Creating temp_vcf4_Sample1.vcf...\n",
      "\n",
      "\n",
      "###########################  CURRENTLY PROCESSING:   ###########################\n",
      "temp_vcf4_Sample1.vcf... VCF format detected\n",
      "\n",
      "Converting temp_vcf4_Sample1.vcf to SNPGenie format...\n",
      "\n",
      "In file temp_vcf4_Sample1.vcf, the newline type is: Unix (LF, \\n)\n",
      "\n",
      "Calculating and storing protein-coding genome and variant data (that dogma stuff)... COMPLETED.\n",
      "\n",
      "Ligating gene segments (computationally, of course!)... COMPLETED.\n",
      "\n",
      "Calculating and storing non-protein-coding genome and variant data (\"junk\" gets a bad rap!)... COMPLETED.\n",
      "\n",
      "Processing all individual sites (nucleotides take time and memory; use a high-memory machine for eukaryotic chromosomes!)... COMPLETED.\n",
      "\n",
      "Processing population genetic estimates codon-by-codon (beware stochasticity!)... COMPLETED.\n",
      "\n",
      "Performing final calculations, noncoding overlap analysis, and output... temp_vcf4_Sample1.vcf COMPLETED.\n",
      "\n",
      "################################################################################\n",
      "##                      SNPGenie completed successfully.                      ##\n",
      "##             Please find results in the SNPGenie_Results folder.            ##\n",
      "################################################################################\n",
      "\n",
      "\n",
      "Results for be_w2_replicate-1_pb1 saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w2_replicate-1_pb1_results/\n",
      "Running SNPGenie for be_w1_replicate-1_pb2...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Creating temp_vcf4_Sample1.vcf...\n",
      "\n",
      "\n",
      "###########################  CURRENTLY PROCESSING:   ###########################\n",
      "temp_vcf4_Sample1.vcf... VCF format detected\n",
      "\n",
      "Converting temp_vcf4_Sample1.vcf to SNPGenie format...\n",
      "\n",
      "In file temp_vcf4_Sample1.vcf, the newline type is: Unix (LF, \\n)\n",
      "\n",
      "Calculating and storing protein-coding genome and variant data (that dogma stuff)... COMPLETED.\n",
      "\n",
      "Ligating gene segments (computationally, of course!)... COMPLETED.\n",
      "\n",
      "Calculating and storing non-protein-coding genome and variant data (\"junk\" gets a bad rap!)... COMPLETED.\n",
      "\n",
      "Processing all individual sites (nucleotides take time and memory; use a high-memory machine for eukaryotic chromosomes!)... COMPLETED.\n",
      "\n",
      "Processing population genetic estimates codon-by-codon (beware stochasticity!)... \n",
      "## WARNING: In temp_vcf4_Sample1.vcf, PB2, site 408,\n",
      "## the reference nucleotide in the SNP Report does not match the FASTA file.\n",
      "## Results at this site are unreliable: proceed with caution, and consider\n",
      "## re-calling SNPs (or fixing the SNP Reports) to address the issue.\n",
      "\n",
      "## WARNING: In temp_vcf4_Sample1.vcf, PB2, site 639,\n",
      "## the reference nucleotide in the SNP Report does not match the FASTA file.\n",
      "## Results at this site are unreliable: proceed with caution, and consider\n",
      "## re-calling SNPs (or fixing the SNP Reports) to address the issue.\n",
      "\n",
      "## WARNING: In temp_vcf4_Sample1.vcf, PB2, site 786,\n",
      "## the reference nucleotide in the SNP Report does not match the FASTA file.\n",
      "## Results at this site are unreliable: proceed with caution, and consider\n",
      "## re-calling SNPs (or fixing the SNP Reports) to address the issue.\n",
      "\n",
      "## WARNING: In temp_vcf4_Sample1.vcf, PB2, site 939,\n",
      "## the reference nucleotide in the SNP Report does not match the FASTA file.\n",
      "## Results at this site are unreliable: proceed with caution, and consider\n",
      "## re-calling SNPs (or fixing the SNP Reports) to address the issue.\n",
      "\n",
      "## WARNING: In temp_vcf4_Sample1.vcf, PB2, site 1050,\n",
      "## the reference nucleotide in the SNP Report does not match the FASTA file.\n",
      "## Results at this site are unreliable: proceed with caution, and consider\n",
      "## re-calling SNPs (or fixing the SNP Reports) to address the issue.\n",
      "\n",
      "## WARNING: In temp_vcf4_Sample1.vcf, PB2, site 2010,\n",
      "## the reference nucleotide in the SNP Report does not match the FASTA file.\n",
      "## Results at this site are unreliable: proceed with caution, and consider\n",
      "## re-calling SNPs (or fixing the SNP Reports) to address the issue.\n",
      "COMPLETED.\n",
      "\n",
      "Performing final calculations, noncoding overlap analysis, and output... temp_vcf4_Sample1.vcf COMPLETED.\n",
      "\n",
      "################################################################################\n",
      "##                      SNPGenie completed successfully.                      ##\n",
      "##             Please find results in the SNPGenie_Results folder.            ##\n",
      "################################################################################\n",
      "\n",
      "\n",
      "Results for be_w1_replicate-1_pb2 saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-1_pb2_results/\n",
      "Running SNPGenie for be_w2_replicate-1_pb2...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Creating temp_vcf4_Sample1.vcf...\n",
      "\n",
      "\n",
      "###########################  CURRENTLY PROCESSING:   ###########################\n",
      "temp_vcf4_Sample1.vcf... VCF format detected\n",
      "\n",
      "Converting temp_vcf4_Sample1.vcf to SNPGenie format...\n",
      "\n",
      "In file temp_vcf4_Sample1.vcf, the newline type is: Unix (LF, \\n)\n",
      "\n",
      "Calculating and storing protein-coding genome and variant data (that dogma stuff)... COMPLETED.\n",
      "\n",
      "Ligating gene segments (computationally, of course!)... COMPLETED.\n",
      "\n",
      "Calculating and storing non-protein-coding genome and variant data (\"junk\" gets a bad rap!)... \n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1485\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1491\n",
      "\n",
      "## Variant 'T' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1497\n",
      "\n",
      "## Variant 'C' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1506\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1512\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1524\n",
      "\n",
      "## Variant 'A' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1530\n",
      "\n",
      "## Variant 'T' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1548\n",
      "\n",
      "## Variant 'A' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1566\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1569\n",
      "\n",
      "## Variant 'C' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1647\n",
      "\n",
      "## Variant 'C' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1773\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1776\n",
      "\n",
      "## Variant 'T' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1797\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1809\n",
      "\n",
      "## Variant 'C' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1813\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1824\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1827\n",
      "\n",
      "## Variant 'T' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1839\n",
      "\n",
      "## Variant 'C' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1842\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|1854\n",
      "\n",
      "## In temp_vcf4_Sample1.vcf|N/A\n",
      "## A total of 21 variants have been excluded because they\n",
      "## fall below the minimum minor allele frequency.\n",
      "COMPLETED.\n",
      "\n",
      "Processing all individual sites (nucleotides take time and memory; use a high-memory machine for eukaryotic chromosomes!)... COMPLETED.\n",
      "\n",
      "Processing population genetic estimates codon-by-codon (beware stochasticity!)... \n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1485\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1491\n",
      "\n",
      "## Variant 'T' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1497\n",
      "\n",
      "## Variant 'C' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1506\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1512\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1524\n",
      "\n",
      "## Variant 'A' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1530\n",
      "\n",
      "## Variant 'T' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1548\n",
      "\n",
      "## Variant 'A' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1566\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1569\n",
      "\n",
      "## Variant 'C' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1647\n",
      "\n",
      "## Variant 'C' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1773\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1776\n",
      "\n",
      "## Variant 'T' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1797\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1809\n",
      "\n",
      "## Variant 'C' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1813\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1824\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1827\n",
      "\n",
      "## Variant 'T' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1839\n",
      "\n",
      "## Variant 'C' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1842\n",
      "\n",
      "## Variant 'G' excluded from analysis because it falls below the\n",
      "## minimum minor allele frequency at:\n",
      "## temp_vcf4_Sample1.vcf|PB2|1854\n",
      "\n",
      "## In temp_vcf4_Sample1.vcf|PB2|N/A\n",
      "## A total of 21 variants have been excluded because they\n",
      "## fall below the minimum minor allele frequency.\n",
      "COMPLETED.\n",
      "\n",
      "Performing final calculations, noncoding overlap analysis, and output... temp_vcf4_Sample1.vcf COMPLETED.\n",
      "\n",
      "################################################################################\n",
      "##                      SNPGenie completed successfully.                      ##\n",
      "##             Please find results in the SNPGenie_Results folder.            ##\n",
      "################################################################################\n",
      "\n",
      "\n",
      "Results for be_w2_replicate-1_pb2 saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w2_replicate-1_pb2_results/\n",
      "Running SNPGenie for be_w1_replicate-1_pb1...\n",
      "\n",
      "\n",
      "################################################################################\n",
      "##                                                                            ##\n",
      "##                             SNPGenie Initiated!                            ##\n",
      "##                                                                            ##\n",
      "################################################################################\n",
      "\n",
      "  ###############################  LICENSE:  #################################\n",
      "  ##            SNPGenie Copyright (C) 2015-19 Chase W. Nelson              ##\n",
      "  ##            This program comes with ABSOLUTELY NO WARRANTY;             ##\n",
      "  ##     This is free software, and you are welcome to redistribute it      ##\n",
      "  ##               under certain conditions; see LICENSE.txt.               ##\n",
      "  ############################################################################\n",
      "\n",
      "WORKING_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA\n",
      "RESULTS_DIRECTORY=/Users/asjaeger/Desktop/project_source/hpai_PA/SNPGenie_Results\n",
      "\n",
      "Your MIN. MINOR ALLELE FREQ. is 0.01. All variants falling below this frequency will be ignored.\n",
      "\n",
      "Reading in FASTA file... COMPLETED.\n",
      "Creating temp_vcf4_Sample1.vcf...\n",
      "\n",
      "\n",
      "###########################  CURRENTLY PROCESSING:   ###########################\n",
      "temp_vcf4_Sample1.vcf... VCF format detected\n",
      "\n",
      "Converting temp_vcf4_Sample1.vcf to SNPGenie format...\n",
      "\n",
      "In file temp_vcf4_Sample1.vcf, the newline type is: Unix (LF, \\n)\n",
      "\n",
      "Calculating and storing protein-coding genome and variant data (that dogma stuff)... COMPLETED.\n",
      "\n",
      "Ligating gene segments (computationally, of course!)... COMPLETED.\n",
      "\n",
      "Calculating and storing non-protein-coding genome and variant data (\"junk\" gets a bad rap!)... COMPLETED.\n",
      "\n",
      "Processing all individual sites (nucleotides take time and memory; use a high-memory machine for eukaryotic chromosomes!)... COMPLETED.\n",
      "\n",
      "Processing population genetic estimates codon-by-codon (beware stochasticity!)... COMPLETED.\n",
      "\n",
      "Performing final calculations, noncoding overlap analysis, and output... temp_vcf4_Sample1.vcf COMPLETED.\n",
      "\n",
      "################################################################################\n",
      "##                      SNPGenie completed successfully.                      ##\n",
      "##             Please find results in the SNPGenie_Results folder.            ##\n",
      "################################################################################\n",
      "\n",
      "\n",
      "Results for be_w1_replicate-1_pb1 saved in /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-1_pb1_results/\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from subprocess import call\n",
    "\n",
    "# Define the specific output directory where results will go\n",
    "output_directory = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/\"\n",
    "os.makedirs(output_directory, exist_ok=True)\n",
    "\n",
    "# Define the base path for the reference FASTA and GTF files\n",
    "reference_base_path = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf/reference/\"\n",
    "\n",
    "# Define the mapping between sample suffixes and reference file names\n",
    "reference_mapping = {\n",
    "    'pa': 'pa',  # Reference files for 'pa' samples\n",
    "    'pb1': 'pb1',  # Reference files for 'pb1' samples\n",
    "    'pb2': 'pb2',  # Reference files for 'pb2' samples\n",
    "    'ha': 'ha',  # Reference files for 'ha' samples\n",
    "    'np': 'np',  # Reference files for 'np' samples\n",
    "    'ns': 'ns',  # Reference files for 'ns' samples\n",
    "    'na': 'na',  # Reference files for 'na' samples\n",
    "}\n",
    "\n",
    "# Run SNPGenie for each VCF file in the file list\n",
    "for vcf in file_list:\n",
    "    # Extract the sample suffix (e.g., 'pa', 'pb1', 'ha' etc. from sample name)\n",
    "    sample_suffix = vcf.split('_')[-1]  # Assuming the format ends with 'pa', 'pb1', etc.\n",
    "\n",
    "    # Check if the sample_suffix has a valid mapping\n",
    "    if sample_suffix in reference_mapping:\n",
    "        reference_name = reference_mapping[sample_suffix]\n",
    "\n",
    "        # Construct the paths to the reference FASTA and GTF files\n",
    "        reference_fasta = os.path.join(reference_base_path, f\"{reference_name}.fasta\")\n",
    "        reference_gtf = os.path.join(reference_base_path, f\"{reference_name}.gtf\")\n",
    "\n",
    "        # Check if the reference files exist (optional, for error handling)\n",
    "        if not os.path.exists(reference_fasta) or not os.path.exists(reference_gtf):\n",
    "            print(f\"Warning: Missing reference files for {sample_suffix}. Skipping {vcf}.\")\n",
    "            continue\n",
    "\n",
    "        # Construct the path to the VCF file (make sure to specify the correct path to your VCF files)\n",
    "        vcf_file = os.path.join('/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/parsed_vcf/', f\"{vcf}.vcf\")\n",
    "        \n",
    "        # Check if the VCF file exists\n",
    "        if not os.path.exists(vcf_file):\n",
    "            print(f\"VCF file missing for {vcf}. Skipping.\")\n",
    "            continue\n",
    "\n",
    "        # Define the SNPGenie output directory for the sample\n",
    "        snpgenie_output_dir = os.path.join(output_directory, f\"{vcf}_results/\")\n",
    "        os.makedirs(snpgenie_output_dir, exist_ok=True)\n",
    "\n",
    "        # Define the SNPGenie command\n",
    "        snpgenie_command = f\"perl /Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/snpgenie.pl --minfreq=0.01 --snpreport={vcf_file} --fastafile={reference_fasta} --gtffile={reference_gtf} --vcfformat=4\"\n",
    "\n",
    "        # Execute the SNPGenie command\n",
    "        print(f\"Running SNPGenie for {vcf}...\")\n",
    "        call(snpgenie_command, shell=True)\n",
    "\n",
    "        # Move SNPGenie results to the specific output folder (if applicable)\n",
    "        # Assuming SNPGenie generates a directory named SNPGenie_Results by default\n",
    "        call(f\"mv SNPGenie_Results {snpgenie_output_dir}\", shell=True)\n",
    "        \n",
    "        print(f\"Results for {vcf} saved in {snpgenie_output_dir}\")\n",
    "    else:\n",
    "        print(f\"No reference mapping found for sample suffix {sample_suffix}. Skipping.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Organized results and saved combined results for: be_w1_replicate-1\n",
      "Organized results and saved combined results for: be_w1_replicate-2\n",
      "Organized results and saved combined results for: be_w2_replicate-1\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from subprocess import call\n",
    "import glob\n",
    "\n",
    "# Define the base output directory where SNPGenie results are stored\n",
    "output_directory = \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/\"\n",
    "\n",
    "# Get unique sample names (excluding gene names)\n",
    "unique_samples = set(\"_\".join(sample.split(\"_\")[:-1]) for sample in file_list)\n",
    "\n",
    "# Organize results by sample\n",
    "for sample in unique_samples:\n",
    "    # Define the sample-level directory\n",
    "    sample_dir = os.path.join(output_directory, sample)\n",
    "    os.makedirs(sample_dir, exist_ok=True)  # Ensure sample-level directory exists\n",
    "\n",
    "    # Find all gene-specific result directories that belong to this sample\n",
    "    gene_result_dirs = glob.glob(os.path.join(output_directory, f\"{sample}_*_results\"))\n",
    "\n",
    "    # Move gene result directories into the sample-level directory\n",
    "    for gene_dir in gene_result_dirs:\n",
    "        if os.path.isdir(gene_dir):  # Ensure it's a directory\n",
    "            call(f\"mv {gene_dir} {sample_dir}/\", shell=True)  # Move it inside the sample directory\n",
    "\n",
    "    # Compile all product_results.txt files into a single file within the sample directory\n",
    "    combined_results_file = os.path.join(sample_dir, \"combined_product_results.txt\")\n",
    "\n",
    "    header_written = False  # Flag to track if header is written\n",
    "\n",
    "    with open(combined_results_file, \"w\") as outfile:\n",
    "        for gene_dir in gene_result_dirs:\n",
    "            product_results_file = os.path.join(sample_dir, os.path.basename(gene_dir), \"SNPGenie_Results/product_results.txt\")\n",
    "\n",
    "            if os.path.exists(product_results_file):  # Check if file exists\n",
    "                with open(product_results_file, \"r\") as infile:\n",
    "                    lines = infile.readlines()\n",
    "                    \n",
    "                    # Write header only once (from the first file)\n",
    "                    if not header_written:\n",
    "                        outfile.write(lines[0])  # Write header\n",
    "                        header_written = True\n",
    "                    \n",
    "                    # Write the rest of the file (excluding header)\n",
    "                    outfile.writelines(lines[1:])\n",
    "    \n",
    "    print(f\"Organized results and saved combined results for: {sample}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated file column for: be_w1_replicate-2\n",
      "Updated file column for: be_w1_replicate-1\n",
      "Updated file column for: be_w2_replicate-1\n"
     ]
    }
   ],
   "source": [
    "# Find all sample directories\n",
    "sample_dirs = [d for d in os.listdir(output_directory) if os.path.isdir(os.path.join(output_directory, d))]\n",
    "\n",
    "for sample in sample_dirs:\n",
    "    combined_results_file = os.path.join(output_directory, sample, \"combined_product_results.txt\")\n",
    "\n",
    "    if os.path.exists(combined_results_file):  # Check if file exists\n",
    "        with open(combined_results_file, \"r\") as infile:\n",
    "            lines = infile.readlines()\n",
    "\n",
    "        # Process file to replace the first column with the sample name\n",
    "        with open(combined_results_file, \"w\") as outfile:\n",
    "            header_written = False\n",
    "\n",
    "            for i, line in enumerate(lines):\n",
    "                split_line = line.strip().split(\"\\t\")\n",
    "\n",
    "                if i == 0:\n",
    "                    # Keep the header unchanged\n",
    "                    outfile.write(line)\n",
    "                    header_written = True\n",
    "                elif len(split_line) > 1:\n",
    "                    # Replace first column with sample name\n",
    "                    split_line[0] = sample\n",
    "                    outfile.write(\"\\t\".join(split_line) + \"\\n\")\n",
    "\n",
    "        print(f\"Updated file column for: {sample}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split gtfs and fastas by gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # split up fasta files\n",
    "\n",
    "# for fasta in fastas:\n",
    "#     with open(fasta, \"r\") as infile: \n",
    "#         for seq in SeqIO.parse(infile, \"fasta\"):\n",
    "#             name = seq.description\n",
    "#             sequence = str(seq.seq)\n",
    "        \n",
    "#             if name.endswith(\"MP\"):\n",
    "#                 with open(directory + name.replace(\"MP\",\"M1\") + \".fasta\", \"w\") as outfile:\n",
    "#                     outfile.write(\">\" + name.replace(\"MP\",\"M1\") + \"\\n\")\n",
    "#                     outfile.write(sequence) \n",
    "#                 with open(directory + name.replace(\"MP\",\"M2\") + \".fasta\", \"w\") as outfile:\n",
    "#                     outfile.write(\">\" + name.replace(\"MP\",\"M2\") + \"\\n\")\n",
    "#                     outfile.write(sequence)              \n",
    "                    \n",
    "#             elif name.endswith(\"NS\"):\n",
    "#                 with open(directory + name.replace(\"NS\",\"NS1\") + \".fasta\", \"w\") as outfile:\n",
    "#                     outfile.write(\">\" + name.replace(\"NS\",\"NS1\") + \"\\n\")\n",
    "#                     outfile.write(sequence) \n",
    "#                 with open(directory + name.replace(\"NS\",\"NEP\") + \".fasta\", \"w\") as outfile:\n",
    "#                     outfile.write(\">\" + name.replace(\"NS\",\"NEP\") + \"\\n\")\n",
    "#                     outfile.write(sequence) \n",
    "                    \n",
    "#             else:\n",
    "#                 with open(directory + name + \".fasta\", \"w\") as outfile:\n",
    "#                     outfile.write(\">\" + name + \"\\n\")\n",
    "#                     outfile.write(sequence) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # split up gtf files \n",
    "# for gtf in gtfs: \n",
    "#     with open(gtf, \"r\") as infile:\n",
    "#         for line in infile:\n",
    "#             gene = line.split(\"\\t\")[0]\n",
    "        \n",
    "#             if gene.endswith(\"MP\") and \"\\\"M1\\\"\" in line:\n",
    "#                 with open(directory + gene.replace(\"MP\",\"M1\") + \".gtf\", \"a\") as outfile:\n",
    "#                         outfile.write(line)\n",
    "            \n",
    "#             elif gene.endswith(\"MP\") and \"\\\"M2\\\"\" in line:\n",
    "#                 with open(directory + gene.replace(\"MP\",\"M2\") + \".gtf\", \"a\") as outfile:\n",
    "#                     outfile.write(line)\n",
    "                    \n",
    "#             elif gene.endswith(\"NS\") and \"\\\"NS1\\\"\" in line:\n",
    "#                 with open(directory + gene.replace(\"NS\",\"NS1\") + \".gtf\", \"a\") as outfile:\n",
    "#                     outfile.write(line)\n",
    "            \n",
    "#             elif gene.endswith(\"NS\") and \"\\\"NEP\\\"\" in line:\n",
    "#                 with open(directory + gene.replace(\"NS\",\"NEP\") + \".gtf\", \"a\") as outfile:\n",
    "#                     outfile.write(line)\n",
    "                    \n",
    "#             else:\n",
    "#                 with open(directory + gene + \".gtf\", \"a\") as outfile:\n",
    "#                     outfile.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtfs = glob.glob(\"test_gtf_parse/*.gtf\")  # Get all .gtf files\n",
    "\n",
    "# Make sure 'gtfs' is a list of file paths\n",
    "for gtf in gtfs:\n",
    "    with open(gtf, \"r\") as infile:\n",
    "        for line in infile:\n",
    "            chrom = line.split(\"\\t\")[0]  # e.g., OL638146.1\n",
    "\n",
    "            # Get gene_id from the attributes field (last column)\n",
    "            attributes = line.strip().split(\"\\t\")[8]\n",
    "            gene_id = None\n",
    "            for field in attributes.split(\";\"):\n",
    "                if \"gene_id\" in field:\n",
    "                    gene_id = field.strip().split(\" \")[1].replace('\"', \"\")\n",
    "                    break\n",
    "\n",
    "            if not gene_id:\n",
    "                continue  # skip if no gene_id found\n",
    "\n",
    "            # Handle special cases for MP and NS segments\n",
    "            if chrom.endswith(\"MP\"):\n",
    "                if gene_id == \"M1\":\n",
    "                    outname = gene_id  # Use gene_id directly\n",
    "                elif gene_id == \"M2\":\n",
    "                    outname = gene_id  # Use gene_id directly\n",
    "                else:\n",
    "                    outname = gene_id  # Default to gene_id for MP cases\n",
    "            elif chrom.endswith(\"NS\"):\n",
    "                if gene_id == \"NS1\":\n",
    "                    outname = gene_id  # Use gene_id directly\n",
    "                elif gene_id == \"NEP\":\n",
    "                    outname = gene_id  # Use gene_id directly\n",
    "                else:\n",
    "                    outname = gene_id  # Default to gene_id for NS cases\n",
    "            else:\n",
    "                outname = gene_id  # Use gene_id for other cases\n",
    "\n",
    "            # Write to a new file named after the gene_id\n",
    "            output_dir = \"test_gtf_parse/\"  # Define the directory where you want to save files\n",
    "            if not os.path.exists(output_dir):\n",
    "                os.makedirs(output_dir)  # Ensure the directory exists\n",
    "            with open(os.path.join(output_dir, f\"{outname}.gtf\"), \"a\") as outfile:\n",
    "                outfile.write(line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/\n",
      "<class 'str'>\n"
     ]
    }
   ],
   "source": [
    "print(directory)\n",
    "print(type(directory))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Consolidate files together into folders by sample name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # make a list of file names, which are just the gene names that the gtf file, fasta file and vcf files share\n",
    "# file_list = []\n",
    "# for f in glob.glob(directory + \"*.vcf\"):\n",
    "#     f = f.replace(\".vcf\",\"\")\n",
    "#     file_list.append(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # consolidate fastas, vcfs, and gtfs all into 1 folder together \n",
    "# for f in file_list:\n",
    "#     call(\"mkdir {f}\".format(f=f), shell=True)\n",
    "#     for n in glob.glob(directory + \"*\"):\n",
    "#         if f in n:\n",
    "#             call(\"mv {n} {f}\".format(n=n,f=f), shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Note: \n",
    "\n",
    "At this point, there will be gtf and fasta files for genes that are not in folders. That is ok. That means that there were no SNPs called for that gene in that sample. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # run SNP Genie on all of these\n",
    "# for f in file_list:\n",
    "#     f1 = f.replace(directory, \"\")\n",
    "#     call(\"snpgenie.pl --vcfformat=4 --gtffile={f}/{f1}.gtf --fastafile={f}/{f1}.fasta --snpreport={f}/{f1}.vcf\".format(f=f,f1=f1),shell=True)\n",
    "#     call(\"mv SNPGenie_Results {f}/\".format(f=f,f1=f1), shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 352,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # compile all of the output files from the various genes\n",
    "# call(\"grep -i '' */SNPGenie_Results/product_results.txt >> combined.txt\".format(directory=directory),shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merged seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-1/combined_product_results.txt\n",
      "Merged seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w1_replicate-2/combined_product_results.txt\n",
      "Merged seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/be_w2_replicate-1/combined_product_results.txt\n",
      "✅ Done. Merged into product_results_allgenes_allsamps.txt\n"
     ]
    }
   ],
   "source": [
    "# Output file\n",
    "output_file = \"seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/product_results_allgenes_allsamps.txt\"\n",
    "\n",
    "# Delete existing output file if present\n",
    "if os.path.exists(output_file):\n",
    "    os.remove(output_file)\n",
    "\n",
    "# Track if we've written the header\n",
    "header_written = False\n",
    "\n",
    "# Get all combined.txt files in subdirectories\n",
    "file_paths = sorted(glob.glob(\"**/combined_product_results.txt\", recursive=True))\n",
    "\n",
    "for file_path in file_paths:\n",
    "    with open(file_path, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "\n",
    "    if not lines:\n",
    "        continue  # skip empty files\n",
    "\n",
    "    header, *data = lines\n",
    "\n",
    "    with open(output_file, \"a\") as out:\n",
    "        if not header_written:\n",
    "            out.write(header)\n",
    "            header_written = True\n",
    "        out.writelines(data)\n",
    "\n",
    "    print(f\"Merged {file_path}\")\n",
    "\n",
    "print(f\"✅ Done. Merged into {output_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 583,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # take out all of the unwanted lines from the combined file\n",
    "# def cleanup_snpgenie_output(directory, infile):\n",
    "#     with open(directory + infile, \"r\") as infile: \n",
    "#         linenumber = 0\n",
    "#         for line in infile:\n",
    "#             linenumber += 1\n",
    "#             if linenumber == 1:\n",
    "#                 new_line = line.split(\":\")[1]\n",
    "#                 with open(directory + \"combined_pi_formatted.txt\", \"a\") as outfile:\n",
    "#                     outfile.write(new_line)\n",
    "        \n",
    "#             elif \"N_diffs\" not in line:\n",
    "#                 new_line = line.replace(\"/SNPGenie_Results/product_results.txt:temp_vcf4_Sample1.vcf\",\"\")\n",
    "            \n",
    "#                 genes = [\"_PB2\",\"_PB1\",\"_PA\",\"_H5\",\"_NP\",\"_N1\",\"_M1\",\"_M2\",\"_NS1\",\"_NEP\"]\n",
    "#                 for g in genes: \n",
    "#                     if g + \"\\t\" in new_line:\n",
    "#                         new_line2 = new_line.replace(g, \"\")\n",
    "#                 with open(directory + \"combined_pi_formatted.txt\", \"a\") as outfile:\n",
    "#                     outfile.write(new_line2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 584,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cleanup_snpgenie_output(directory, \"combined.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "source": [
    "## Results:\n",
    "\n",
    "This actually all worked really well! Except for the grep, which for some reason works totally fine in shell but isn't working for me when run from within this notebook. Otherwise though, things look good! I will now go through and see which genes didn't work/have empty values. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'file\\tproduct\\tN_diffs\\tS_diffs\\tN_diffs_vs_ref\\tS_diffs_vs_ref\\tN_sites\\tS_sites\\tpiN\\tpiS\\tmean_dN_vs_ref\\tmean_dS_vs_ref\\tmean_gdiv_polymorphic\\tmean_N_gdiv\\tmean_S_gdiv\\n'\n",
      "'be_w1_replicate-1\\tNA\\t0.0853980187327366\\t0.0907949386651212\\t0.0446096654275093\\t0.0476190476190476\\t1091.69640644362\\t309.303593556382\\t7.82250616826109e-05\\t0.000293546342676328\\t4.08627024548269e-05\\t0.000153955688233436\\t0.0879711171008575\\t0.0852392863559099\\t0.090702947845805\\n'\n",
      "'be_w1_replicate-1\\tHA\\t0.265355035769179\\t0.178263169906338\\t0.125038063631619\\t0.0910180479947746\\t1324.50435673024\\t376.464064322391\\t0.000200342893868807\\t0.000473519750755492\\t9.44036635261026e-05\\t0.000241770879668424\\t0.0341100099901029\\t0.0331508513202019\\t0.0356446638619444\\n'\n",
      "'be_w1_replicate-1\\tPB1\\t0.516780723002237\\t0.173763677941048\\t0.275539380691041\\t0.0927385605512713\\t1764.75758000292\\t506.242419997088\\t0.000292833831036092\\t0.000343242034008228\\t0.000156134408381794\\t0.000183190022977144\\t0.0985907556314371\\t0.103339395090085\\t0.0867191569848175\\n'\n",
      "'be_w1_replicate-1\\tPB1-F2\\t0\\t0\\t0\\t0\\t216.166666666667\\t53.8333333333334\\t0\\t0\\t0\\t0\\t*\\t*\\t*\\n'\n"
     ]
    }
   ],
   "source": [
    "with open(\"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/product_results_allgenes_allsamps.txt\", \"r\") as f:\n",
    "    for i in range(5):\n",
    "        print(repr(f.readline()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file                      object\n",
      "product                   object\n",
      "N_diffs                  float64\n",
      "S_diffs                  float64\n",
      "N_diffs_vs_ref           float64\n",
      "S_diffs_vs_ref           float64\n",
      "N_sites                  float64\n",
      "S_sites                  float64\n",
      "piN                      float64\n",
      "piS                      float64\n",
      "mean_dN_vs_ref           float64\n",
      "mean_dS_vs_ref           float64\n",
      "mean_gdiv_polymorphic    float64\n",
      "mean_N_gdiv              float64\n",
      "mean_S_gdiv              float64\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_table(\n",
    "    \"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/product_results_allgenes_allsamps.txt\",\n",
    "    sep=\"\\t\",\n",
    "    na_values=[\"NA\", \"*\"]  # treat both as NaN\n",
    ")\n",
    "\n",
    "# Optional: check if columns read correctly\n",
    "print(df.dtypes)\n",
    "\n",
    "# Fill missing 'product' values with 'neuram'\n",
    "df['product'] = df['product'].fillna('neuram')\n",
    "\n",
    "# Convert piN and piS to numeric (if not already)\n",
    "df['piN'] = pd.to_numeric(df['piN'], errors='coerce')\n",
    "df['piS'] = pd.to_numeric(df['piS'], errors='coerce')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>product</th>\n",
       "      <th>N_diffs</th>\n",
       "      <th>S_diffs</th>\n",
       "      <th>N_diffs_vs_ref</th>\n",
       "      <th>S_diffs_vs_ref</th>\n",
       "      <th>N_sites</th>\n",
       "      <th>S_sites</th>\n",
       "      <th>piN</th>\n",
       "      <th>piS</th>\n",
       "      <th>mean_dN_vs_ref</th>\n",
       "      <th>mean_dS_vs_ref</th>\n",
       "      <th>mean_gdiv_polymorphic</th>\n",
       "      <th>mean_N_gdiv</th>\n",
       "      <th>mean_S_gdiv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>be_w1_replicate-1</td>\n",
       "      <td>neuram</td>\n",
       "      <td>0.085398</td>\n",
       "      <td>0.090795</td>\n",
       "      <td>0.044610</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>1091.696406</td>\n",
       "      <td>309.303594</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000294</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>0.087971</td>\n",
       "      <td>0.085239</td>\n",
       "      <td>0.090703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>be_w1_replicate-1</td>\n",
       "      <td>HA</td>\n",
       "      <td>0.265355</td>\n",
       "      <td>0.178263</td>\n",
       "      <td>0.125038</td>\n",
       "      <td>0.091018</td>\n",
       "      <td>1324.504357</td>\n",
       "      <td>376.464064</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.000474</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.000242</td>\n",
       "      <td>0.034110</td>\n",
       "      <td>0.033151</td>\n",
       "      <td>0.035645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>be_w1_replicate-1</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0.516781</td>\n",
       "      <td>0.173764</td>\n",
       "      <td>0.275539</td>\n",
       "      <td>0.092739</td>\n",
       "      <td>1764.757580</td>\n",
       "      <td>506.242420</td>\n",
       "      <td>0.000293</td>\n",
       "      <td>0.000343</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>0.098591</td>\n",
       "      <td>0.103339</td>\n",
       "      <td>0.086719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>be_w1_replicate-1</td>\n",
       "      <td>PB1-F2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>216.166667</td>\n",
       "      <td>53.833333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>be_w1_replicate-1</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0.362772</td>\n",
       "      <td>0.295541</td>\n",
       "      <td>0.152741</td>\n",
       "      <td>0.151378</td>\n",
       "      <td>1753.581936</td>\n",
       "      <td>517.321389</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000571</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>0.000293</td>\n",
       "      <td>0.038718</td>\n",
       "      <td>0.036272</td>\n",
       "      <td>0.042213</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                file product   N_diffs   S_diffs  N_diffs_vs_ref  \\\n",
       "0  be_w1_replicate-1  neuram  0.085398  0.090795        0.044610   \n",
       "1  be_w1_replicate-1      HA  0.265355  0.178263        0.125038   \n",
       "2  be_w1_replicate-1     PB1  0.516781  0.173764        0.275539   \n",
       "3  be_w1_replicate-1  PB1-F2  0.000000  0.000000        0.000000   \n",
       "4  be_w1_replicate-1     PB2  0.362772  0.295541        0.152741   \n",
       "\n",
       "   S_diffs_vs_ref      N_sites     S_sites       piN       piS  \\\n",
       "0        0.047619  1091.696406  309.303594  0.000078  0.000294   \n",
       "1        0.091018  1324.504357  376.464064  0.000200  0.000474   \n",
       "2        0.092739  1764.757580  506.242420  0.000293  0.000343   \n",
       "3        0.000000   216.166667   53.833333  0.000000  0.000000   \n",
       "4        0.151378  1753.581936  517.321389  0.000207  0.000571   \n",
       "\n",
       "   mean_dN_vs_ref  mean_dS_vs_ref  mean_gdiv_polymorphic  mean_N_gdiv  \\\n",
       "0        0.000041        0.000154               0.087971     0.085239   \n",
       "1        0.000094        0.000242               0.034110     0.033151   \n",
       "2        0.000156        0.000183               0.098591     0.103339   \n",
       "3        0.000000        0.000000                    NaN          NaN   \n",
       "4        0.000087        0.000293               0.038718     0.036272   \n",
       "\n",
       "   mean_S_gdiv  \n",
       "0     0.090703  \n",
       "1     0.035645  \n",
       "2     0.086719  \n",
       "3          NaN  \n",
       "4     0.042213  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data wrangling and plotting:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>product</th>\n",
       "      <th>N_diffs</th>\n",
       "      <th>S_diffs</th>\n",
       "      <th>N_diffs_vs_ref</th>\n",
       "      <th>S_diffs_vs_ref</th>\n",
       "      <th>N_sites</th>\n",
       "      <th>S_sites</th>\n",
       "      <th>piN</th>\n",
       "      <th>piS</th>\n",
       "      <th>mean_dN_vs_ref</th>\n",
       "      <th>mean_dS_vs_ref</th>\n",
       "      <th>mean_gdiv_polymorphic</th>\n",
       "      <th>mean_N_gdiv</th>\n",
       "      <th>mean_S_gdiv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>be_w1_replicate-1</td>\n",
       "      <td>neuram</td>\n",
       "      <td>0.085398</td>\n",
       "      <td>0.090795</td>\n",
       "      <td>0.044610</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>1091.696406</td>\n",
       "      <td>309.303594</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000294</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>0.0879711171008575</td>\n",
       "      <td>0.0852392863559099</td>\n",
       "      <td>0.090702947845805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>be_w1_replicate-1</td>\n",
       "      <td>HA</td>\n",
       "      <td>0.265355</td>\n",
       "      <td>0.178263</td>\n",
       "      <td>0.125038</td>\n",
       "      <td>0.091018</td>\n",
       "      <td>1324.504357</td>\n",
       "      <td>376.464064</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.000474</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.000242</td>\n",
       "      <td>0.0341100099901029</td>\n",
       "      <td>0.0331508513202019</td>\n",
       "      <td>0.0356446638619444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>be_w1_replicate-1</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0.516781</td>\n",
       "      <td>0.173764</td>\n",
       "      <td>0.275539</td>\n",
       "      <td>0.092739</td>\n",
       "      <td>1764.757580</td>\n",
       "      <td>506.242420</td>\n",
       "      <td>0.000293</td>\n",
       "      <td>0.000343</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>0.0985907556314371</td>\n",
       "      <td>0.103339395090085</td>\n",
       "      <td>0.0867191569848175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>be_w1_replicate-1</td>\n",
       "      <td>PB1-F2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>216.166667</td>\n",
       "      <td>53.833333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>be_w1_replicate-1</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0.362772</td>\n",
       "      <td>0.295541</td>\n",
       "      <td>0.152741</td>\n",
       "      <td>0.151378</td>\n",
       "      <td>1753.581936</td>\n",
       "      <td>517.321389</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000571</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>0.000293</td>\n",
       "      <td>0.0387180552498716</td>\n",
       "      <td>0.0362719251297004</td>\n",
       "      <td>0.0422125268501162</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                file product   N_diffs   S_diffs  N_diffs_vs_ref  \\\n",
       "0  be_w1_replicate-1  neuram  0.085398  0.090795        0.044610   \n",
       "1  be_w1_replicate-1      HA  0.265355  0.178263        0.125038   \n",
       "2  be_w1_replicate-1     PB1  0.516781  0.173764        0.275539   \n",
       "3  be_w1_replicate-1  PB1-F2  0.000000  0.000000        0.000000   \n",
       "4  be_w1_replicate-1     PB2  0.362772  0.295541        0.152741   \n",
       "\n",
       "   S_diffs_vs_ref      N_sites     S_sites       piN       piS  \\\n",
       "0        0.047619  1091.696406  309.303594  0.000078  0.000294   \n",
       "1        0.091018  1324.504357  376.464064  0.000200  0.000474   \n",
       "2        0.092739  1764.757580  506.242420  0.000293  0.000343   \n",
       "3        0.000000   216.166667   53.833333  0.000000  0.000000   \n",
       "4        0.151378  1753.581936  517.321389  0.000207  0.000571   \n",
       "\n",
       "   mean_dN_vs_ref  mean_dS_vs_ref mean_gdiv_polymorphic         mean_N_gdiv  \\\n",
       "0        0.000041        0.000154    0.0879711171008575  0.0852392863559099   \n",
       "1        0.000094        0.000242    0.0341100099901029  0.0331508513202019   \n",
       "2        0.000156        0.000183    0.0985907556314371   0.103339395090085   \n",
       "3        0.000000        0.000000                     *                   *   \n",
       "4        0.000087        0.000293    0.0387180552498716  0.0362719251297004   \n",
       "\n",
       "          mean_S_gdiv  \n",
       "0   0.090702947845805  \n",
       "1  0.0356446638619444  \n",
       "2  0.0867191569848175  \n",
       "3                   *  \n",
       "4  0.0422125268501162  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # read in the SNPGenie pi results as a dataframe\n",
    "# df = pd.read_table(\"/Users/asjaeger/Desktop/project_source/hpai_PA/seq/illumina-pipeline/for_SNPgenie/sample_vcfs_test/product_results_allgenes_allsamps.txt\", sep=\"\\t\")\n",
    "# df['product'] = df['product'].fillna('neuram')                                     # replace NA with neuram\n",
    "# # df['species'] = np.where((df['file'].str.contains(\"duck\")), \"duck\",\"human\")        # add in a species column\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 15 entries, 0 to 14\n",
      "Data columns (total 15 columns):\n",
      " #   Column                 Non-Null Count  Dtype  \n",
      "---  ------                 --------------  -----  \n",
      " 0   file                   15 non-null     object \n",
      " 1   product                15 non-null     object \n",
      " 2   N_diffs                15 non-null     float64\n",
      " 3   S_diffs                15 non-null     float64\n",
      " 4   N_diffs_vs_ref         15 non-null     float64\n",
      " 5   S_diffs_vs_ref         15 non-null     float64\n",
      " 6   N_sites                15 non-null     float64\n",
      " 7   S_sites                15 non-null     float64\n",
      " 8   piN                    15 non-null     float64\n",
      " 9   piS                    15 non-null     float64\n",
      " 10  mean_dN_vs_ref         15 non-null     float64\n",
      " 11  mean_dS_vs_ref         15 non-null     float64\n",
      " 12  mean_gdiv_polymorphic  12 non-null     float64\n",
      " 13  mean_N_gdiv            12 non-null     float64\n",
      " 14  mean_S_gdiv            10 non-null     float64\n",
      "dtypes: float64(13), object(2)\n",
      "memory usage: 1.9+ KB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  product       piN       piS   piN/piS\n",
      "0      HA  0.000185  0.000809  0.228970\n",
      "1     PB1  0.000158  0.000212  0.747176\n",
      "2  PB1-F2  0.000000  0.000000       NaN\n",
      "3     PB2  0.000203  0.000446  0.456088\n",
      "4  neuram  0.000175  0.000125  1.402513\n"
     ]
    }
   ],
   "source": [
    "# Assuming your DataFrame is called df and has columns: 'product', 'piN', 'piS'\n",
    "\n",
    "# Group by product and calculate mean piN and piS\n",
    "grouped = df.groupby('product')[['piN', 'piS']].mean().reset_index()\n",
    "\n",
    "# Calculate the ratio\n",
    "grouped['piN/piS'] = grouped['piN'] / grouped['piS']\n",
    "\n",
    "# View the result\n",
    "print(grouped)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  product  piN_mean   piN_std  piS_mean   piS_std   piN/piS\n",
      "0      HA  0.000185  0.000135  0.000809  0.000881  0.228970\n",
      "1     PB1  0.000158  0.000140  0.000212  0.000185  0.747176\n",
      "2  PB1-F2  0.000000  0.000000  0.000000  0.000000       NaN\n",
      "3     PB2  0.000203  0.000003  0.000446  0.000120  0.456088\n",
      "4  neuram  0.000175  0.000141  0.000125  0.000152  1.402513\n"
     ]
    }
   ],
   "source": [
    "# Group by 'product' and calculate both mean and std for piN and piS\n",
    "grouped = df.groupby('product')[['piN', 'piS']].agg(['mean', 'std']).reset_index()\n",
    "\n",
    "\n",
    "# Flatten MultiIndex columns\n",
    "grouped.columns = ['_'.join(col).strip('_') for col in grouped.columns.values]\n",
    "\n",
    "# Rename for clarity\n",
    "grouped.rename(columns={\n",
    "    'piN_mean': 'piN_mean',\n",
    "    'piN_std': 'piN_std',\n",
    "    'piS_mean': 'piS_mean',\n",
    "    'piS_std': 'piS_std'\n",
    "}, inplace=True)\n",
    "\n",
    "# Calculate the ratio of the means\n",
    "grouped['piN/piS'] = grouped['piN_mean'] / grouped['piS_mean']\n",
    "\n",
    "# View result\n",
    "print(grouped)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA90AAAJOCAYAAACqS2TfAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAAU5RJREFUeJzt3Xl8Tdf+//H3yWxKDCWECErV0KJBG2oWNetITTGllVJTKA2tuUWrxKzakBpLjdfwRW7NRUlRLdqi2qREFZUElUiyf3/45dyeJkg023Hi9Xw8zuPevfbaZ39Oupvmfdbaa1sMwzAEAAAAAABynJO9CwAAAAAAILcidAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAeCJGRkbJYLLJYLNqxY0eG/YZhqHz58rJYLGrYsOF9r88Mv/zyiywWiyIjI61t6T8HDw8P/frrrxmOadiwoapWrZrp+02fPl1FihRRSkpKturo3r27ypQpY9N27do1TZo0SdWqVZOnp6cKFCigRx99VO3bt9fOnTuz9LkmT56crTqya8eOHdZrxmKxyNnZWd7e3nrllVd04sQJU8+drkyZMurevbtp77906VKFh4eb9v4AAPMRugEAD5QCBQooIiIiQ/vOnTt1+vRpFShQwA5VmaNEiRLat2+fWrVqlWFfUlKS3nnnnWy936pVq9SuXTu5uLhk67h3331Xa9assW6npqaqWbNmeu+99/Tyyy/riy++0MqVKzVo0CDFx8dr9+7d2Xp/s73//vvat2+ftm/frmHDhikqKkp169bV2bNn7V3av0boBgDHl73/KgMAYLIOHTpoyZIlmjVrljw9Pa3tERERCggIUEJCgh2ry1nu7u565plnMt3XvHlzLV26VEOGDFG1atXu+l6///679uzZo6FDh2a7jkcffdRme9euXdq7d6/mz5+vHj16WNufe+45vfnmm0pLS8v2OcxUoUIF68+xfv36KliwoHr16qXIyEiNGDEi02OuX7+uvHnz3s8yAQAPKUa6AQAPlI4dO0qSli1bZm2Lj4/XqlWr1LNnz0yPSU5O1vjx4/X444/L3d1dRYsWVY8ePfTHH3/Y9Fu+fLmaNWumEiVKKE+ePKpUqZLefvttXbt2zaZf9+7dlT9/fp06dUotW7ZU/vz55evrq8GDByspKemun6FMmTJq3bq11qxZoyeffFIeHh4qV66cpk+fbtMvs+nl6YYOHaoiRYpo2LBhdz2fJK1Zs0b58+dX06ZNJf1vmnpUVJR69OihwoULK1++fGrTpo1+/vnnDJ/379PLL126JOnWSHxmnJyy9udDWlqa3nvvPZUuXVoeHh6qWbOmvvzyS+v+3bt3y2Kx2PyzTrdw4UJZLBYdPHgwS+f6u/QAnj49f/To0bJYLDp06JBefvllFSpUyPpFw40bNxQWFqayZcvKzc1NJUuWVN++fXXlyhWb97x586aGDh2q4sWLK2/evHr22Wd14MCBDOdOP9c/pf/z+OWXX2zaly5dqoCAAOXPn1/58+dX9erVrTM9GjZsqI0bN+rXX3+1mUYPAHAshG4AwAPF09NTL7/8subPn29tW7ZsmZycnNShQ4cM/dPS0tSuXTtNnDhRnTp10saNGzVx4kRFRUWpYcOG+uuvv6x9T548qZYtWyoiIkKbN2/WwIEDtWLFCrVp0ybD+968eVNt27ZVkyZNtG7dOvXs2VNTp07VpEmTsvQ5jhw5ooEDB2rQoEFas2aN6tSpowEDBmT5PucCBQronXfe0ZYtW7Rt27a79l+1apVat24td3d3m/ZevXrJycnJOk35wIEDatiwYYZQ+Xc1a9aUq6urBgwYoCVLliguLi5LNf/TzJkztXnzZoWHh2vx4sVycnJSixYttG/fPklSvXr1VKNGDc2aNSvTY2vVqqVatWpl+7ynTp2SJBUtWtSm/cUXX1T58uX1xRdfaO7cuTIMQ88//7wmT56srl27auPGjQoNDdVnn32mxo0b23zB8tprr2ny5MkKCgrSunXr9NJLL+nFF1/Un3/+me360o0cOVKdO3eWj4+PIiMjtWbNGnXr1s36ZcHs2bNVt25dFS9eXPv27bO+AAAOxgAA4AGwYMECQ5Jx8OBBY/v27YYk4/vvvzcMwzBq1apldO/e3TAMw6hSpYrRoEED63HLli0zJBmrVq2yeb+DBw8akozZs2dner60tDTj5s2bxs6dOw1Jxrfffmvd161bN0OSsWLFCptjWrZsaVSsWPGun8XPz8+wWCzGkSNHbNoDAwMNT09P49q1a4ZhGMaZM2cMScaCBQsy/TkkJSUZ5cqVM2rWrGmkpaUZhmEYDRo0MKpUqWLzvhcvXjRcXFxsfgbp7/PCCy/Y9P3qq68MScb48eNtPq+fn59Nv4iICCN//vyGJEOSUaJECSMoKMjYtWvXXT9/+ufy8fEx/vrrL2t7QkKCUbhwYaNp06YZ6jx8+LC17cCBA4Yk47PPPrvjedKvk+XLlxs3b940rl+/buzatcsoX7684ezsbP1nOmrUKEOSMXLkSJvjN2/ebEgyPvjgA5v25cuXG5KMefPmGYZhGCdOnDAkGYMGDbLpt2TJEkOS0a1bN2tb+rn+Kf1znjlzxjAMw/j5558NZ2dno3Pnznf8jK1atcrwzwYA4FgY6QYAPHAaNGigRx99VPPnz9d3332ngwcP3nZq+YYNG1SwYEG1adNGKSkp1lf16tVVvHhxm5XQf/75Z3Xq1EnFixeXs7OzXF1d1aBBA0nKsNq1xWLJMAL+5JNPZrqieGaqVKmS4V7sTp06KSEhQYcOHcrSe7i5uWn8+PGKjo7WihUrbttv3bp1cnNzU/PmzTPs69y5s812nTp15Ofnp+3bt9/x3D179tRvv/2mpUuXqn///vL19dXixYvVoEEDffjhh1mq/8UXX5SHh4d1u0CBAmrTpo127dql1NRUSbduJyhWrJjNaPeMGTNUtGjRTGc2ZKZDhw5ydXVV3rx5Vb9+faWmpmrlypV68sknbfq99NJLNtvpMwj+ufr4K6+8onz58lmnwqf/rP75s2zfvn22F61LFxUVpdTUVPXt2/eejgcAOA5CNwDggWOxWNSjRw8tXrxYc+fO1WOPPaZ69epl2vf333/XlStX5ObmJldXV5vX+fPndfHiRUnS1atXVa9ePX399dcaP368duzYoYMHD2r16tWSZDMNXZLy5s1rExilWwuf3bhxI0ufoXjx4rdtS79nOiteffVVPfXUUxoxYoRu3ryZaZ+VK1eqRYsWmS4Mdrs6slKDl5eXOnbsqGnTpunrr7/W0aNH5e3trREjRtxxevrdzp2cnKyrV69KuvUz7d27t5YuXaorV67ojz/+0IoVKxQcHJxhqvztTJo0SQcPHtShQ4cUExOjn3/+Wc8//3yGfv+8R/3SpUtycXHJMA3dYrHY/IzS//efn8fFxUVFihTJUo3/lL7eQKlSpe7peACA42D1cgDAA6l79+4aOXKk5s6dq/fee++2/R555BEVKVJEmzdvznR/+iPGtm3bpnPnzmnHjh3W0W1JWQqP9+L8+fO3bctOULNYLJo0aZICAwM1b968DPvj4+P15ZdfZroY253qKF++fJZrSFelShW9+uqrCg8P108//aTatWvfsf/tzu3m5qb8+fNb29544w1NnDhR8+fP140bN5SSkqKQkJAs11WuXDnVrFnzrv3+uQhZ+jPN//jjD5vgbRiGzp8/b72fPP2f1/nz51WyZElrv5SUlAxfXqR/UZOUlGTzpUH6lz/p0s/322+/ydfX9661AwAcFyPdAIAHUsmSJfXWW2+pTZs26tat2237tW7dWpcuXVJqaqpq1qyZ4VWxYkVJ/wtc/xw9/fjjj02p/9ixY/r2229t2pYuXaoCBQroqaeeytZ7NW3aVIGBgRo7dqx1hDjd+vXrZbFY1Lp160yPXbJkic323r179euvv6phw4a3Pd+lS5eUnJyc6b4ffvhBkuTj43PXulevXm0zMyAxMVHr169XvXr15OzsbG0vUaKEXnnlFc2ePVtz585VmzZtVLp06bu+/7/VpEkTSdLixYtt2letWqVr165Z96f/rP75s1yxYoVSUlJs2tJXgT969KhN+/r16222mzVrJmdnZ82ZM+eONbq7u2eYhQEAcCyMdAMAHlgTJ068a59XX31VS5YsUcuWLTVgwADVrl1brq6u+u2337R9+3a1a9dOL7zwgurUqaNChQopJCREo0aNkqurq5YsWZIhGOcUHx8ftW3bVqNHj1aJEiW0ePFiRUVFadKkSff0fOhJkybJ399fFy5cUJUqVaztK1euVGBgoHVE/5+io6MVHBysV155RbGxsRoxYoRKliypPn363PZc27dv14ABA9S5c2fVqVNHRYoU0YULF7Rs2TJt3rxZQUFBWZoW7ezsrMDAQIWGhiotLU2TJk1SQkKCxowZk6HvgAED9PTTT0uSFixYcNf3zgmBgYF67rnnNGzYMCUkJKhu3bo6evSoRo0apRo1aqhr166SpEqVKqlLly4KDw+Xq6urmjZtqu+//16TJ0+2eZa8JLVs2VKFCxdWr169NHbsWLm4uCgyMlKxsbE2/cqUKaPhw4dr3Lhx+uuvv9SxY0d5eXnp+PHjunjxovVn9MQTT2j16tWaM2eO/P395eTklKVRfQDAg4PQDQBwaM7OzvrPf/6jadOmadGiRZowYYJcXFxUqlQpNWjQQE888YSkW1OEN27cqMGDB6tLly7Kly+f2rVrp+XLl2d75Dkrqlevrh49emjUqFE6efKkfHx8NGXKFA0aNOie3q9GjRrq2LGjli5dam27du2atmzZcsfR0oiICC1atEivvvqqkpKS1KhRI02bNk2FCxe+7THPPPOMevbsqe3bt2vRokW6ePGi8uTJo8qVK2vGjBl64403slTzm2++qRs3bqh///7WLws2btyounXrZuhbu3ZtlSlTRnny5LGOMJvNYrFo7dq1Gj16tBYsWKD33ntPjzzyiLp27ar333/fZlZERESEvL29FRkZqenTp6t69epatWqVXn31VZv39PT0tD6OrkuXLipYsKCCg4PVokULBQcH2/QdO3asKlSooBkzZqhz585ycXFRhQoV1L9/f2ufAQMG6NixYxo+fLji4+NlGIYMwzD3BwMAyFEWg9/cAADkqDJlyqhq1arasGGDqedZsWKFOnfurN9//z1DiI6MjFSPHj108OBBhxgZPXr0qKpVq6ZZs2bdcRQeAABHwz3dAAA4qPbt2+vmzZt3HLV+0J0+fVrbtm3T66+/rhIlSmR4fBcAAI6O0A0AAOxm3LhxCgwM1NWrV/XFF1/c0/3uAAA8yJheDgAAAACASRjpBgAAAADAJIRuAAAAAABMQugGAAAAAMAkD91zutPS0nTu3DkVKFBAFovF3uUAAAAAAByQYRhKTEyUj4+PnJxuP5790IXuc+fOydfX195lAAAAAABygdjYWJUqVeq2+x+60F2gQAFJt34wnp6edq4GAAAAAOCIEhIS5Ovra82Yt/PQhe70KeWenp6EbgAAAADAv3K325ZZSA0AAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwiV1D965du9SmTRv5+PjIYrFo7dq1WT72q6++kouLi6pXr25afQAAAAAA/Bt2Dd3Xrl1TtWrVNHPmzGwdFx8fr6CgIDVp0sSkygAAAAAA+Pdc7HnyFi1aqEWLFtk+rnfv3urUqZOcnZ2zNToOAAAAAMD95HD3dC9YsECnT5/WqFGj7F0KAAAAAAB3ZNeR7uw6efKk3n77be3evVsuLlkrPSkpSUlJSdbthISEHK2pdZPhOfp+eHht+PJ9e5cAAAAAIIc5zEh3amqqOnXqpDFjxuixxx7L8nETJkyQl5eX9eXr62tilQAAAAAA/I/DhO7ExERFR0frzTfflIuLi1xcXDR27Fh9++23cnFx0bZt2zI9LiwsTPHx8dZXbGzsfa4cAAAAAPCwcpjp5Z6envruu+9s2mbPnq1t27Zp5cqVKlu2bKbHubu7y93d/X6UCAAAAACADbuG7qtXr+rUqVPW7TNnzujIkSMqXLiwSpcurbCwMJ09e1YLFy6Uk5OTqlatanN8sWLF5OHhkaEdAAAAAIAHgV1Dd3R0tBo1amTdDg0NlSR169ZNkZGRiouLU0xMjL3KAwAAAADgX7EYhmHYu4j7KSEhQV5eXoqPj5enp+e/fj9WL0dOYfVyAAAAwHFkNVs6zEJqAAAAAAA4GkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBK7hu5du3apTZs28vHxkcVi0dq1a+/Yf/Xq1QoMDFTRokXl6empgIAAbdmy5f4UCwAAAABANtk1dF+7dk3VqlXTzJkzs9R/165dCgwM1KZNm/TNN9+oUaNGatOmjQ4fPmxypQAAAAAAZJ+LPU/eokULtWjRIsv9w8PDbbbff/99rVu3TuvXr1eNGjVyuDoAAAAAAP4dh76nOy0tTYmJiSpcuLC9SwEAAAAAIAO7jnT/Wx999JGuXbum9u3b37ZPUlKSkpKSrNsJCQn3ozQAAAAAABx3pHvZsmUaPXq0li9frmLFit2234QJE+Tl5WV9+fr63scqAQAAAAAPM4cM3cuXL1evXr20YsUKNW3a9I59w8LCFB8fb33FxsbepyoBAAAAAA87h5tevmzZMvXs2VPLli1Tq1at7trf3d1d7u7u96EyAAAAAABs2TV0X716VadOnbJunzlzRkeOHFHhwoVVunRphYWF6ezZs1q4cKGkW4E7KChI06ZN0zPPPKPz589LkvLkySMvLy+7fAYAAAAAAG7HrtPLo6OjVaNGDevjvkJDQ1WjRg2NHDlSkhQXF6eYmBhr/48//lgpKSnq27evSpQoYX0NGDDALvUDAAAAAHAndh3pbtiwoQzDuO3+yMhIm+0dO3aYWxAAAAAAADnIIRdSAwAAAADAERC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJPYNXTv2rVLbdq0kY+PjywWi9auXXvXY3bu3Cl/f395eHioXLlymjt3rvmFAgAAAABwD+wauq9du6Zq1app5syZWep/5swZtWzZUvXq1dPhw4c1fPhw9e/fX6tWrTK5UgAAAAAAss/Fnidv0aKFWrRokeX+c+fOVenSpRUeHi5JqlSpkqKjozV58mS99NJLJlUJAAAAAMC9cah7uvft26dmzZrZtD333HOKjo7WzZs37VQVAAAAAACZs+tId3adP39e3t7eNm3e3t5KSUnRxYsXVaJEiQzHJCUlKSkpybqdkJBgep0AAAAAAEgONtItSRaLxWbbMIxM29NNmDBBXl5e1pevr6/pNQIAAAAAIDlY6C5evLjOnz9v03bhwgW5uLioSJEimR4TFham+Ph46ys2NvZ+lAoAAAAAgGNNLw8ICND69ett2rZu3aqaNWvK1dU102Pc3d3l7u5+P8oDAAAAAMCGXUe6r169qiNHjujIkSOSbj0S7MiRI4qJiZF0a5Q6KCjI2j8kJES//vqrQkNDdeLECc2fP18REREaMmSIPcoHAAAAAOCO7DrSHR0drUaNGlm3Q0NDJUndunVTZGSk4uLirAFcksqWLatNmzZp0KBBmjVrlnx8fDR9+nQeFwYAAAAAeCDZNXQ3bNjQuhBaZiIjIzO0NWjQQIcOHTKxKgAAAAAAcoZDLaQGAAAAAIAjIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYJIsh+7Lly/rt99+s2k7duyYevToofbt22vp0qU5XhwAAAAAAI4sy6G7b9++mjJlinX7woULqlevng4ePKikpCR1795dixYtMqVIAAAAAAAcUZZD9/79+9W2bVvr9sKFC1W4cGEdOXJE69at0/vvv69Zs2aZUiQAAAAAAI4oy6H7/PnzKlu2rHV727ZteuGFF+Ti4iJJatu2rU6ePJnzFQIAAAAA4KCyHLo9PT115coV6/aBAwf0zDPPWLctFouSkpJytDgAAAAAABxZlkN37dq1NX36dKWlpWnlypVKTExU48aNrft/+ukn+fr6mlIkAAAAAACOyCWrHceNG6emTZtq8eLFSklJ0fDhw1WoUCHr/s8//1wNGjQwpUgAAAAAABxRlkN39erVdeLECe3du1fFixfX008/bbP/1VdfVeXKlXO8QAAAAAAAHFWWQ7ckFS1aVO3atct0X6tWrXKkIAAAAAAAcosshe7p06fr9ddfl4eHh6ZPn37Hvv3798+RwgAAAAAAcHRZCt1Tp05V586d5eHhoalTp962n8ViIXQDAAAAAPD/ZSl0nzlzJtP/DwAAAAAAbi/LjwzLjGEYMgwjp2oBAAAAACBXuafQHRERoapVq8rDw0MeHh6qWrWqPv3005yuDQAAAAAAh5at1csl6d1339XUqVPVr18/BQQESJL27dunQYMG6ZdfftH48eNzvEgAAAAAABxRtkP3nDlz9Mknn6hjx47WtrZt2+rJJ59Uv379CN0AAAAAAPx/2Z5enpqaqpo1a2Zo9/f3V0pKSo4UBQAAAABAbpDt0N2lSxfNmTMnQ/u8efPUuXPnHCkKAAAAAIDcINvTy6VbC6lt3bpVzzzzjCRp//79io2NVVBQkEJDQ639pkyZkjNVAgAAAADggLIdur///ns99dRTkqTTp09LkooWLaqiRYvq+++/t/azWCw5VCIAAAAAAI4p26F7+/btOVrA7Nmz9eGHHyouLk5VqlRReHi46tWrd9v+S5Ys0QcffKCTJ0/Ky8tLzZs31+TJk1WkSJEcrQsAAAAAgH8ry/d0+/j46I033tDmzZuVnJycIydfvny5Bg4cqBEjRujw4cOqV6+eWrRooZiYmEz779mzR0FBQerVq5eOHTumL774QgcPHlRwcHCO1AMAAAAAQE7KcuheunSp8ubNq379+umRRx7RK6+8okWLFuny5cv3fPIpU6aoV69eCg4OVqVKlRQeHi5fX99MF2qTbt07XqZMGfXv319ly5bVs88+q969eys6OvqeawAAAAAAwCxZDt0NGzbURx99pJMnT2rfvn166qmnNGvWLJUoUUINGzbU1KlTrfd4Z0VycrK++eYbNWvWzKa9WbNm2rt3b6bH1KlTR7/99ps2bdokwzD0+++/a+XKlWrVqtVtz5OUlKSEhASbFwAAAAAA90O2HxkmSVWqVFFYWJj279+vmJgYde7cWdu2bdMTTzyhqlWrauPGjXd9j4sXLyo1NVXe3t427d7e3jp//nymx9SpU0dLlixRhw4d5ObmpuLFi6tgwYKaMWPGbc8zYcIEeXl5WV++vr7Z+7AAAAAAANyjewrdf+ft7a3XXntN69ev1x9//KFx48bJ3d09y8f/c5VzwzBuu/L58ePH1b9/f40cOVLffPONNm/erDNnzigkJOS27x8WFqb4+HjrKzY2Nsu1AQAAAADwb9zTc7pvJ1++fHrhhRey1PeRRx6Rs7NzhlHtCxcuZBj9TjdhwgTVrVtXb731liTpySefVL58+VSvXj2NHz9eJUqUyHCMu7t7tr4EAAAAAAAgp2R5pNvJyUnOzs53fLm4ZD3Du7m5yd/fX1FRUTbtUVFRqlOnTqbHXL9+XU5OtiU7OztLujVCDgAAAADAgyTLKXnNmjW33bd3717NmDEj28E3NDRUXbt2Vc2aNRUQEKB58+YpJibGOl08LCxMZ8+e1cKFCyVJbdq00WuvvaY5c+boueeeU1xcnAYOHKjatWvLx8cnW+cGAAAAAMBsWQ7d7dq1y9D2ww8/KCwsTOvXr1fnzp01bty4bJ28Q4cOunTpksaOHau4uDhVrVpVmzZtkp+fnyQpLi7O5pnd3bt3V2JiombOnKnBgwerYMGCaty4sSZNmpSt8wIAAAAAcD9YjHuYl33u3DmNGjVKn332mZ577jlNmDBBVatWNaO+HJeQkCAvLy/Fx8fL09PzX79f6ybDc6AqQNrw5fv2LgEAAABAFmU1W2Zr9fL4+HgNGzZM5cuX17Fjx/Tll19q/fr1DhO4AQAAAAC4n7I8vfyDDz7QpEmTVLx4cS1btizT6eYAAAAAAOB/shy63377beXJk0fly5fXZ599ps8++yzTfqtXr86x4gAAAAAAcGRZDt1BQUGyWCxm1gIAAAAAQK6S5dAdGRlpYhkAAAAAAOQ+2VpIDQAAAAAAZF2WRrpffPFFRUZGytPTUy+++OId+3JPNwAAAAAAt2QpdHt5eVnv5/by8jK1IAAAAAAAcosshe4FCxZk+v8BAAAAAMDtZXkhtX+6cOGCfvzxR1ksFj322GMqVqxYTtYFAAAAAIDDy3boTkhIUN++ffX5558rNTVVkuTs7KwOHTpo1qxZTD8Hcol6vcfZuwTkErs/ftfeJQAAANhNtlcvDw4O1tdff60NGzboypUrio+P14YNGxQdHa3XXnvNjBoBAAAAAHBI2R7p3rhxo7Zs2aJnn33W2vbcc8/pk08+UfPmzXO0OAAAAAAAHFm2R7qLFCmS6RRyLy8vFSpUKEeKAgAAAAAgN8h26H7nnXcUGhqquLg4a9v58+f11ltv6d13uW8PAAAAAIB02Z5ePmfOHJ06dUp+fn4qXbq0JCkmJkbu7u76448/9PHHH1v7Hjp0KOcqBQAAAADAwWQ7dD///PMmlAEAAAAAQO6T7dA9atQoM+oAAAAAACDXyfY93QAAAAAAIGsI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJsr16eWpqqiIjI/Xll1/qwoULSktLs9m/bdu2HCsOAAAAAABHlu3QPWDAAEVGRqpVq1aqWrWqLBaLGXUBAAAAAODwsh26P//8c61YsUItW7Y0ox4AAAAAAHKNbN/T7ebmpvLly5tRCwAAAAAAuUq2Q/fgwYM1bdo0GYZhRj0AAAAAAOQa2Z5evmfPHm3fvl3/93//pypVqsjV1dVm/+rVq3OsOAAAAAAAHFm2Q3fBggX1wgsvmFELAAAAAAC5SrZD94IFC8yoAwAAAACAXCfb93QDAAAAAICsyfZItyStXLlSK1asUExMjJKTk232HTp0KEcKAwAAAADA0WV7pHv69Onq0aOHihUrpsOHD6t27doqUqSIfv75Z7Vo0cKMGgEAAAAAcEjZDt2zZ8/WvHnzNHPmTLm5uWno0KGKiopS//79FR8fb0aNAAAAAAA4pGyH7piYGNWpU0eSlCdPHiUmJkqSunbtqmXLluVsdQAAAAAAOLBsh+7ixYvr0qVLkiQ/Pz/t379fknTmzBkZhpGz1QEAAAAA4MCyHbobN26s9evXS5J69eqlQYMGKTAwUB06dOD53QAAAAAA/E22Vy+fN2+e0tLSJEkhISEqXLiw9uzZozZt2igkJCTHCwQAAAAAwFFlO3Q7OTnJyel/A+Tt27dX+/btc7QoAAAAAAByg2xPL5ek3bt3q0uXLgoICNDZs2clSYsWLdKePXtytDgAAAAAABxZtkP3qlWr9NxzzylPnjw6fPiwkpKSJEmJiYl6//33c7xAAAAAAAAcVbZD9/jx4zV37lx98skncnV1tbbXqVNHhw4dytHiAAAAAABwZNkO3T/++KPq16+fod3T01NXrlzJiZoAAAAAAMgVsh26S5QooVOnTmVo37Nnj8qVK5cjRQEAAAAAkBtkO3T37t1bAwYM0Ndffy2LxaJz585pyZIlGjJkiPr06WNGjQAAAAAAOKRsPzJs6NChio+PV6NGjXTjxg3Vr19f7u7uGjJkiN58800zagQAAAAAwCFlO3RL0nvvvacRI0bo+PHjSktLU+XKlZU/f/6crg0AAAAAAId2T6FbkvLmzauaNWvmZC0AAAAAAOQqWQ7dPXv2zFK/+fPn33MxAAAAAADkJlkO3ZGRkfLz81ONGjVkGIaZNQEAAAAAkCtkOXSHhITo888/188//6yePXuqS5cuKly4sJm1AQAAAADg0LL8yLDZs2crLi5Ow4YN0/r16+Xr66v27dtry5YtjHwDAAAAAJCJbD2n293dXR07dlRUVJSOHz+uKlWqqE+fPvLz89PVq1fvqYDZs2erbNmy8vDwkL+/v3bv3n3H/klJSRoxYoT8/Pzk7u6uRx99lPvIAQAAAAAPpHtevdxischiscgwDKWlpd3TeyxfvlwDBw7U7NmzVbduXX388cdq0aKFjh8/rtKlS2d6TPv27fX7778rIiJC5cuX14ULF5SSknKvHwMAAAAAANNka6Q7KSlJy5YtU2BgoCpWrKjvvvtOM2fOVExMzD09p3vKlCnq1auXgoODValSJYWHh8vX11dz5szJtP/mzZu1c+dObdq0SU2bNlWZMmVUu3Zt1alTJ9vnBgAAAADAbFkO3X369FGJEiU0adIktW7dWr/99pu++OILtWzZUk5O2crukqTk5GR98803atasmU17s2bNtHfv3kyP+c9//qOaNWvqgw8+UMmSJfXYY49pyJAh+uuvv257nqSkJCUkJNi8AAAAAAC4H7I8vXzu3LkqXbq0ypYtq507d2rnzp2Z9lu9enWW3u/ixYtKTU2Vt7e3Tbu3t7fOnz+f6TE///yz9uzZIw8PD61Zs0YXL15Unz59dPny5dve1z1hwgSNGTMmSzUBAAAAAJCTshy6g4KCZLFYcryAf76nYRi3PU9aWposFouWLFkiLy8vSbemqL/88suaNWuW8uTJk+GYsLAwhYaGWrcTEhLk6+ubg58AAAAAAIDMZTl0R0ZG5uiJH3nkETk7O2cY1b5w4UKG0e90JUqUUMmSJa2BW5IqVaokwzD022+/qUKFChmOcXd3l7u7e47WDgAAAABAVmT/Zuwc4ubmJn9/f0VFRdm0R0VF3XZhtLp16+rcuXM2jyf76aef5OTkpFKlSplaLwAAAAAA2WW30C1JoaGh+vTTTzV//nydOHFCgwYNUkxMjEJCQiTdmhoeFBRk7d+pUycVKVJEPXr00PHjx7Vr1y699dZb6tmzZ6ZTywEAAAAAsKd7fk53TujQoYMuXbqksWPHKi4uTlWrVtWmTZvk5+cnSYqLi1NMTIy1f/78+RUVFaV+/fqpZs2aKlKkiNq3b6/x48fb6yMAAAAAAHBbdg3d0q1HkfXp0yfTfZndR/74449nmJIOAAAAAMCDyK7TywEAAAAAyM0I3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASVzsXQAAAPdb9fGj7V0Ccokj74y2dwkAgAccI90AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJjE7qF79uzZKlu2rDw8POTv76/du3dn6bivvvpKLi4uql69urkFAgAAAABwj+waupcvX66BAwdqxIgROnz4sOrVq6cWLVooJibmjsfFx8crKChITZo0uU+VAgAAAACQfXYN3VOmTFGvXr0UHBysSpUqKTw8XL6+vpozZ84dj+vdu7c6deqkgICA+1QpAAAAAADZZ7fQnZycrG+++UbNmjWzaW/WrJn27t172+MWLFig06dPa9SoUVk6T1JSkhISEmxeAAAAAADcD3YL3RcvXlRqaqq8vb1t2r29vXX+/PlMjzl58qTefvttLVmyRC4uLlk6z4QJE+Tl5WV9+fr6/uvaAQAAAADICrsvpGaxWGy2DcPI0CZJqamp6tSpk8aMGaPHHnssy+8fFham+Ph46ys2NvZf1wwAAAAAQFZkbbjYBI888oicnZ0zjGpfuHAhw+i3JCUmJio6OlqHDx/Wm2++KUlKS0uTYRhycXHR1q1b1bhx4wzHubu7y93d3ZwPAQAAAADAHdhtpNvNzU3+/v6KioqyaY+KilKdOnUy9Pf09NR3332nI0eOWF8hISGqWLGijhw5oqeffvp+lQ4AAAAAQJbYbaRbkkJDQ9W1a1fVrFlTAQEBmjdvnmJiYhQSEiLp1tTws2fPauHChXJyclLVqlVtji9WrJg8PDwytAMAAAAA8CCwa+ju0KGDLl26pLFjxyouLk5Vq1bVpk2b5OfnJ0mKi4u76zO7AQAAAAB4UNk1dEtSnz591KdPn0z3RUZG3vHY0aNHa/To0TlfFAAAAAAAOcDuq5cDAAAAAJBbEboBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwiYu9CwAAAEDOafZ5mL1LQC6x9dUJ9i4ByBUY6QYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMYvfQPXv2bJUtW1YeHh7y9/fX7t27b9t39erVCgwMVNGiReXp6amAgABt2bLlPlYLAAAAAEDW2TV0L1++XAMHDtSIESN0+PBh1atXTy1atFBMTEym/Xft2qXAwEBt2rRJ33zzjRo1aqQ2bdro8OHD97lyAAAAAADuzq6he8qUKerVq5eCg4NVqVIlhYeHy9fXV3PmzMm0f3h4uIYOHapatWqpQoUKev/991WhQgWtX7/+PlcOAAAAAMDd2S10Jycn65tvvlGzZs1s2ps1a6a9e/dm6T3S0tKUmJiowoULm1EiAAAAAAD/iou9Tnzx4kWlpqbK29vbpt3b21vnz5/P0nt89NFHunbtmtq3b3/bPklJSUpKSrJuJyQk3FvBAAAAAABkk90XUrNYLDbbhmFkaMvMsmXLNHr0aC1fvlzFihW7bb8JEybIy8vL+vL19f3XNQMAAAAAkBV2C92PPPKInJ2dM4xqX7hwIcPo9z8tX75cvXr10ooVK9S0adM79g0LC1N8fLz1FRsb+69rBwAAAAAgK+wWut3c3OTv76+oqCib9qioKNWpU+e2xy1btkzdu3fX0qVL1apVq7uex93dXZ6enjYvAAAAAADuB7vd0y1JoaGh6tq1q2rWrKmAgADNmzdPMTExCgkJkXRrlPrs2bNauHChpFuBOygoSNOmTdMzzzxjHSXPkyePvLy87PY5AAAAAADIjF1Dd4cOHXTp0iWNHTtWcXFxqlq1qjZt2iQ/Pz9JUlxcnM0zuz/++GOlpKSob9++6tu3r7W9W7duioyMvN/lAwAAAABwR3YN3ZLUp08f9enTJ9N9/wzSO3bsML8gAAAAAAByiN1XLwcAAAAAILcidAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmsXvonj17tsqWLSsPDw/5+/tr9+7dd+y/c+dO+fv7y8PDQ+XKldPcuXPvU6UAAAAAAGSPXUP38uXLNXDgQI0YMUKHDx9WvXr11KJFC8XExGTa/8yZM2rZsqXq1aunw4cPa/jw4erfv79WrVp1nysHAAAAAODu7Bq6p0yZol69eik4OFiVKlVSeHi4fH19NWfOnEz7z507V6VLl1Z4eLgqVaqk4OBg9ezZU5MnT77PlQMAAAAAcHd2C93Jycn65ptv1KxZM5v2Zs2aae/evZkes2/fvgz9n3vuOUVHR+vmzZum1QoAAAAAwL1wsdeJL168qNTUVHl7e9u0e3t76/z585kec/78+Uz7p6Sk6OLFiypRokSGY5KSkpSUlGTdjo+PlyQlJCT8248gSbqZknT3TkAW5NQ1mVNSkm/YuwTkEg/atS1JqTf43Y2c8SBe3ynXub6RMx7E6xt4kKT/O2IYxh372S10p7NYLDbbhmFkaLtb/8za002YMEFjxozJ0O7r65vdUgFTeXlNsXcJgCm8It+3dwmAabzem2jvEgDTePWaau8SAIeQmJgoLy+v2+63W+h+5JFH5OzsnGFU+8KFCxlGs9MVL1480/4uLi4qUqRIpseEhYUpNDTUup2WlqbLly+rSJEidwz3yBkJCQny9fVVbGysPD097V0OkKO4vpGbcX0jt+LaRm7G9X1/GYahxMRE+fj43LGf3UK3m5ub/P39FRUVpRdeeMHaHhUVpXbt2mV6TEBAgNavX2/TtnXrVtWsWVOurq6ZHuPu7i53d3ebtoIFC/674pFtnp6e/IuPXIvrG7kZ1zdyK65t5GZc3/fPnUa409l19fLQ0FB9+umnmj9/vk6cOKFBgwYpJiZGISEhkm6NUgcFBVn7h4SE6Ndff1VoaKhOnDih+fPnKyIiQkOGDLHXRwAAAAAA4Lbsek93hw4ddOnSJY0dO1ZxcXGqWrWqNm3aJD8/P0lSXFyczTO7y5Ytq02bNmnQoEGaNWuWfHx8NH36dL300kv2+ggAAAAAANyW3RdS69Onj/r06ZPpvsjIyAxtDRo00KFDh0yuCjnF3d1do0aNyjDFH8gNuL6Rm3F9I7fi2kZuxvX9YLIYd1vfHAAAAAAA3BO73tMNAAAAAEBuRugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgDgIcODSwAAuH/s/pxuAHhQJCcny83NTdKtUGKxWOxcEZBz4uPjdfXqVRmGoVKlStm7HMA0aWlpcnJiXAm5E3+fOCZ+I+Ge/PLLLzp9+rQSEhKsbYycwJH9+OOPGjJkiA4ePChJslgsXNPINY4dO6bmzZurRYsWqlixosaMGaOLFy/auywgRxw/flyLFy/WyZMnJYnAjVwlJiZGCxYs0Lhx47R3717+PnFQjHQj21atWqUhQ4YoISFBDRo0UPv27fXqq69afwnw7RsczenTp9W4cWPFxcXp6tWrcnV1VfXq1bmmkSv88MMPatSokbp3767WrVvryJEjCgsLU40aNdS2bVt7lwf8K4mJiWrUqJEkqXXr1vrrr780ceJEFSpUSAUKFLD+Dud3ORzRd999p7Zt28rHx0dXrlzRmDFjtGLFCr344otc0w7GYvBVCbIhNjZWLVq00IABA1SwYEEtW7ZMFy5cUNeuXdW7d29JTHuBY7lx44beeustXbx4Ua1bt9Y777yjZ599VoMHD1b16tUlcU3DcSUkJKh79+7y9vbWnDlzrO1dunRRfHy81q9fz/UNhzdgwAB9++23Gj16tCZOnKjLly+rTJky6tu3r5555hm5u7tL+t+MPK53OIIzZ86ocePG6tChg0aOHCkXFxeNHTtWy5Yt0/79+1W0aFF7l4hsYKQb2eLq6qpq1aopKChI7u7uql69uiZMmKDPPvtMktS7d2++UYZDcXJyUuPGjXXt2jV17txZpUqVUvfu3WUYhoYMGcKINxzan3/+qZs3b6pdu3aSpNTUVDk7O+vxxx/Xl19+KYkAAsd18+ZNubq66sUXX1RcXJyqVKmizZs3a+vWrdq7d68aNWqkTp06qXbt2urfv78Mw2DqORzCzZs3FRERoaeeekrvvvuu8ubNK0lq1qyZlixZwvRyB8RvHmTJli1b1LlzZw0fPlyXL1+2fmtcoUIFhYWFqWLFilqyZImmTZsmiT/i4Djc3NzUtGlTdenSRZLUoEEDzZ8/X1999ZUmT56sI0eOWPsePXrUTlUC98bPz09DhgxR8+bNJf1vpM/Pz8/6ezzdhQsX7nt9wL345ZdfFB8fL1dXV0mSv7+/fvrpJ40cOVLSrWDy66+/qmjRoipUqJAmTJigihUrasaMGYQVOARXV1dVqVJFjz32mPLly2dtr169upKTk3X27FmlpaXZsUJkF6Ebd7Vt2za1bNlSSUlJOnDggP773/9qwoQJ1v0VKlTQiBEj5O3trU2bNunKlSv2KxbIgoSEBJ0/f16XLl2SJBUoUEDSrRVvDcNQo0aNFBERYQ3eBw8e1KBBg9StWzfFx8fbs3TgrtKv7/SF0ho0aCDp1vXt4nJrgttff/2lc+fOWf9oGzdunMaMGaMbN27Yp2ggi27evKmePXuqUqVK1r838ufPrylTpui7775TTEyMevTooS1btmjHjh2aMWOGvvnmG9WrV0+tW7dmUAAPtOjoaDVp0kRpaWnq2LGj9e/t9C+L0mdrWCwW66yNr7/+WomJiXarGVlD6MYdnT59WpcvX9a0adO0cuVKbdiwQYMHD9aiRYv04YcfWvuVL19eEydOVGRkpAoWLGi/goG7OHbsmJ5//nk9++yz6tChg5YsWWLdl/4fsrS0NDVu3Fjz58/X119/reeff17z5s1TRESEvLy87Fg9cGd/v75fffXVDNd3OovFImdnZzk5OWnUqFEaNWqUgoOD5eHhYY+ygSxzdXXV9OnTVapUKdWtW1d//vmnJMnX11cuLi5q2LChdu/erXXr1qlSpUpKS0uTj4+PPv30Uz366KN2rh64vW+//VaNGjVS5cqVbX5fp6WlyWKxKCUlRTdu3JCzs7N1sODtt99WYGAgX5g6AEI3buvXX39VrVq11Lt3b+u9JGXKlNEbb7yhdu3aKSIiQlOmTLH2f/TRR1WiRAl7lQvc1bfffqtnnnlGTzzxhMaMGSMPDw8tXLhQycnJ1j7pz3dNH/GuVKmS/vrrLx04cEBPPfWUHasH7iwr13f6/8+bN6/KlCmj8ePHa9KkSYqOjlaNGjXsVTqQJemjfZUrV9bChQtVsGBBNW3aVH/++acqVKigl156Sb/88otmzpypWrVqSeLxYXAM3377rQICAtS/f3/NmDHDZl/6Nezi4iI3NzelpqbKMAyNGjVKs2bN0n//+18WVXMALKSG28qbN6+GDRumjz76SIcPH7a2+/n56Y033pCzs7MmTZokV1dX9evXz46VAnd39OhR1a9fX/3799d7770n6dbIyODBg3X8+HH99ddfCggIkJOTk3XK7bBhw7RhwwYdPnxYVatWtWf5wB1l9fp2c3OTJKWkpGjdunXatm2bvvrqK75QwgPtxo0b8vDwkMVisS6e9vjjjysgIEBTpkxRgwYNtHPnTvXo0UNr167V8ePH1bx5cxbAhEM4ffq0GjRooB49elh/f0vS/Pnz5erqqq5du1rbXF1d5enpqb59+2rHjh3au3ev/P397VE2somv/3BbRYsWVffu3TV06FBFRERo7Nix1n2lS5dWr1691LdvX7Vs2dKOVQJ3l5iYqMDAQPn5+dn8B23z5s06efKk2rVrp7Zt26p+/fpKTU2Vk5OTnJyc1KpVKx05ckTVqlWzY/XAnWX3+pakwoULq1ixYtq/fz9/sOGBdvbsWQUFBWn79u2SZF087YMPPlBkZKQ++eQTubu7q379+kpJSVHNmjU1ffp065Rc4EG3detWubq6qnDhwjp37pwk6f3339fAgQNVoUIFa7+0tDRdvnxZJ06c0J49e3Tw4EF+fzsSA/ib48ePG19++aWxdetWa9vvv/9uTJ482ShYsKAxZswYm/43b9683yUC92T+/PmGu7u7MXbsWMMwDGPixImGp6ensWLFCuPw4cPG8uXLDU9PTyMkJMTOlQLZl9Xru3fv3tZjLl26ZK9ygSw7ffq0ERAQYLRs2dLYs2ePYRiGMWHCBKNw4cJGVFSUYRi3/napVq2aUa9ePePQoUNGrVq1jJiYGHuWDWTL+++/b9SoUcMYM2aM8fbbbxvFihUz/u///i/TvlOnTjWOHj16nyvEv2UxDJ6dgFvWrFmjt956S9KtqeUFChTQ1q1blS9fPl24cEGLFi3SxIkT1atXL02cONHO1QJ3Fxsbq82bN+v69evy9/fXlStX1K5dO9WtW1c//vijFi9erMDAQEnStWvX1KZNGxUsWFCrV6+2c+XA3XF942Fx8uRJ9e/fX+7u7ipWrJjWrl2rxYsXq1mzZtY+P/zwg1q0aKGiRYsqKiqKRS/hEFJTU+Xs7CxJGj9+vBYtWqTY2Fh9+umn6tSpE7dI5CLc0w1J0n//+191795dH374oYKCgvTll1+qTZs2ql+/vrZu3apixYqpa9euun79uiIiIjRkyBAVKVKEXwR4YB09elTt2rVTwYIFdfr0aUnS9OnTtWHDBrVp00YdO3ZUYGCgdWGefPnyqVChQipVqpR1WiLXNx5U//b6ZnEpOJIKFSpo2rRpevPNN7V48WKNGzfOGrjTr+fHH39cmzdvlru7O4EbDsPZ2dl6Db/zzjvKkyePIiIi9P333+vs2bMqWbIkwTuX4L+60F9//aXPP/9cYWFhev3113X58mX17dtXHTp0UHJysho3bqzLly+rWLFiCgkJ0aFDh/TII4/wCwAPrKNHjyogIEAdOnTQtm3bFBUVpeeff15Dhw6Vr6+vPvnkEy1dulTvvfee9T9mI0eO1K5du9SnTx/ro8OAB1FOXN+Ao3nsscc0Z84c1atXT19++aX27NkjSTaLX1asWFFlypSxY5VA9qVfv5I0ePBgdenSRZs3b9a0adN07tw5/h7JJRjpfsgdOHBA1atX1/PPP6+SJUvq8uXLat26tZo3b645c+Zo8eLF6tatm2rVqqXo6GgeSYAHXmxsrJo0aaJWrVpZb4N4+umnFRcXpzVr1ujq1avq0aOHUlJSFBISorx58+rKlSv64IMPtHfvXlWsWNHOnwC4Pa5vPMweffRRzZw5U/3799f48eP17rvvqm7dunyRBIeVkpIiFxcXnTlzRpMmTdLkyZM1fPhwGYahdevW6dq1a3rnnXd4JG8uQOh+iG3ZskUvvfSS1qxZo9atW0uS1q9fr3z58mn48OGyWCwqWrSomjdvrtTUVF26dEmFChWyc9XAnaWmpqps2bJKSkrSnj179Oyzz0qSihUrJmdnZ+sfZ6+99posFotef/11SVJ0dDSPTcIDj+sbD7sKFSpo+vTpCg0N1ZAhQzR16lQ988wz9i4LuKPk5GTrIxv/Pl3cxcVFv/zyi5599lkFBgbKw8NDkjRixAhdv35de/bssd7zDcfGQmoPqd9++00TJ05UxYoVbZ6xHR4erjFjxujPP/+UdOtf+suXLys8PFzu7u72KhfIlvRFd9LS0hQeHq5SpUqpfPnyCgoK0ocffmjTd9WqVapcubIqVapkp2qB7OH6Bm4tnPbuu+/qo48+UunSpe1dDnBbP/74o2bNmqWuXbuqVq1aNvuuX7+u1q1bq0yZMoqIiJDFYrFZd+PSpUsqUqSIPcpGDiN0P4Sio6M1ZswYxcbGatq0aWrQoIF19cRff/1VjRs3VmpqqipVqqTdu3dr3759euKJJ+xdNpAtJ0+e1IABA3T9+nUdPXpU3bp109SpUyWJhaTg8Li+AdvRQ+BBdPr0adWvX19xcXHq3r27+vfvr+rVq1v3X79+XSdOnNBTTz1lc+82v8dzH/5pPoRcXV31559/6tixY9q/f7+kW6snGoahUqVKaeXKlWrTpo0qV66sAwcOELjhkNJXu3V2dpanp6deeOEF6z4WJYGj4/oGRODGA+3GjRsKDw9X/fr1tWjRIm3fvl0fffSRjhw5Yu2TN29e+fv7Z/i9TeDOfRjpfogcOXJERYsWVcmSJfXjjz+qf//+SkhI0ODBg/Xyyy9n6P/3ZwcCjurUqVPq16+fDMOwLroD5BZc3wDwYEpOTtbGjRt17do1denSRTt37lT37t1Vt25dDRkyxDri/fd7vHk8WO5F6H5InD17Vu3bt5enp6ciIiLk4+Oj77//XqGhoZKkkJAQvfjii5II28h9Tp48qdDQUF28eJFFd5DrcH0DwIMpMTFRBQoUsG5v375dPXv2tAnehmHou+++05NPPmnHSmE25i48JEqWLKmuXbsqKSlJ/fr107lz51S1alV99NFHkqRPP/1Uy5YtkyQCN3KdChUq6MMPP1SpUqXk4+Nj73KAHMX1DQAPhoSEBJ0/f16XLl2SJGvgTktLk2EYatSokSIiIvTVV19p8uTJOnjwoAYNGqRu3bopPj7enqXDZIx053L/XIhh/vz5WrBggYoVK6YZM2ZYR7x79eqlUqVKKTIy0uYbOSA3YdEd5GZc3wBgP8eOHVO/fv0UExOjMmXKqEePHurcubNNn/S/y7dv367XX39d169f159//qk9e/bwWMdcjtCdi3399ddauHChJkyYIE9PT2v7ggULNGvWLJUrV04zZsyQt7e3jh8/rvz58/PYDQAAACAbvv32Wz377LPq2bOnateurWXLlunmzZtav3699cvQ9MCdft9227ZttWfPHu3atUtVq1a18yeA2ZhenotFRUVp165devfdd5WYmGht79Gjhxo2bKj169crKChI586dU+XKlQncAAAAQDYcPXpU9evXV//+/TVt2jR17txZQ4cO1eXLl3X8+HHt27dP0q0VydOnmQ8dOlQbNmzQ9u3bCdwPCUJ3LjZ06FB16dJF+/fvV1hYmM29IrVr11aVKlVUqFAhpaam2rFKAAAAwPEkJiYqMDBQfn5+eu+996ztmzdv1smTJ9WuXTu1bdtW9evXV2pqqpycnOTk5KRWrVrpyJEjqlatmh2rx/1E6M5lTpw4of3792vLli1yc3PT0KFD9corryg6OtomeB89elQtW7bU3Llz5evra+eqAQAAAMdSoEABTZw4UT/99JPGjRsnSZo0aZJmzZqlTz75ROvWrdOsWbP07bff6s0337Qe16BBA1Yrf8i42LsA5JzVq1drwIABKlWqlH788UcFBARowIABCg0NVVpamtasWaPHHntM/v7+2rFjhw4dOqSCBQvau2wAAADAYcTGxmrz5s26fv26/P39tXLlSrVr105RUVH68ccftXLlSgUGBkq69YQJf39//f7773auGvZE6M4l9u3bp9dee02TJk1ScHCwtm/friZNmqhVq1Zq1qyZBg8erPr162v9+vVKSUnRlClT9Pjjj9u7bAAAAMBhHD16VO3atVPBggV1+vRpSdL06dO1YcMGtWnTRh07dlRgYKDS16rOly+fChUqpFKlSiktLU0Wi0UWi8WeHwF2wOrlucT06dO1Y8cOrV69WidPnlTLli3VqFEjzZs3T9Kte07+/qzAvz9GDAAAAMCdHT16VAEBAerXr5+GDRumn376SbNmzdLmzZu1bds2HTx4UMHBwRo7dqzCwsLk5OSkkSNHas6cOdqzZ48qVqxo748AO2GkO5c4deqU9d7sRo0aqVWrVpo7d64kaeXKlUpISFDXrl3l6upK4AYAAACyITY21jqLdOLEiZKkp59+WnFxcVqzZo2uXr2qHj16KCUlRSEhIcqbN6+uXLmiDz74QHv37iVwP+QI3Q7s8uXL8vDwUN68eRUYGKjOnTsrMjJSPXv21EcffWSdurJ161bdvHlTHTt2lKurq52rBgAAABxLamqqypYtq6SkJO3Zs0fPPvusJKlYsWJydna2Dmq99tprslgsev311yVJ0dHReuqpp+xWNx4MDHk6qLVr16pt27aqXr26Ro0aJTc3N/Xs2VPu7u5q3ry5nJycdPnyZQ0fPlzr1q3TsGHDlCdPHnuXDQAAADicMmXKaMmSJUpOTta4ceN04sQJJSYm6oUXXtBrr72m2rVrW/sGBwfriy++0LFjxwjckMQ93Q7p0KFDaty4sQYPHqxLly5pz549qlChgp588kn99ttv+uSTT1S5cmV5eHgoLi5Oa9euVY0aNexdNgAAAODQTp48qQEDBuj69es6evSounXrpqlTp0pi3STcHqHbwZw+fVrLli2TxWLRiBEjJEnr16/XjBkzVKhQIXXu3FlFihTR7t275efnp7p166p06dJ2rhoAAADIHU6ePKmQkBCdPn1aCxcuVP369SVJhmGwMjkyReh2IAkJCWrSpIliYmLUs2dPTZgwwbrvP//5j8LDw1WoUCGNGDGCqSwAAACASU6dOqV+/frJMAy9++67qlu3rr1LwgOM+Q8OxNPTU/PmzVPBggW1e/duHTt2zLqvbdu2GjJkiH7++WdNmTJF169fF9+nAAAAADmvfPnymj59ulxdXTVkyBDt37/f3iXhAcZItwNKv3+kdu3a6t+/v6pUqWLdt3XrVlWsWFF+fn52rBAAAADI/X744Qe9++67+uijj7ilE7dF6HZQhw8fVnBwsJ566ikNGjRIlStXtndJAAAAwEMnOTlZbm5u9i4DDzBCtwM7fPiwQkJCVK5cOY0aNUqPP/64vUsCAAAAAPwN93Q7sBo1amjmzJmKi4uTl5eXvcsBAAAAAPwDI925wI0bN+Th4WHvMgAAAAAA/0DoBgAAAADAJEwvBwAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAtzV69GhVr17d3mUAAOCwCN0AADig7t27y2KxyGKxyNXVVeXKldOQIUN07do1e5d2Rzt27JDFYtGVK1fsXQoAAPeFi70LAAAA96Z58+ZasGCBbt68qd27dys4OFjXrl3TnDlzbPrdvHlTrq6udqoSAICHGyPdAAA4KHd3dxUvXly+vr7q1KmTOnfurLVr11qnhM+fP1/lypWTu7u7DMNQTEyM2rVrp/z588vT01Pt27fX77//bvOeEydOlLe3twoUKKBevXrpxo0bNvsbNmyogQMH2rQ9//zz6t69u3U7KSlJQ4cOla+vr9zd3VWhQgVFRETol19+UaNGjSRJhQoVksVisTkOAIDciJFuAAByiTx58ujmzZuSpFOnTmnFihVatWqVnJ2dJd0Kx/ny5dPOnTuVkpKiPn36qEOHDtqxY4ckacWKFRo1apRmzZqlevXqadGiRZo+fbrKlSuXrTqCgoK0b98+TZ8+XdWqVdOZM2d08eJF+fr6atWqVXrppZf0448/ytPTU3ny5MnRnwEAAA8aQjcAALnAgQMHtHTpUjVp0kSSlJycrEWLFqlo0aKSpKioKB09elRnzpyRr6+vJGnRokWqUqWKDh48qFq1aik8PFw9e/ZUcHCwJGn8+PH673//m2G0+05++uknrVixQlFRUWratKkk2YT2woULS5KKFSumggUL/uvPDQDAg47p5QAAOKgNGzYof/788vDwUEBAgOrXr68ZM2ZIkvz8/KyBW5JOnDghX19fa+CWpMqVK6tgwYI6ceKEtU9AQIDNOf65fTdHjhyRs7OzGjRocK8fCwCAXIWRbgAAHFSjRo00Z84cubq6ysfHx2axtHz58tn0NQxDFoslw3vcrv12nJycZBiGTVv6lHZJTBcHAOAfGOkGAMBB5cuXT+XLl5efn99dVyevXLmyYmJiFBsba207fvy44uPjValSJUlSpUqVtH//fpvj/rldtGhRxcXFWbdTU1P1/fffW7efeOIJpaWlaefOnZnW4ebmZj0OAICHAaEbAICHQNOmTfXkk0+qc+fOOnTokA4cOKCgoCA1aNBANWvWlCQNGDBA8+fP1/z58/XTTz9p1KhROnbsmM37NG7cWBs3btTGjRv1ww8/qE+fPjbP3C5Tpoy6deumnj17au3atTpz5ox27NihFStWSLo17d1isWjDhg36448/dPXq1fv2MwAAwB4I3QAAPAQsFovWrl2rQoUKqX79+mratKnKlSun5cuXW/t06NBBI0eO1LBhw+Tv769ff/1Vb7zxhs379OzZU926dbMG9rJly1ofA5Zuzpw5evnll9WnTx89/vjjeu2113Tt2jVJUsmSJTVmzBi9/fbb8vb21ptvvmn+hwcAwI4sxj9vzAIAAAAAADmCkW4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAk/w9Owh2yNBYFwwAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 1000x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Sort by ratio if you want cleaner x-axis order\n",
    "grouped_sorted = grouped.sort_values('piN/piS', ascending=False)\n",
    "\n",
    "# Plot\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.barplot(data=grouped_sorted, x='product', y='piN/piS', hue = 'product', palette='viridis')\n",
    "\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.ylabel('Mean piN/piS')\n",
    "plt.xlabel('Product')\n",
    "plt.title('Mean piN/piS by Product')\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA90AAAJOCAYAAACqS2TfAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAAV8xJREFUeJzt3Xl8Tdf+//H3ySyJxDxHUKrmWYuihqiZDtcshqatFtGiWtVSqhdtKTG1NLgxu6aW9ou0KKpqKKKoqmqjRBWVRELG/fsjvxz3NMIJ2TlJvJ6Px3k87lln7X0+R073zTtr7bUshmEYAgAAAAAA2c7J0QUAAAAAAJBfEboBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAecaSJUtksVhksVi0c+fODK8bhqHKlSvLYrHoiSeeyPH6zPDbb7/JYrFoyZIl1rb0fwcPDw/9/vvvGY554oknVLNmzdueLyQkREWLFlVycnKW6hg4cKAqVKhg0xYXF6dp06apTp068vHxUcGCBfXQQw+pR48e+uabb+z6XB9++GGW6rhXZ8+eVXBwsKpVqyYvLy95eHioQoUK6tevn3bs2CHDMHKkDgDAg8fF0QUAAJBVBQsWVGhoaIZg/c033+jMmTMqWLCgYwozQenSpfXdd9/poYceyvBaQkKC3nrrLS1dutTu861bt07dunWTi0vWfgV4++23NWLECOvzlJQUtWvXTseOHdNrr72mxo0bS5JOnz6tTZs2affu3WrZsmWW3sMsn3/+ufr06aNixYppyJAhql+/vtzd3fXLL79o7dq1at26tb766iu1adPG0aUCAPIhQjcAIM/p2bOnli9frrlz58rHx8faHhoaqiZNmigmJsaB1WUvd3d3PfbYY7d9rX379lqxYoVGjx6tOnXq3PVcf/75p/bs2aMxY8ZkuY5/hv5du3Zp7969WrRokQYNGmRtf/LJJzVs2DClpqZm+T3McObMGfXu3Vs1atTQV199ZfN9admypZ577jnt3LlThQsXdmCVAID8jOnlAIA8p3fv3pKklStXWtuio6O1bt06DR48+LbHJCYmavLkyXrkkUfk7u6u4sWLa9CgQfrrr79s+q1evVrt2rVT6dKlVaBAAVWrVk1vvPGG4uLibPoNHDhQ3t7e+uWXX9SxY0d5e3vLz89Po0aNUkJCwl0/Q4UKFdS5c2dt2LBBtWvXloeHhypVqqSQkBCbfrebXp5uzJgxKlq0qF5//fW7vp8kbdiwQd7e3mrbtq2kW9PUw8PDNWjQIBUpUkReXl7q0qWLfv311wyf93+nl1+5ckVS2kj87Tg52fcrRmpqqt577z2VL19eHh4eatiwob7++mvr67t375bFYrH5WacLCwuTxWLRgQMHMj3/jBkzFB8fr3nz5tkE7v/1xBNPZPijxenTp9WnTx+VKFFC7u7uqlatmubOnWvTZ+fOndbaxo0bpzJlysjHx0dt27bVqVOnMrxP+mi6j4+PPD091axZM5vPCgDInwjdAIA8x8fHR88++6wWLVpkbVu5cqWcnJzUs2fPDP1TU1PVrVs3TZ06VX369NEXX3yhqVOnKjw8XE888YRu3Lhh7Xv69Gl17NhRoaGh2rJli1555RWtWbNGXbp0yXDepKQkde3aVW3atNFnn32mwYMH66OPPtK0adPs+hxHjhzRK6+8oldffVUbNmxQ06ZNNWLECLvvcy5YsKDeeustbd26Vdu3b79r/3Xr1qlz585yd3e3aX/uuefk5OSkFStWaObMmdq/f7+eeOIJXbt2LdNzNWzYUK6urhoxYoSWL1+uqKgou2r+pzlz5mjLli2aOXOmli1bJicnJ3Xo0EHfffedJKl58+aqV69ehsCbfmyjRo3UqFGjTM8fHh6u0qVLq2HDhnbXdOLECTVq1Eg//vijpk+frs2bN6tTp04KDg7WxIkTM/R/88039fvvv+vTTz/VggULdPr0aXXp0kUpKSnWPsuWLVO7du3k4+Oj//znP1qzZo2KFCmiJ598kuANAPmdAQBAHrF48WJDknHgwAFjx44dhiTjxx9/NAzDMBo1amQMHDjQMAzDqFGjhtGyZUvrcStXrjQkGevWrbM534EDBwxJxrx58277fqmpqUZSUpLxzTffGJKMo0ePWl8bMGCAIclYs2aNzTEdO3Y0qlatetfP4u/vb1gsFuPIkSM27QEBAYaPj48RFxdnGIZhnD171pBkLF68+Lb/DgkJCUalSpWMhg0bGqmpqYZhGEbLli2NGjVq2Jz38uXLhouLi82/Qfp5nnrqKZu+3377rSHJmDx5ss3n9ff3t+kXGhpqeHt7G5IMSUbp0qWNwMBAY9euXXf9/Omfq0yZMsaNGzes7TExMUaRIkWMtm3bZqjz8OHD1rb9+/cbkoz//Oc/d3wfDw8P47HHHsvQnpKSYiQlJVkfKSkp1teefPJJo1y5ckZ0dLTNMcOGDTM8PDyMq1evGoZhWL+DHTt2tOm3Zs0aQ5Lx3XffGYZhGHFxcUaRIkWMLl26ZKihTp06RuPGje/4GQAAeRsj3QCAPKlly5Z66KGHtGjRIh07dkwHDhzIdGr55s2bVahQIXXp0kXJycnWR926dVWqVCmbldB//fVX9enTR6VKlZKzs7NcXV2tC4KdPHnS5rwWiyXDCHjt2rVvu6L47dSoUSPDtOY+ffooJiZGP/zwg13ncHNz0+TJk3Xw4EGtWbMm036fffaZ3Nzc1L59+wyv9e3b1+Z506ZN5e/vrx07dtzxvQcPHqw//vhDK1asUHBwsPz8/LRs2TK1bNlSH3zwgV31P/300/Lw8LA+L1iwoLp06aJdu3ZZR4p79+6tEiVK2Ix2z549W8WLF7/tzAZ739fV1dX6CA4OliTdvHlTX3/9tZ566il5enrafF86duyomzdvat++fTbn6tq1q83z2rVrS5L1e7B3715dvXpVAwYMsDlfamqq2rdvrwMHDmS4fQEAkH8QugEAeZLFYtGgQYO0bNkyffzxx3r44YfVvHnz2/b9888/de3aNbm5udkELVdXV128eFGXL1+WJF2/fl3NmzfX999/r8mTJ2vnzp06cOCA1q9fL0k209AlydPT0yYwSmkLn928edOuz1CqVKlM29LvmbZHr169VL9+fY0bN05JSUm37bN27Vp16NBBnp6edtdhTw2+vr7q3bu3Zs2ape+//14REREqWbKkxo0bd8fp6Xd778TERF2/fl1S2r/piy++qBUrVujatWv666+/tGbNGgUFBWWYKv9P5cuXv+0fQaZPn64DBw5kuB/8ypUrSk5O1uzZszN8Vzp27ChJ1u9LuqJFi9o8T68p/fvy559/SpKeffbZDOecNm2aDMPQ1atX7/g5AAB5F6uXAwDyrIEDB2r8+PH6+OOP9d5772Xar1ixYipatKi2bNly29fTtxjbvn27Lly4oJ07d9psd2VPeLwXFy9ezLTtn0HuTiwWi6ZNm6aAgAAtWLAgw+vR0dH6+uuvb7sY253qqFy5st01pKtRo4Z69eqlmTNn6ueff7ZuJZaZzN7bzc1N3t7e1raXXnpJU6dO1aJFi3Tz5k0lJydryJAhd60nICBAc+fO1cGDB23u677dFmySVLhwYTk7O6t///4aOnTobftUrFjxru/7v4oVKyYpbXQ+s5XoS5YsmaVzAgDyDkI3ACDPKlu2rF577TX99NNPGjBgQKb9OnfurFWrViklJUWPPvpopv0sFoskZRg9/eSTT7Kn4H84fvy4jh49ajPFfMWKFSpYsKDq16+fpXO1bdtWAQEBmjRpkvz8/Gxe27RpkywWizp37nzbY5cvX65nnnnG+nzv3r36/fffFRQUlOn7XblyRQULFpSbm1uG13766SdJUpkyZe5a9/r16/XBBx9YZwzExsZq06ZNat68uZydna39SpcurX/961+aN2+eEhMT1aVLF5UvX/6u53/11Ve1ePFiDR06VF999dVd93D39PRUq1atdPjwYdWuXfu2ny+rmjVrpkKFCunEiRMaNmzYfZ8PAJC3ELoBAHna1KlT79qnV69eWr58uTp27KgRI0aocePGcnV11R9//KEdO3aoW7dueuqpp9S0aVMVLlxYQ4YM0YQJE+Tq6qrly5fr6NGjptRepkwZde3aVe+8845Kly6tZcuWKTw8XNOmTbvtNPC7mTZtmho0aKBLly6pRo0a1va1a9cqICAg08B58OBBBQUF6V//+pfOnTuncePGqWzZsnr55Zczfa8dO3ZoxIgR6tu3r5o2baqiRYvq0qVLWrlypbZs2aLAwECVK1furjU7OzsrICBAI0eOVGpqqqZNm6aYmJjbrhI+YsQI6x9NFi9efNdzS2kj2itXrlTv3r1Vq1YtvfTSS6pfv77c3d116dIlbdu2TZJsthObNWuWHn/8cTVv3lwvvfSSKlSooNjYWP3yyy/atGmTXSvF/y9vb2/Nnj1bAwYM0NWrV/Xss8+qRIkS+uuvv3T06FH99ddfmj9/fpbOCQDIOwjdAIB8z9nZWZ9//rlmzZqlpUuXasqUKXJxcVG5cuXUsmVL1apVS1LalO4vvvhCo0aNUr9+/eTl5aVu3bpp9erVWR55tkfdunU1aNAgTZgwQadPn1aZMmU0Y8YMvfrqq/d0vnr16ql3795asWKFtS0uLk5bt269Y6gLDQ3V0qVL1atXLyUkJKhVq1aaNWuWihQpkukxjz32mAYPHqwdO3Zo6dKlunz5sgoUKKDq1atr9uzZeumll+yqediwYbp586aCg4Otfyz44osv1KxZswx9GzdurAoVKqhAgQJq06aNXeeX0hY6O3bsmGbOnKnFixdr4sSJSk1NValSpdS4cWNt2LBB3bp1s/avXr26fvjhB7377rt66623dOnSJRUqVEhVqlSx3tedVf369VP58uX1/vvv68UXX1RsbKxKlCihunXrauDAgfd0TgBA3mAxDMNwdBEAADxoKlSooJo1a2rz5s2mvs+aNWvUt29f/fnnnxlC9JIlSzRo0CAdOHAgS/tYO0pERITq1KmjuXPn3nEUHgCA3ITVywEAyMd69OihpKSkO45a53ZnzpzR9u3b9cILL6h06dKMDAMA8hRCNwAAyNXeffddBQQE6Pr16/rvf/97T/e7AwDgKEwvBwAAAADAJIx0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGCSB26f7tTUVF24cEEFCxaUxWJxdDkAAAAAgDzIMAzFxsaqTJkycnLKfDz7gQvdFy5ckJ+fn6PLAAAAAADkA+fOnVO5cuUyff2BC90FCxaUlPYP4+Pj4+BqAAAAAAB5UUxMjPz8/KwZMzMPXOhOn1Lu4+ND6AYAAAAA3Je73bbMQmoAAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAA+UR8fLyjS8A/ELoBAAAAIA+LiIhQUFCQvL295eXlJW9vbwUFBSkiIsLRpUEODt27du1Sly5dVKZMGVksFm3cuNHuY7/99lu5uLiobt26ptUHAAAAALnZqlWr1LBhQ4WGhiouLk6SFBcXp9DQUDVs2FCrVq1ycIVwaOiOi4tTnTp1NGfOnCwdFx0drcDAQLVp08akygAAAAAgd4uIiFBgYKCSkpJu+3pSUpICAwN17NixHK4M/8vFkW/eoUMHdejQIcvHvfjii+rTp4+cnZ2zNDoOAAAAAPlFSEhIpoE7XVJSkkJCQrRw4cIcqgr/lOfu6V68eLHOnDmjCRMmOLoUAAAAAHAYe6eOr1y50uRKcCcOHenOqtOnT+uNN97Q7t275eJiX+kJCQlKSEiwPo+JiTGrPAAAAADIEfHx8dZ7uO8mLi5ON27cUIECBUyuCreTZ0a6U1JS1KdPH02cOFEPP/yw3cdNmTJFvr6+1oefn5+JVQIAAACA+Tw9PeXl5WVXXy8vLwK3A+WZ0B0bG6uDBw9q2LBhcnFxkYuLiyZNmqSjR4/KxcVF27dvv+1xY8eOVXR0tPVx7ty5HK4cAAAAALJfr1697OrXu3dvkyvBneSZ6eU+Pj4ZVt2bN2+etm/frrVr16pixYq3Pc7d3V3u7u45USIAAAAA5Jjg4GCFhYXdcTE1V1dXBQcH52BV+CeHhu7r16/rl19+sT4/e/asjhw5oiJFiqh8+fIaO3aszp8/r7CwMDk5OalmzZo2x5coUUIeHh4Z2gEAAAAgv6tdu7bCwsIy3TbM1dVVYWFhqlWrlgOqQzqHTi8/ePCg6tWrp3r16kmSRo4cqXr16mn8+PGSpKioKEVGRjqyRAAAAADItXr16qVDhw4pKCjIeo+3l5eXgoKCdOjQIbunoMM8FsMwDEcXkZNiYmLk6+ur6Oho+fj4OLocAAAAAMg2rFKec+zNlnlmITUAAAAAwJ0RuHMfQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACZxaOjetWuXunTpojJlyshisWjjxo137L9+/XoFBASoePHi8vHxUZMmTbR169acKRYAAAAAgCxyaOiOi4tTnTp1NGfOHLv679q1SwEBAfryyy916NAhtWrVSl26dNHhw4dNrhQAAAAAgKyzGIZhOLoISbJYLNqwYYO6d++epeNq1Kihnj17avz48Xb1j4mJka+vr6Kjo+Xj43MPlQIAAAAAHnT2Zss8fU93amqqYmNjVaRIEUeXAgAAAABABi6OLuB+TJ8+XXFxcerRo0emfRISEpSQkGB9HhMTkxOlAQAAAACQd0e6V65cqXfeeUerV69WiRIlMu03ZcoU+fr6Wh9+fn45WCUAAAAA4EGWJ0P36tWr9dxzz2nNmjVq27btHfuOHTtW0dHR1se5c+dyqEoAAAAAwIMuz00vX7lypQYPHqyVK1eqU6dOd+3v7u4ud3f3HKgMAAAAAABbDg3d169f1y+//GJ9fvbsWR05ckRFihRR+fLlNXbsWJ0/f15hYWGS0gJ3YGCgZs2apccee0wXL16UJBUoUEC+vr4O+QwAAAAAAGTGodPLDx48qHr16qlevXqSpJEjR6pevXrW7b+ioqIUGRlp7f/JJ58oOTlZQ4cOVenSpa2PESNGOKR+AAAAAADuJNfs051T2KcbAAAAAHC/Hoh9ugEAAAAAyM0I3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASRwaunft2qUuXbqoTJkyslgs2rhx412P+eabb9SgQQN5eHioUqVK+vjjj80vFAAAAACAe+DQ0B0XF6c6depozpw5dvU/e/asOnbsqObNm+vw4cN68803FRwcrHXr1plcKQAAAAAAWefiyDfv0KGDOnToYHf/jz/+WOXLl9fMmTMlSdWqVdPBgwf14Ycf6plnnsnam8fFSQULShZL2vPERCkpSXJxkdzdbftJUoECktP//xtFUlJaf2dnycPj3vrGx0uGkdbm7JzWlpwsJSSkHVugwL31vXFDSk1N+wwu///Hm5Ii3byZtb4Wi+TpeavvzZtpr7m5Sa6uWe+bmpr2fpLk5XWrb0JC2mdxdU3rn9W+hpH27yOl1fDPn2dW+trzs8+O78ntfp7Z8T1J/3ne7/fknz/P+/2eZPbzvN/vyf/+PO/3e5LZz/NevydcI7Lel2vErb5cI7Lel2tEWhvXiKz35Rpxqy/XCPv6co3gGiHd+nkmJsouRi4hydiwYcMd+zRv3twIDg62aVu/fr3h4uJiJCYm3vaYmzdvGtHR0dbHuXPnDElGtGQYly7d6jh5smFIhhEUZHsCT8+09rNnb7V99FFaW58+tn2LFUtr//HHW20LFqS1detm29ffP619//5bbcuWpbW1bWvbt3r1tPYdO261bdiQ1ta0qW3fhg3T2jdvvtW2bVtaW506tn1btkxrX7PmVtuePWltlSvb9u3YMa198eJbbYcPp7WVKWPb99ln09rnzLnV9vPPaW2+vrZ9BwxIa3///Vttf/yR1ubiYtv35ZfT2idMuNX2999pbZJh/O93YPTotLbRo2+1JSbe6vv337faJ0xIa3v5Zdv3c3FJa//jj1tt77+f1jZggG1fX9+09p9/vtU2Z05a27PP2vYtUyat/fDhW22LF6e1dexo27dy5bT2PXtuta1Zk9bWsqVt3zp10tq3bbvVtnlzWlvDhrZ9mzZNa//f/+Z27Ehrq17dtm/btmnty5bdatu/P63N39+2b7duae0LFtxq+/HHtLZixWz79umT1v7RR7fazp5Na/P0tO0bFJTWPnnyrbZLl279PP/XiBFpbW++eavt+vVbfa9fv9X+5ptpbSNG2J4jvS/XCK4RhsE1Ih3XiFu4RqThGpGGa0QarhG3cI1I8wBcI6KHDzckGdHR0cad5KmF1C5evKiSJUvatJUsWVLJycm6fPnybY+ZMmWKfH19rQ8/P7+cKBUAAAAAAFkMwzAcXYQkWSwWbdiwQd27d8+0z8MPP6xBgwZp7Nix1rZvv/1Wjz/+uKKiolSqVKkMxyQkJCghIcH6PCYmRn5+foq+cEE+pUox5SO/T/lgWhjTwv7Zl2lhXCMkrhFcI7hGcI1Ia+Macasv1wj7+nKN4BohWX+eMTduyLd4cUVHR8vHx0eZyVOhu0WLFqpXr55mzZplbduwYYN69Oih+Ph4uab/o99BTEyMfH197/oPAwAAAABAZuzNlnlqenmTJk0UHh5u07Zt2zY1bNjQrsANAAAAAEBOcmjovn79uo4cOaIjR45IStsS7MiRI4qMjJQkjR07VoGBgdb+Q4YM0e+//66RI0fq5MmTWrRokUJDQzV69GhHlA8AAAAAwB05dMuwgwcPqlWrVtbnI0eOlCQNGDBAS5YsUVRUlDWAS1LFihX15Zdf6tVXX9XcuXNVpkwZhYSEZH27MAAAAAAAckCuuac7p3BPNwAAAADgfuXLe7oBAAAAAMhLCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmMTu0H316lX98ccfNm3Hjx/XoEGD1KNHD61YsSLbiwMAAAAAIC+zO3QPHTpUM2bMsD6/dOmSmjdvrgMHDighIUEDBw7U0qVLTSkSAAAAAIC8yO7QvW/fPnXt2tX6PCwsTEWKFNGRI0f02Wef6d///rfmzp1rSpEAAAAAAORFdofuixcvqmLFitbn27dv11NPPSUXFxdJUteuXXX69OnsrxAAAAAAgDzK7tDt4+Oja9euWZ/v379fjz32mPW5xWJRQkJCthYHAAAAAEBeZnfobty4sUJCQpSamqq1a9cqNjZWrVu3tr7+888/y8/Pz5QiAQAAAADIi1zs7fjuu++qbdu2WrZsmZKTk/Xmm2+qcOHC1tdXrVqlli1bmlIkAAAAAAB5kd2hu27dujp58qT27t2rUqVK6dFHH7V5vVevXqpevXq2FwgAAAAAQF5lMQzDcHQROSkmJka+vr6Kjo6Wj4+Po8sBAAAAAORB9mZLu0a6Q0JC9MILL8jDw0MhISF37BscHJy1SgEAAAAAyKfsGumuWLGiDh48qKJFi9psG5bhZBaLfv3112wtMLsx0g0AAAAAuF/ZOtJ99uzZ2/5vAAAAAACQObu3DLsdwzD0gN0SDgAAAACA3e4pdIeGhqpmzZry8PCQh4eHatasqU8//TS7awMAAAAAIE+ze8uwdG+//bY++ugjDR8+XE2aNJEkfffdd3r11Vf122+/afLkydleJAAAAAAAeVGWtwwrVqyYZs+erd69e9u0r1y5UsOHD9fly5eztcDsxkJqAAAAAID7ZW+2zPL08pSUFDVs2DBDe4MGDZScnJzV0wEAAAAAkG9lOXT369dP8+fPz9C+YMEC9e3bN1uKAgAAAAAgP8jyPd1S2kJq27Zt02OPPSZJ2rdvn86dO6fAwECNHDnS2m/GjBnZUyUAAAAAAHlQlkP3jz/+qPr160uSzpw5I0kqXry4ihcvrh9//NHaz2KxZFOJAAAAAADkTVkO3Tt27DCjDgAAAAAA8h277+kuU6aMXnrpJW3ZskWJiYlm1gQAAAAAQL5gd+hesWKFPD09NXz4cBUrVkz/+te/tHTpUl29etXM+gAAAAAAyLOyvE+3JB0/flyff/65PvvsMx0+fFhNmjRRt27d1LVrVz300ENm1Jlt2KcbAAAAAHC/TNunW5Jq1KihsWPHat++fYqMjFTfvn21fft21apVSzVr1tQXX3xxz4UDAAAAAJBf3NNId2bi4uK0bds2FSxYUG3bts2u02YrRroBAAAAAPfL3mx5T/t0Z8bLy0tPPfVUdp4SAAAAAIA8y+7Q7eTkdNe9ty0Wi5KTk++7KAAAAAAA8gO7Q/eGDRsyfW3v3r2aPXu2snGmOgAAAAAAeZ7dobtbt24Z2n766SeNHTtWmzZtUt++ffXuu+9ma3EAAAAAAORl97R6+YULF/T888+rdu3aSk5O1pEjR/Sf//xH5cuXz+76AAAAAADIs7IUuqOjo/X666+rcuXKOn78uL7++mtt2rRJNWvWNKs+AAAAAADyLLunl7///vuaNm2aSpUqpZUrV952ujkAAAAAALjF7n26nZycVKBAAbVt21bOzs6Z9lu/fn22FWcG9ukGAAAAANyvbN+nOzAw8K5bhgEAAAAAgFvsDt1LliwxsQwAAAAAAPKfe1q9HAAAAAAA3J1dI91PP/20lixZIh8fHz399NN37Jvb7+kGAAAAACCn2BW6fX19rfdz+/r6mloQAAAAAAD5hd2rl+cXrF4OAAAAALhf2b56+T9dunRJp06dksVi0cMPP6wSJUrc66kAAAAAAMiXsryQWkxMjPr376+yZcuqZcuWatGihcqWLat+/fopOjrajBoBAAAAAMiTshy6g4KC9P3332vz5s26du2aoqOjtXnzZh08eFDPP/+8GTUCAAAAAJAnZfmebi8vL23dulWPP/64Tfvu3bvVvn17xcXFZWuB2Y17ugEAAAAA98vebJnlke6iRYvedgVzX19fFS5cOKunAwAAAAAg38py6H7rrbc0cuRIRUVFWdsuXryo1157TW+//Xa2FgcAAAAAQF6W5dA9f/587du3T/7+/qpcubIqV66s8uXLa+/evfrkk09Uv35968Me8+bNU8WKFeXh4aEGDRpo9+7dd+y/fPly1alTR56enipdurQGDRqkK1euZPVjAAAAAABguixvGda9e/dse/PVq1frlVde0bx589SsWTN98skn6tChg06cOKHy5ctn6L9nzx4FBgbqo48+UpcuXXT+/HkNGTJEQUFB2rBhQ7bVBQAAAABAdsjyQmrZ6dFHH1X9+vU1f/58a1u1atXUvXt3TZkyJUP/Dz/8UPPnz9eZM2esbbNnz9b777+vc+fO2fWeLKQGAAAAALhfpi2kll0SExN16NAhtWvXzqa9Xbt22rt3722Padq0qf744w99+eWXMgxDf/75p9auXatOnTpl+j4JCQmKiYmxeQAAAAAAkBMcFrovX76slJQUlSxZ0qa9ZMmSunjx4m2Padq0qZYvX66ePXvKzc1NpUqVUqFChTR79uxM32fKlCny9fW1Pvz8/LL1cwAAAAAAkBmHhe50FovF5rlhGBna0p04cULBwcEaP368Dh06pC1btujs2bMaMmRIpucfO3asoqOjrQ97p6EDAAAAAHC/sryQWnYpVqyYnJ2dM4xqX7p0KcPod7opU6aoWbNmeu211yRJtWvXlpeXl5o3b67JkyerdOnSGY5xd3eXu7t79n8AAAAAAADuwmEj3W5ubmrQoIHCw8Nt2sPDw9W0adPbHhMfHy8nJ9uSnZ2dJaWNkAMAAAAAkJtkeaQ7JSVFS5Ys0ddff61Lly4pNTXV5vXt27fbfa6RI0eqf//+atiwoZo0aaIFCxYoMjLSOl187NixOn/+vMLCwiRJXbp00fPPP6/58+frySefVFRUlF555RU1btxYZcqUyepHAQAAAADAVFkO3SNGjNCSJUvUqVMn1axZM9P7r+3Rs2dPXblyRZMmTVJUVJRq1qypL7/8Uv7+/pKkqKgoRUZGWvsPHDhQsbGxmjNnjkaNGqVChQqpdevWmjZt2j3XAAAAAACAWbK8T3exYsUUFhamjh07mlWTqdinGwAAAABwv0zbp9vNzU2VK1e+r+IAAAAAAHgQZDl0jxo1SrNmzWLhMgAAAAAA7iLL93Tv2bNHO3bs0P/93/+pRo0acnV1tXl9/fr12VYcAAAAAAB5WZZDd6FChfTUU0+ZUQsAAAAAAPlKlkP34sWLzagDAAAAAIB8J8v3dAMAAAAAAPtkeaRbktauXas1a9YoMjJSiYmJNq/98MMP2VIYAAAAAAB5XZZHukNCQjRo0CCVKFFChw8fVuPGjVW0aFH9+uuv6tChgxk1AkCuFR8f7+gSAAAAkItlOXTPmzdPCxYs0Jw5c+Tm5qYxY8YoPDxcwcHBio6ONqNGAMhVIiIiFBQUJG9vb3l5ecnb21tBQUGKiIhwdGkAAADIZbIcuiMjI9W0aVNJUoECBRQbGytJ6t+/v1auXJm91QFALrNq1So1bNhQoaGhiouLkyTFxcUpNDRUDRs21KpVqxxcIQAAAHKTLIfuUqVK6cqVK5Ikf39/7du3T5J09uxZGYaRvdUBQC4SERGhwMBAJSUl3fb1pKQkBQYG6tixYzlcGQAAAHKrLIfu1q1ba9OmTZKk5557Tq+++qoCAgLUs2dP9u8GkK+FhIRkGrjTJSUlKSQkJIcqAgAAQG5nMbI4PJ2amqrU1FS5uKQtfL5mzRrt2bNHlStX1pAhQ+Tm5mZKodklJiZGvr6+io6Olo+Pj6PLAZCHeHt7W6eU34mXl5euX7+eAxUBAADAUezNllkO3XkdoRvAvYiPj5eXl1eW+hcoUMDEigAAAOBI9mbLLE8vl6Tdu3erX79+atKkic6fPy9JWrp0qfbs2XNv1QJALufp6Wl36Pby8iJwAwAAQNI9hO5169bpySefVIECBXT48GElJCRIkmJjY/Xvf/872wsEgNyiV69edvXr3bu3yZUAAAAgr8hy6J48ebI+/vhjLVy4UK6urtb2pk2b6ocffsjW4gAgNwkODra57t2Oq6urgoODc6giAAAA5HZZDt2nTp1SixYtMrT7+Pjo2rVr2VETAORKtWvXVlhYWKbB29XVVWFhYapVq1YOVwYAAIDcKsuhu3Tp0vrll18ytO/Zs0eVKlXKlqIAILfq1auXDh06pKCgIOs93l5eXgoKCtKhQ4fsnoIOAACAB4NLVg948cUXNWLECC1atEgWi0UXLlzQd999p9GjR2v8+PFm1AgAuUqtWrW0cOFCLVy4UDdu3GDRNAAAAGQqy6F7zJgxio6OVqtWrXTz5k21aNFC7u7uGj16tIYNG2ZGjQCQaxG4AQAAcCf3vE93fHy8Tpw4odTUVFWvXl3e3t7ZXZsp2KcbAAAAAHC/7M2WWR7pTufp6amGDRve6+EAAAAAAOR7dofuwYMH29Vv0aJF91wMAAAAAAD5id2he8mSJfL391e9evV0jzPSAQAAAAB4oNgduocMGaJVq1bp119/1eDBg9WvXz8VKVLEzNoAAAAAAMjT7N6ne968eYqKitLrr7+uTZs2yc/PTz169NDWrVsZ+QYAAAAA4DbuefXy33//XUuWLFFYWJiSkpJ04sSJPLGCOauXAwAAAADul73Z0u6R7n+yWCyyWCwyDEOpqan3ehoAAAAAAPKtLIXuhIQErVy5UgEBAapataqOHTumOXPmKDIyMk+McgMAAAAAkJPsXkjt5Zdf1qpVq1S+fHkNGjRIq1atUtGiRc2sDQAAAACAPM3ue7qdnJxUvnx51atXTxaLJdN+69evz7bizMA93QAAAACA+2VvtrR7pDswMPCOYRsAAAAAANiyO3QvWbLExDIAAAAAAMh/7nn1cgAAAAAAcGeEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABM4vDQPW/ePFWsWFEeHh5q0KCBdu/efcf+CQkJGjdunPz9/eXu7q6HHnpIixYtyqFqAQAAAACwn4sj33z16tV65ZVXNG/ePDVr1kyffPKJOnTooBMnTqh8+fK3PaZHjx76888/FRoaqsqVK+vSpUtKTk7O4coBAAAAALg7i2EYhqPe/NFHH1X9+vU1f/58a1u1atXUvXt3TZkyJUP/LVu2qFevXvr1119VpEiRe3rPmJgY+fr6Kjo6Wj4+PvdcOwAAAADgwWVvtnTY9PLExEQdOnRI7dq1s2lv166d9u7de9tjPv/8czVs2FDvv/++ypYtq4cfflijR4/WjRs3Mn2fhIQExcTE2DwAAAAAAMgJDptefvnyZaWkpKhkyZI27SVLltTFixdve8yvv/6qPXv2yMPDQxs2bNDly5f18ssv6+rVq5ne1z1lyhRNnDgx2+sHAAAAAOBuHL6QmsVisXluGEaGtnSpqamyWCxavny5GjdurI4dO2rGjBlasmRJpqPdY8eOVXR0tPVx7ty5bP8MAAAAAADcjsNGuosVKyZnZ+cMo9qXLl3KMPqdrnTp0ipbtqx8fX2tbdWqVZNhGPrjjz9UpUqVDMe4u7vL3d09e4sHAAAAAMAODhvpdnNzU4MGDRQeHm7THh4erqZNm972mGbNmunChQu6fv26te3nn3+Wk5OTypUrZ2q9AAAAAABklUOnl48cOVKffvqpFi1apJMnT+rVV19VZGSkhgwZIiltanhgYKC1f58+fVS0aFENGjRIJ06c0K5du/Taa69p8ODBKlCggKM+BgAAAAAAt+XQfbp79uypK1euaNKkSYqKilLNmjX15Zdfyt/fX5IUFRWlyMhIa39vb2+Fh4dr+PDhatiwoYoWLaoePXpo8uTJjvoIAAAAAABkyqH7dDsC+3QDAAAAAO5Xrt+nGwAAAACA/I7QDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAB4oMTHxzu6BAAPEEI3AAAA8r2IiAgFBQXJ29tbXl5e8vb2VlBQkCIiIhxdGoB8zmIYhuHoInJSTEyMfH19FR0dLR8fH0eXAwAAAJOtWrVKgYGBSkpKyvCaq6urwsLC1KtXLwdUBiAvszdbMtINAACAfCsiIiLTwC1JSUlJCgwM1LFjx3K4MgAPCkI3AAAA8q2QkJBMA3e6pKQkhYSE5FBFAB40TC8HAABAvuXt7a24uLi79vPy8tL169dzoCIA+QXTywEAAPBAi4+PtytwS1JcXJxu3LhhckUAHkSEbgAAAORLnp6e8vLysquvl5eXChQoYHJFAB5EhG4AAADkW/auSt67d2+TKwHwoCJ0AwAAIN8KDg6Wq6vrHfu4uroqODg4hyoC8KAhdAMAACDfql27tsLCwjIN3un7dNeqVSuHKwPwoCB0AwAAIF/r1auXDh06pKCgIOs93l5eXgoKCtKhQ4fsnoIOAPeCLcMAAADwQLlx4waLpgG4b2wZBgAAANwGgRtATiJ0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmcXjonjdvnipWrCgPDw81aNBAu3fvtuu4b7/9Vi4uLqpbt665BQIAAAAAcI8cGrpXr16tV155RePGjdPhw4fVvHlzdejQQZGRkXc8Ljo6WoGBgWrTpk0OVQoAAAAAQNZZDMMwHPXmjz76qOrXr6/58+db26pVq6bu3btrypQpmR7Xq1cvValSRc7Oztq4caOOHDli93vGxMTI19dX0dHR8vHxuZ/yAQAAAAAPKHuzpcNGuhMTE3Xo0CG1a9fOpr1du3bau3dvpsctXrxYZ86c0YQJE8wuEQAAAACA++LiqDe+fPmyUlJSVLJkSZv2kiVL6uLFi7c95vTp03rjjTe0e/duubjYV3pCQoISEhKsz2NiYu69aAAAAAAAssDhC6lZLBab54ZhZGiTpJSUFPXp00cTJ07Uww8/bPf5p0yZIl9fX+vDz8/vvmsGAAAAAMAeDgvdxYoVk7Ozc4ZR7UuXLmUY/Zak2NhYHTx4UMOGDZOLi4tcXFw0adIkHT16VC4uLtq+fftt32fs2LGKjo62Ps6dO2fK5wEAAAAA4J8cNr3czc1NDRo0UHh4uJ566ilre3h4uLp165ahv4+Pj44dO2bTNm/ePG3fvl1r165VxYoVb/s+7u7ucnd3z97iAQAAAACwg8NCtySNHDlS/fv3V8OGDdWkSRMtWLBAkZGRGjJkiKS0Uerz588rLCxMTk5Oqlmzps3xJUqUkIeHR4Z2AAAAAAByA4eG7p49e+rKlSuaNGmSoqKiVLNmTX355Zfy9/eXJEVFRd11z24AAAAAAHIrh+7T7Qjs0w0AAAAAuF+5fp9uAAAAAADyO0I3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXTDdPHx8Y4uAQAAAAAcgtANU0RERCgoKEje3t7y8vKSt7e3goKCFBER4ejSAAAAACDHWAzDMBxdRE6KiYmRr6+voqOj5ePj4+hy8qVVq1YpMDBQSUlJGV5zdXVVWFiYevXq5YDKAAAAACB72JstGelGtoqIiMg0cEtSUlKSAgMDdezYsRyuDAAAAAByHqEb2SokJCTTwJ0uKSlJISEhOVQRAAAAADgO08uRrby9vRUXF3fXfl5eXrp+/XoOVAQAAAAA2Y/p5chx8fHxdgVuSYqLi9ONGzdMrggAAAAAHIvQjWzj6ekpLy8vu/p6eXmpQIECJlcEAAAAAI5F6Ea2sndV8t69e5tcCQAAAAA4HqEb2So4OFiurq537OPq6qrg4OAcqggAAAAAHIfQjWxVu3ZthYWFZRq80/fprlWrVg5XBgAAAAA5j9CNbNerVy8dOnRIQUFB1nu8vby8FBQUpEOHDtk9BR0AAAAA8jqHh+558+apYsWK8vDwUIMGDbR79+5M+65fv14BAQEqXry4fHx81KRJE23dujUHq4W9atWqpYULF+r69euKj4/X9evXtXDhQka4AQAAADxQHBq6V69erVdeeUXjxo3T4cOH1bx5c3Xo0EGRkZG37b9r1y4FBAToyy+/1KFDh9SqVSt16dJFhw8fzuHKkRWsUg4AAADgQWUxDMNw1Js/+uijql+/vubPn29tq1atmrp3764pU6bYdY4aNWqoZ8+eGj9+vF397d3AHAAAAACAzNibLR020p2YmKhDhw6pXbt2Nu3t2rXT3r177TpHamqqYmNjVaRIETNKBAAAAADgvrg46o0vX76slJQUlSxZ0qa9ZMmSunjxol3nmD59uuLi4tSjR49M+yQkJCghIcH6PCYm5t4KBgAAAAAgixy+kJrFYrF5bhhGhrbbWblypd555x2tXr1aJUqUyLTflClT5Ovra334+fndd80AAAAAANjDYaG7WLFicnZ2zjCqfenSpQyj3/+0evVqPffcc1qzZo3atm17x75jx45VdHS09XHu3Ln7rh0AAAAAAHs4LHS7ubmpQYMGCg8Pt2kPDw9X06ZNMz1u5cqVGjhwoFasWKFOnTrd9X3c3d3l4+Nj8wAAAAAAICc47J5uSRo5cqT69++vhg0bqkmTJlqwYIEiIyM1ZMgQSWmj1OfPn1dYWJiktMAdGBioWbNm6bHHHrOOkhcoUEC+vr4O+xwAAAAAANyOQ0N3z549deXKFU2aNElRUVGqWbOmvvzyS/n7+0uSoqKibPbs/uSTT5ScnKyhQ4dq6NCh1vYBAwZoyZIlOV0+AAAAAAB35NB9uh2BfboBAAAAAPcr1+/TDQAAAABAfkfoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkDl293BHS142LiYlxcCUAAAAAgLwqPVPebW3yBy50x8bGSpL8/PwcXAkAAAAAIK+LjY2Vr69vpq8/cFuGpaam6sKFCypYsKAsFoujy8n3YmJi5Ofnp3PnzrFFG/Idvt/Iz/h+I7/iu438jO93zjIMQ7GxsSpTpoycnDK/c/uBG+l2cnJSuXLlHF3GA8fHx4f/8JFv8f1Gfsb3G/kV323kZ3y/c86dRrjTsZAaAAAAAAAmIXQDAAAAAGASQjdM5e7urgkTJsjd3d3RpQDZju838jO+38iv+G4jP+P7nTs9cAupAQAAAACQUxjpBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAOQbbMwDILdxcXQByLtSU1Pl5MTfbZB/JCYmys3NTVLaL20Wi8XBFQHm4PuN/CY6OlrXr1+XYRgqV66co8sBABskJtjtxIkTWrZsmU6fPi1JBG7kK6dOndLo0aN14MABSZLFYmG0BPlGZGSkFi9erHfffVd79+7l+4185fjx42rfvr06dOigqlWrauLEibp8+bKjywKy1W+//aYzZ84oJibG2sZ1PO9gpBt2iY2NVatWrSRJnTt31o0bNzR16lQVLlxYBQsWtI6aMHqCvOjMmTNq3bq1oqKidP36dbm6uqpu3bp8p5EvHDt2TF27dlWZMmV07do1TZw4UWvWrNHTTz/N9xt53k8//aRWrVpp4MCB6ty5s44cOaKxY8eqXr166tq1q6PLA7LFunXrNHr0aMXExKhly5bq0aOHevXqxe8peYjF4E8ksNOIESN09OhRvfPOO5o6daquXr2qChUqaOjQoXrsscfk7u4u6dZf3bgAIC+4efOmXnvtNV2+fFmdO3fWW2+9pccff1yjRo1S3bp1JTEVF3nX2bNn1bp1a/Xs2VPjx4+Xi4uLJk2apJUrV2rfvn0qXry4o0sE7llMTIwGDhyokiVLav78+db2fv36KTo6Wps2beL6jTzv3Llz6tChg0aMGKFChQpp5cqVunTpkvr3768XX3xREr+n5AWMdOOukpKS5OrqqqefflpRUVGqUaOGtmzZom3btmnv3r1q1aqV+vTpo8aNGys4OFiGYTD1HHmGk5OTWrdurbi4OPXt21flypXTwIEDZRiGRo8ezYg38qykpCSFhoaqfv36evvtt+Xp6SlJateunZYvX860ROR5f//9t5KSktStWzdJUkpKipydnfXII4/o66+/lsQAAPI+V1dX1alTR4GBgXJ3d1fdunU1ZcoU/ec//5Ekvfjii/yekgeQjJCp3377TdHR0XJ1dZUkNWjQQD///LPGjx8vKe0Xt99//13FixdX4cKFNWXKFFWtWlWzZ8/mlznkGW5ubmrbtq369esnSWrZsqUWLVqkb7/9Vh9++KGOHDli7RsREeGgKoGsc3V1VY0aNfTwww/Ly8vL2l63bl0lJibq/PnzSk1NdWCFwP3x9/fX6NGj1b59e0m3Ztr5+/tbZ9+lu3TpUo7XB9yPrVu3qm/fvnrzzTd19epV63e6SpUqGjt2rKpWrarly5dr1qxZkvgDU25H6MZtJSUlafDgwapWrZquXbsmSfL29taMGTN07NgxRUZGatCgQdq6dat27typ2bNn69ChQ2revLk6d+7Mf/jI1WJiYnTx4kVduXJFklSwYEFJaSvyG4ahVq1aKTQ01Bq8Dxw4oFdffVUDBgxQdHS0I0sH7urgwYNq06aNUlNT1bt3b02ZMkXSrUCSPhvJYrFYZyV9//33io2NdVjNgL3Sr9/pC6W1bNlSUtr128UlbQLnjRs3dOHCBesfld59911NnDhRN2/edEzRQBZt375dHTt2VEJCgvbv36+vvvrKei2X0oL3uHHjVLJkSX355ZfW39WRexG6cVuurq4KCQlRuXLl1KxZM/3999+SJD8/P7m4uOiJJ57Q7t279dlnn6latWpKTU1VmTJl9Omnn+qhhx5ycPVA5o4fP67u3bvr8ccfV8+ePbV8+XLra+lBJDU1Va1bt9aiRYv0/fffq3v37lqwYIFCQ0Pl6+vrwOqBOzt69KhatWql6tWr29zmk5qaKovFouTkZN28eVPOzs7WPza98cYbCggIIJAg1/vf63evXr0yXL/TWSwWOTs7y8nJSRMmTNCECRMUFBQkDw8PR5QNZMmZM2d09epVzZo1S2vXrtXmzZs1atQoLV26VB988IG1X+XKlTV16lQtWbJEhQoVclzBsAuhGxmkj4ZUr15dYWFhKlSokNq2bau///5bVapU0TPPPKPffvtNc+bMUaNGjSSxfRjyhqNHj+qxxx5TrVq1NHHiRHl4eCgsLEyJiYnWPun7z6ePeFerVk03btzQ/v37Vb9+fQdWD9zZ0aNH1aRJEwUHB2v27Nk2r6Vfo11cXOTm5qaUlBQZhqEJEyZo7ty5+uqrr1hUDbmaPdfv9P/t6empChUqaPLkyZo2bZoOHjyoevXqOap0wG6///67GjVqpBdffNG6DkeFChX00ksvqVu3bgoNDdWMGTOs/R966CGVLl3aUeUiC1hIDVY3b96Uh4eHLBaLdfG0Rx55RE2aNNGMGTPUsmVLffPNNxo0aJA2btyoEydOqH379izcgDwhIiJCLVq0UHBwsN577z1JaTM3Ro0apRMnTujGjRtq0qSJnJycrFMSX3/9dW3evFmHDx9WzZo1HVk+cEdnzpxRy5YtNWjQIOv3W5IWLVokV1dX9e/f39rm6uoqHx8fDR06VDt37tTevXvVoEEDR5QN2MXe67ebm5skKTk5WZ999pm2b9+ub7/9lj+YIs/w9PTU66+/runTp+vw4cPWdn9/f7300ktydnbWtGnT5OrqquHDhzuwUmQVw5OQJJ0/f16BgYHasWOHJFkXT3v//fe1ZMkSLVy4UO7u7mrRooWSk5PVsGFDhYSEWKcsArlZbGysAgIC5O/vbxNItmzZotOnT6tbt27q2rWrWrRooZSUFDk5OcnJyUmdOnXSkSNHVKdOHQdWD9zdtm3b5OrqqiJFiujChQuSpH//+9965ZVXVKVKFWu/1NRUXb16VSdPntSePXt04MABAjdytaxevyWpSJEiKlGihPbt28f3G3lK8eLFNXDgQI0ZM0ahoaGaNGmS9bXy5cvrueee09ChQ9WxY0cHVol7YgCGYZw5c8Zo0qSJ0bFjR2PPnj2GYRjGlClTjCJFihjh4eGGYRjGiRMnjDp16hjNmzc3fvjhB6NRo0ZGZGSkI8sG7LZo0SLD3d3dmDRpkmEYhjF16lTDx8fHWLNmjXH48GFj9erVho+PjzFkyBAHVwrcm3//+99GvXr1jIkTJxpvvPGGUaJECeP//u//btv3o48+MiIiInK4QuDe2Hv9fvHFF63HXLlyxVHlAlly4sQJ4+uvvza2bdtmbfvzzz+NDz/80ChUqJAxceJEm/5JSUk5XSKygcUw2NsJaU6fPq3g4GC5u7urRIkS2rhxo5YtW6Z27dpZ+/z000/q0KGDihcvrvDwcBaVQq527tw5bdmyRfHx8WrQoIGuXbumbt26qVmzZjp16pSWLVumgIAASVJcXJy6dOmiQoUKaf369Q6uHLBf+t7EkjR58mQtXbpU586d06effqo+ffpwCxDyJK7feBBs2LBBr732mqS0qeUFCxbUtm3b5OXlpUuXLmnp0qWaOnWqnnvuOU2dOtXB1eJ+cE83rKpUqaJZs2Zp2LBhWrZsmd59911r4E5fXOqRRx7Rli1b5O7uTuBGrhYREaFu3bqpUKFCOnPmjCQpJCREmzdvVpcuXdS7d28FBARYFw708vJS4cKFVa5cOettEwQV5AXOzs7Wa/Rbb72lAgUKKDQ0VD/++KPOnz+vsmXLEryRp9zv9ZvFXZEXfPXVVxo4cKA++OADBQYG6uuvv1aXLl3UokULbdu2TSVKlFD//v0VHx+v0NBQjR49WkWLFuVankdxVYKNhx9+WPPnz1fz5s319ddfa8+ePZJks7hU1apVVaFCBQdWCdxZRESEmjRpop49e2r79u0KDw9X9+7dNWbMGPn5+WnhwoVasWKF3nvvPWsYGT9+vHbt2qWXX37ZunUYkFekX58ladSoUerXr5+2bNmiWbNm6cKFC3yfkWdkx/UbyO1u3LihVatWaezYsXrhhRd09epVDR06VD179lRiYqJat26tq1evqkSJEhoyZIh++OEHFStWjGt5HsZINzJ46KGHNGfOHAUHB2vy5Ml6++231axZM/6PDHnCuXPn1KZNG3Xq1Mk6FevRRx9VVFSUNmzYoOvXr2vQoEFKTk7WkCFD5OnpqWvXrun999/X3r17VbVqVQd/AiBrkpOT5eLiorNnz2ratGn68MMP9eabb8owDH322WeKi4vTW2+9xbYyyPW4fuNBsH//ftWtW1fdu3dX2bJldfXqVXXu3Fnt27fX/PnztWzZMg0YMECNGjXSwYMH2c4xnyB047aqVKmikJAQjRw5UqNHj9ZHH32kxx57zNFlAXeVkpKiihUrKiEhQXv27NHjjz8uSSpRooScnZ2tfzx6/vnnZbFY9MILL0iSDh48yLYyyNUSExOtWyL973RxFxcX/fbbb3r88ccVEBAgDw8PSdK4ceMUHx+vPXv2WO/5BnIzrt/I77Zu3apnnnlGGzZsUOfOnSVJmzZtkpeXl958801ZLBYVL15c7du3V0pKiq5cuaLChQs7uGpkBxZSwx399NNPevvttzV9+nSVL1/e0eUAdklfFDA1NVUzZ85UuXLlVLlyZQUGBuqDDz6w6btu3TpVr15d1apVc1C1wN2dOnVKc+fOVf/+/dWoUSOb1+Lj49W5c2dVqFBBoaGhslgsNve1XrlyRUWLFnVE2UCWcf1GfvXHH39o6tSpqlq1qs0e2zNnztTEiRP1999/S0r7g+nVq1c1c+ZMubu7O6pcZDNCN+7qf0dXgLzi9OnTGjFihOLj4xUREaEBAwboo48+kiQW2kGecubMGbVo0UJRUVEaOHCggoODVbduXevr8fHxOnnypOrXr29zvx/fc+RVXL+R3xw8eFATJ07UuXPnNGvWLLVs2dK688Tvv/+u1q1bKyUlRdWqVdPu3bv13XffqVatWo4uG9mIqxbuisCNvCh9NX5nZ2f5+Pjoqaeesr7GQiTIK27evKmZM2eqRYsWWrp0qXbs2KHp06fryJEj1j6enp5q0KBBhu81wQR5Fddv5Deurq76+++/dfz4ce3bt09S2s4ThmGoXLlyWrt2rbp06aLq1atr//79BO58iJFuAPnaL7/8ouHDh8swDOuigEBekZiYqC+++EJxcXHq16+fvvnmGw0cOFDNmjXT6NGjrSPe/3uPN9uDIb/g+o287siRIypevLjKli2rU6dOKTg4WDExMRo1apSeffbZDP3TR7+R/xC6AeR7p0+f1siRI3X58mUWBUSeExsbq4IFC1qf79ixQ4MHD7YJ3oZh6NixY6pdu7YDKwWyH9dv5FXnz59Xjx495OPjo9DQUJUpU0Y//vijRo4cKUkaMmSInn76aUmE7QcBc88A5HtVqlTRBx98oHLlyqlMmTKOLge4o5iYGF28eFFXrlyRJGvgTk1NlWEYatWqlUJDQ/Xtt9/qww8/1IEDB/Tqq69qwIABio6OdmTpQLbj+o28qmzZsurfv78SEhI0fPhwXbhwQTVr1tT06dMlSZ9++qlWrlwpSQTuBwAj3QAeGCwKiNzu+PHjGj58uCIjI1WhQgUNGjRIffv2temTvpDUjh079MILLyg+Pl5///239uzZw7ZJyLe4fiOv+Odif4sWLdLixYtVokQJzZ492zri/dxzz6lcuXJasmSJzWwm5E+EbgAAcoGjR4/q8ccf1+DBg9W4cWOtXLlSSUlJ2rRpkzVspP8yl37fdteuXbVnzx7t2rVLNWvWdPAnAIAH2/fff6+wsDBNmTJFPj4+1vbFixdr7ty5qlSpkmbPnq2SJUvqxIkT8vb2ZkveBwTTywEAcLCIiAi1aNFCwcHBmjVrlvr27asxY8bo6tWrOnHihL777jtJaSuSp08zHzNmjDZv3qwdO3YQuAEgFwgPD9euXbv09ttvKzY21to+aNAgPfHEE9q0aZMCAwN14cIFVa9encD9ACF0AwDgQLGxsQoICJC/v7/ee+89a/uWLVt0+vRpdevWTV27dlWLFi2UkpIiJycnOTk5qVOnTjpy5Ijq1KnjwOoBAOnGjBmjfv36ad++fRo7dqzNOhuNGzdWjRo1VLhwYaWkpDiwSjgCoRsAAAcqWLCgpk6dqp9//lnvvvuuJGnatGmaO3euFi5cqM8++0xz587V0aNHNWzYMOtxLVu2ZLVyAHCwkydPat++fdq6davc3Nw0ZswY/etf/9LBgwdtgndERIQ6duyojz/+WH5+fg6uGjnNxdEFAADwIDp37py2bNmi+Ph4NWjQQGvXrlW3bt0UHh6uU6dOae3atQoICJCUtoJzgwYN9Oeffzq4agBAuvXr12vEiBEqV66cTp06pSZNmmjEiBEaOXKkUlNTtWHDBj388MNq0KCBdu7cqR9++EGFChVydNlwAEI3AAA5LCIiQt26dVOhQoV05swZSVJISIg2b96sLl26qHfv3goICFD6WqdeXl4qXLiwypUrp9TUVFksFlksFkd+BAB4oH333Xd6/vnnNW3aNAUFBWnHjh1q06aNOnXqpHbt2mnUqFFq0aKFNm3apOTkZM2YMUOPPPKIo8uGg7B6OQAAOSgiIkJNmjTR8OHD9frrr+vnn3/W3LlztWXLFm3fvl0HDhxQUFCQJk2apLFjx8rJyUnjx4/X/PnztWfPHlWtWtXRHwEAHnghISHauXOn1q9fr9OnT6tjx45q1aqVFixYICltvY70rcD+uY0YHjyMdAMAkEPOnTtnHQmZOnWqJOnRRx9VVFSUNmzYoOvXr2vQoEFKTk7WkCFD5OnpqWvXrun999/X3r17CdwAkEv88ssv1nuzW7VqpU6dOunjjz+WJK1du1YxMTHq37+/XF1dCdwgdAMAkFNSUlJUsWJFJSQkaM+ePXr88cclSSVKlJCzs7P1F7Pnn39eFotFL7zwgiTp4MGDql+/vsPqBgBIV69elYeHhzw9PRUQEKC+fftqyZIlGjx4sKZPn2697Wfbtm1KSkpS79695erq6uCqkRvwZxcAAHJIhQoVtHz5ciUmJurdd9/VyZMnFRsbq6eeekrPP/+8GjdubO0bFBSk//73vzp+/DiBGwAcbOPGjeratavq1q2rCRMmyM3NTYMHD5a7u7vat28vJycnXb16VW+++aY+++wzvf766ypQoICjy0YuwT3dAADksNOnT2vEiBGKj49XRESEBgwYoI8++kgS9/4BQG7zww8/qHXr1ho1apSuXLmiPXv2qEqVKqpdu7b++OMPLVy4UNWrV5eHh4eioqK0ceNG1atXz9FlIxchdAMA4ACnT5/WkCFDdObMGYWFhalFixaSJMMwWJkcAHKJM2fOaOXKlbJYLBo3bpwkadOmTZo9e7YKFy6svn37qmjRotq9e7f8/f3VrFkzlS9f3sFVI7chdAMA4CC//PKLhg8fLsMw9Pbbb6tZs2aOLgkA8P/FxMSoTZs2ioyM1ODBgzVlyhTra59//rlmzpypwoULa9y4cdwGhDti/hoAAA5SuXJlhYSEyNXVVaNHj9a+ffscXRIA4P/z8fHRggULVKhQIe3evVvHjx+3vta1a1eNHj1av/76q2bMmKH4+HgxlonMMNINAICD/fTTT3r77bc1ffp0piUCQC6TvvZG48aNFRwcrBo1alhf27Ztm6pWrSp/f38HVojcjtANAEAukJiYKDc3N0eXAQC4jcOHDysoKEj169fXq6++qurVqzu6JOQhhG4AAAAAuIvDhw9ryJAhqlSpkiZMmKBHHnnE0SUhj+CebgAAAAC4i3r16mnOnDmKioqSr6+vo8tBHsJINwAAAADY6ebNm/Lw8HB0GchDCN0AAAAAAJiE6eUAAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AACBT77zzjurWrevoMgAAyLMI3QAA5EEDBw6UxWKRxWKRq6urKlWqpNGjRysuLs7Rpd3Rzp07ZbFYdO3aNUeXAgBAjnBxdAEAAODetG/fXosXL1ZSUpJ2796toKAgxcXFaf78+Tb9kpKS5Orq6qAqAQB4sDHSDQBAHuXu7q5SpUrJz89Pffr0Ud++fbVx40brlPBFixapUqVKcnd3l2EYioyMVLdu3eTt7S0fHx/16NFDf/75p805p06dqpIlS6pgwYJ67rnndPPmTZvXn3jiCb3yyis2bd27d9fAgQOtzxMSEjRmzBj5+fnJ3d1dVapUUWhoqH777Te1atVKklS4cGFZLBab4wAAyI8Y6QYAIJ8oUKCAkpKSJEm//PKL1qxZo3Xr1snZ2VlSWjj28vLSN998o+TkZL388svq2bOndu7cKUlas2aNJkyYoLlz56p58+ZaunSpQkJCVKlSpSzVERgYqO+++04hISGqU6eOzp49q8uXL8vPz0/r1q3TM888o1OnTsnHx0cFChTI1n8DAAByG0I3AAD5wP79+7VixQq1adNGkpSYmKilS5eqePHikqTw8HBFRETo7Nmz8vPzkyQtXbpUNWrU0IEDB9SoUSPNnDlTgwcPVlBQkCRp8uTJ+uqrrzKMdt/Jzz//rDVr1ig8PFxt27aVJJvQXqRIEUlSiRIlVKhQofv+3AAA5HZMLwcAII/avHmzvL295eHhoSZNmqhFixaaPXu2JMnf398auCXp5MmT8vPzswZuSapevboKFSqkkydPWvs0adLE5j3++fxujhw5ImdnZ7Vs2fJePxYAAPkKI90AAORRrVq10vz58+Xq6qoyZcrYLJbm5eVl09cwDFkslgznyKw9M05OTjIMw6YtfUq7JKaLAwDwD4x0AwCQR3l5ealy5cry9/e/6+rk1atXV2RkpM6dO2dtO3HihKKjo1WtWjVJUrVq1bRv3z6b4/75vHjx4oqKirI+T0lJ0Y8//mh9XqtWLaWmpuqbb765bR1ubm7W4wAAeBAQugEAeAC0bdtWtWvXVt++ffXDDz9o//79CgwMVMuWLdWwYUNJ0ogRI7Ro0SItWrRIP//8syZMmKDjx4/bnKd169b64osv9MUXX+inn37Syy+/bLPndoUKFTRgwAANHjxYGzdu1NmzZ7Vz506tWbNGUtq0d4vFos2bN+uvv/7S9evXc+zfAAAARyB0AwDwALBYLNq4caMKFy6sFi1aqG3btqpUqZJWr15t7dOzZ0+NHz9er7/+uho0aKDff/9dL730ks15Bg8erAEDBlgDe8WKFa3bgKWbP3++nn32Wb388st65JFH9PzzzysuLk6SVLZsWU2cOFFvvPGGSpYsqWHDhpn/4QEAcCCL8c8bswAAAAAAQLZgpBsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADDJ/wOfCI+J2FS4MQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 1000x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.stripplot(data=grouped, x='product', y='piN/piS', size=8, color='black')\n",
    "\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.ylabel('Mean piN/piS')\n",
    "plt.xlabel('Product')\n",
    "plt.title('Mean piN/piS by Gene')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.axhline(y=1, color='red', linestyle='dotted', linewidth=1.5)\n",
    "\n",
    "plt.show()\n",
    "##>1 = positive selection, <1 = purifying"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 587,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>measure</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HA</th>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000397</td>\n",
       "      <td>piS</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>piS</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>piS</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NEP</th>\n",
       "      <td>0.000262</td>\n",
       "      <td>0.000380</td>\n",
       "      <td>piS</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP</th>\n",
       "      <td>0.000237</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>piS</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         mean        sd measure species\n",
       "HA   0.000207  0.000397     piS    duck\n",
       "M1   0.000000  0.000000     piS    duck\n",
       "M2   0.000000  0.000000     piS    duck\n",
       "NEP  0.000262  0.000380     piS    duck\n",
       "NP   0.000237  0.000469     piS    duck"
      ]
     },
     "execution_count": 587,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # cast the dataframe and calculate some mean values for piN \n",
    "# pS_h = df[df['species'] == 'human']\n",
    "# pS_h = pS_h[['file','product','piS']]\n",
    "# pS_h = pS_h.pivot(index='file', columns='product', values='piS')\n",
    "# pS_h.index.name = None\n",
    "# pS_h.columns.name = None\n",
    "# pS_h = pS_h.fillna(0)\n",
    "# pS_h.loc['mean'] = pS_h.mean()\n",
    "# pS_h.loc['sd'] = pS_h.std()\n",
    "# pS_means_h = pS_h.loc[['mean','sd']]\n",
    "# pS_means_h = pS_means_h.transpose()\n",
    "# pS_means_h[\"measure\"] = \"piS\"\n",
    "# pS_means_h[\"species\"] = \"human\"\n",
    "\n",
    "# pS_d = df[df['species'] == 'duck']\n",
    "# pS_d = pS_d[['file','product','piS']]\n",
    "# pS_d = pS_d.pivot(index='file', columns='product', values='piS')\n",
    "# pS_d.index.name = None\n",
    "# pS_d.columns.name = None\n",
    "# pS_d = pS_d.fillna(0)\n",
    "# pS_d.loc['mean'] = pS_d.mean()\n",
    "# pS_d.loc['sd'] = pS_d.std()\n",
    "# pS_means_d = pS_d.loc[['mean','sd']]\n",
    "# pS_means_d = pS_means_d.transpose()\n",
    "# pS_means_d[\"measure\"] = \"piS\"\n",
    "# pS_means_d[\"species\"] = \"duck\"\n",
    "# pS_means_d.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 588,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>measure</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>piN</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>piN</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M2</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>piN</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NEP</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>piN</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>piN</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  index      mean        sd measure species\n",
       "0    HA  0.000253  0.000260     piN   human\n",
       "1    M1  0.000033  0.000051     piN   human\n",
       "2    M2  0.000123  0.000175     piN   human\n",
       "3   NEP  0.000051  0.000113     piN   human\n",
       "4    NP  0.000036  0.000055     piN   human"
      ]
     },
     "execution_count": 588,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine the mean and stdev values for piN and piS into 1 dataframe \n",
    "df1 = pd.concat([pN_means_h, pN_means_d, pS_means_h, pS_means_d])\n",
    "df1.reset_index(inplace=True)\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/rpy2/rinterface/__init__.py:145: RRuntimeWarning: Warning:\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/rpy2/rinterface/__init__.py:145: RRuntimeWarning:  Ignoring unknown parameters: stat\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "%%R -w 800 -h 500 -u px -i df1  # this sets the size of the plot...otherwise, it will go off the page\n",
    "require(ggplot2)\n",
    "library(ggplot2)\n",
    "\n",
    "df1$index = gsub(\"neuram\", \"NA\", df1$index)\n",
    "df1$gene = factor(df1$index, levels=c(\"PB2\",\"PB1\",\"PA\",\"HA\",\"NP\",\"NA\",\"M1\",\"M2\",\"NS1\",\"NEP\"))\n",
    "\n",
    "p <- ggplot(data=df1, aes(x=gene, y=mean, fill=species)) + \n",
    "    geom_col(position=\"dodge\", stat=\"identity\")+\n",
    "    geom_errorbar(data=df1, aes(x=gene, ymin=mean-sd, ymax=mean+sd, color=species),position=position_dodge(0.9), stat=\"identity\")+\n",
    "    facet_wrap(~measure, scales=\"free\")+\n",
    "    scale_fill_manual(values=c(\"#99bfaa\",\"#5c3d46\"))+\n",
    "    scale_color_manual(values=c(\"#99bfaa\",\"#5c3d46\"))+\n",
    "    labs(x=\"gene\",y=\"diversity\")+\n",
    "    #ggtitle(\"number of shared sites in random simulations\") + \n",
    "    scale_y_continuous(limits=c(-0.0006,0.002))+\n",
    "    #scale_x_continuous(limits=c(0,500))+\n",
    "    theme(plot.title = element_text(size=20, hjust=0.5))+\n",
    "    theme(panel.grid.major.y=element_line(colour=NA))+\n",
    "    theme(panel.grid.minor=element_line(colour=NA,size=NA))+    \n",
    "    theme(strip.background = element_rect(colour=NA, fill=NA))+\n",
    "    theme(axis.line.x=element_line(colour=\"black\"))+\n",
    "    theme(axis.line.y=element_line(colour=\"black\"))+theme(strip.text.x=element_text(size=11))+\n",
    "    theme(axis.title.y=element_text(size=16, vjust=8))+\n",
    "    theme(axis.title.x=element_text(size=16, vjust=-8))+\n",
    "    theme(axis.text=element_text(size=16, colour=\"black\"))+\n",
    "    theme(axis.text.x=element_text(size=16, angle=90))+\n",
    "    theme(legend.text=element_text(size=16))+theme(legend.title=element_text(size=16, face=\"plain\"))+\n",
    "    theme(panel.margin=unit(1, \"lines\"))+theme(plot.margin=unit(c(1,1,1,1),\"cm\"))+\n",
    "    theme(legend.key.size=unit(0.7, \"cm\"))+\n",
    "    theme(panel.background=element_rect(fill=NA))+\n",
    "    theme(legend.key=element_rect(fill=NA))\n",
    "\n",
    "ggsave(\"piN_and_piS_means-2018-05-10.pdf\", p, width = 10, height = 5, device=pdf, path=\"/Users/lmoncla/Documents/H5N1_Cambodian_outbreak_study/paper-and-figure-drafts/figures-2018-05-08\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot number of SNPs/site\n",
    "\n",
    "I am starting to think that maybe using pi for this dataset isn't so great. I think that I am going to try out and see how it looks to just plot the overall number of nonsynonymous and synonymous SNPs per nonsynonymous or synonymou site per gene per species. I can read in the number of SNPs from the combined SNPs output file. For the number of synonymous and nonsynonymous sites, I need to count the samples in df (those that had a snp) but also those that did not. I think that the easiest way to do that will be to just run SNPGenie on a set of dummy data so that I can get the number of synonymous and nonsynonymous sites per gene per sample.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 1: Calculate the average number of synonymous and nonsynonymous sites per gene \n",
    "\n",
    "This will involve using the ones I already have plus generating fake vcf files so that I can run SNPGenie for the samples/genes that didn't have any SNPs. \n",
    "\n",
    "\n",
    "#### Generate dummy vcfs to get the number of nonsynonymous and synonymous sites for genes that don't have snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 589,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_vcfs_directory = \"/Users/lmoncla/Documents/H5N1_Cambodian_outbreak_study/SNPGenie_diversity_estimates/1_percent_SNPs-2018-05-10/samples_and_genes_with_no_snps/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 590,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = []\n",
    "for f in glob.glob(missing_vcfs_directory + \"*.gtf\"):\n",
    "    files.append(f)\n",
    "    sample_name = f.replace(missing_vcfs_directory, \"\")\n",
    "    sample_name = sample_name.replace(\".gtf\",\"\")\n",
    "    samples.append(sample_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 591,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make fake vcf files for each sample \n",
    "for s in samples: \n",
    "    vcfname = missing_vcfs_directory + s + \".vcf\"\n",
    "    with open(vcfname, \"w\") as vcf: \n",
    "        vcf.write(\"#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tSample1\\n\")\n",
    "        vcf.write(s+\"\\t28\\t.\\tC\\tT\\t.\\tPASS\\tADP=147;WT=1000;HET=1;HOM=0;NC=0\\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\\t0/1:37:147:147:135:135,12:8.16%:1.9326E-4:38:37:61:74:5:7\\n\")\n",
    "        vcf.write(s+\"\\t100\\t.\\tC\\tT\\t.\\tPASS\\tADP=147;WT=1000;HET=1;HOM=0;NC=0\\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\\t0/1:37:147:147:135:135,12:8.16%:1.9326E-4:38:37:61:74:5:7\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 593,
   "metadata": {},
   "outputs": [],
   "source": [
    "# consolidate fastas, vcfs, and gtfs all into 1 folder together \n",
    "for s in samples:\n",
    "    call(\"mkdir {missing_vcfs_directory}{s}\".format(missing_vcfs_directory=missing_vcfs_directory,s=s), shell=True)\n",
    "    for n in glob.glob(missing_vcfs_directory + \"*\"):\n",
    "        if s in n:\n",
    "            call(\"mv {n} {missing_vcfs_directory}{s}\".format(s=s, n=n,missing_vcfs_directory=missing_vcfs_directory), shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 594,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run SNPgenie\n",
    "for s in samples:\n",
    "    call(\"snpgenie.pl --vcfformat=4 --gtffile={missing_vcfs_directory}{s}/{s}.gtf --fastafile={missing_vcfs_directory}{s}/{s}.fasta --snpreport={missing_vcfs_directory}{s}/{s}.vcf\".format(missing_vcfs_directory=missing_vcfs_directory,s=s),shell=True)\n",
    "    call(\"mv SNPGenie_Results {missing_vcfs_directory}{s}/\".format(missing_vcfs_directory=missing_vcfs_directory,s=s), shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compile all of the output files from the various genes\n",
    "call(\"grep -i '' */SNPGenie_Results/product_results.txt >> combined.txt\".format(directory=directory),shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 595,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clean up outfile\n",
    "cleanup_snpgenie_output(missing_vcfs_directory, \"combined.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 605,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42\n"
     ]
    }
   ],
   "source": [
    "# read in the SNPGenie pi results as a dataframe\n",
    "df_dummy_values = pd.read_table(missing_vcfs_directory + \"combined_pi_formatted.txt\", sep=\"\\t\")\n",
    "df_dummy_values['product'] = df['product'].fillna('neuram')                                     # replace NA with neuram\n",
    "df_dummy_values['species'] = np.where((df_dummy_values['file'].str.contains(\"duck\")), \"duck\",\"human\")        # add in a species column\n",
    "df_dummy_values = df_dummy_values[['file','product','N_sites','S_sites','species']]\n",
    "df_dummy_values = df_dummy_values.rename(index=str, columns={\"file\":'sample','product':'gene'})\n",
    "df_dummy_values.head()\n",
    "print(len(df_dummy_values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 606,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "78\n"
     ]
    }
   ],
   "source": [
    "# read in the values from the genes that did have SNPs\n",
    "sites = df[['file','product','N_sites','S_sites','species']]\n",
    "sites = sites.rename(index=str, columns={'file':'sample','product':'gene'})\n",
    "sites.head()\n",
    "print(len(sites))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 607,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>N_sites</th>\n",
       "      <th>S_sites</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA</td>\n",
       "      <td>1328.854414</td>\n",
       "      <td>372.145586</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP</td>\n",
       "      <td>1145.707418</td>\n",
       "      <td>348.292582</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PB2</td>\n",
       "      <td>1751.833333</td>\n",
       "      <td>525.166667</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HA</td>\n",
       "      <td>1332.785015</td>\n",
       "      <td>371.214985</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1</td>\n",
       "      <td>576.666667</td>\n",
       "      <td>179.333333</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene      N_sites     S_sites species\n",
       "0   HA  1328.854414  372.145586    duck\n",
       "1   NP  1145.707418  348.292582    duck\n",
       "2  PB2  1751.833333  525.166667    duck\n",
       "3   HA  1332.785015  371.214985   human\n",
       "4   M1   576.666667  179.333333   human"
      ]
     },
     "execution_count": 607,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine dummy and real data into 1 dataframe\n",
    "sites2 = pd.concat([sites,df_dummy_values])\n",
    "sites2 = sites2[['gene','N_sites','S_sites','species']]\n",
    "sites2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 608,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  # Remove the CWD from sys.path while we load stuff.\n",
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  # This is added back by InteractiveShellApp.init_path()\n",
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  if sys.path[0] == '':\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>N_sites</th>\n",
       "      <th>S_sites</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NEP</td>\n",
       "      <td>286.101039</td>\n",
       "      <td>76.898961</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NEP</td>\n",
       "      <td>286.101039</td>\n",
       "      <td>76.898961</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NEP</td>\n",
       "      <td>285.992593</td>\n",
       "      <td>77.007407</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NS1</td>\n",
       "      <td>520.724740</td>\n",
       "      <td>154.275260</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NS1</td>\n",
       "      <td>521.000000</td>\n",
       "      <td>154.000000</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene     N_sites     S_sites species\n",
       "mean  NEP  286.101039   76.898961   human\n",
       "mean  NEP  286.101039   76.898961   human\n",
       "mean  NEP  285.992593   77.007407    duck\n",
       "mean  NS1  520.724740  154.275260   human\n",
       "mean  NS1  521.000000  154.000000    duck"
      ]
     },
     "execution_count": 608,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate the mean number of nonsynonymous and synonymous sites per gene per species and output that into a dataframe\n",
    "genes = ['PB2','PB1','PA','HA','NP','neuram','M1','M2','NS1','NEP']\n",
    "species = ['duck','human']\n",
    "count = 0\n",
    "\n",
    "for g in genes: \n",
    "    for s in species:\n",
    "        count += 1\n",
    "        subsetdf = sites2.loc[(sites2['gene'] == g) & (sites2['species'] == s)]\n",
    "        subsetdf.loc['mean'] =  subsetdf.mean()\n",
    "        subsetdf['gene'] = g\n",
    "        subsetdf['species'] = s\n",
    "        subsetdf = subsetdf.loc[['mean']]\n",
    "        \n",
    "        if count == 1:\n",
    "            mean_sites_df = subsetdf\n",
    "        else:\n",
    "            mean_sites_df = pd.concat([subsetdf, means_df])\n",
    "\n",
    "mean_sites_df['gene'] = mean_sites_df['gene'].str.replace(\"neuram\",\"NA\")\n",
    "mean_sites_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 609,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'duck': {'M1': {'synonymous': 0, 'nonsynonymous': 4}, 'M2': {'synonymous': 0, 'nonsynonymous': 1}, 'NA': {'synonymous': 2, 'nonsynonymous': 2}, 'NP': {'synonymous': 6, 'nonsynonymous': 6}, 'NS1': {'synonymous': 1, 'nonsynonymous': 3}, 'NEP': {'synonymous': 2, 'nonsynonymous': 1}, 'PA': {'synonymous': 6, 'nonsynonymous': 2}, 'PB2': {'synonymous': 8, 'nonsynonymous': 0}, 'HA': {'synonymous': 2, 'nonsynonymous': 4}, 'PB1': {'synonymous': 3, 'nonsynonymous': 1}}, 'human': {'HA': {'synonymous': 9, 'nonsynonymous': 11}, 'M2': {'synonymous': 1, 'nonsynonymous': 3}, 'NA': {'synonymous': 6, 'nonsynonymous': 4}, 'NP': {'synonymous': 8, 'nonsynonymous': 5}, 'PA': {'synonymous': 17, 'nonsynonymous': 9}, 'PB2': {'synonymous': 6, 'nonsynonymous': 10}, 'PB1': {'synonymous': 14, 'nonsynonymous': 10}, 'M1': {'synonymous': 5, 'nonsynonymous': 2}, 'NS1': {'synonymous': 3, 'nonsynonymous': 0}, 'NEP': {'synonymous': 0, 'nonsynonymous': 1}}}\n"
     ]
    }
   ],
   "source": [
    "# read in snps csv file and count the number of nonsynonymous and synonymous per gene per species\n",
    "\n",
    "with open(snps_path, \"r\") as infile: \n",
    "    snps_dict = {}\n",
    "    for line in infile: \n",
    "        if \"reference_position\" not in line:\n",
    "            sample = line.split(\"\\t\")[0]\n",
    "            sample = \"_\".join(sample.split(\"_\")[:-1])\n",
    "        \n",
    "            if \"duck\" in sample:\n",
    "                species = \"duck\"\n",
    "            else:\n",
    "                species = \"human\"\n",
    "        \n",
    "            gene = line.split(\"\\t\")[1]\n",
    "            syn_nonsyn = line.split(\"\\t\")[6]\n",
    "\n",
    "            if species not in snps_dict:\n",
    "                snps_dict[species] = {}\n",
    "            if gene not in snps_dict[species]:\n",
    "                snps_dict[species][gene] = {\"synonymous\":0, \"nonsynonymous\":0}\n",
    "            if syn_nonsyn == \"nonsynonymous\":\n",
    "                snps_dict[species][gene][\"nonsynonymous\"] += 1\n",
    "            elif syn_nonsyn == \"synonymous\":\n",
    "                snps_dict[species][gene][\"synonymous\"] += 1\n",
    "                \n",
    "print(snps_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 610,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>synonymous</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HA</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1</th>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NEP</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     synonymous  nonsynonymous species\n",
       "HA            2              4    duck\n",
       "M1            0              4    duck\n",
       "M2            0              1    duck\n",
       "NA            2              2    duck\n",
       "NEP           2              1    duck"
      ]
     },
     "execution_count": 610,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make into a dataframe; this has the total number of SNPs detected per gene per species, not an average\n",
    "hsnps_df = pd.DataFrame.from_dict(snps_dict[\"human\"], orient=\"index\")\n",
    "hsnps_df['species'] = \"human\"\n",
    "dsnps_df = pd.DataFrame.from_dict(snps_dict[\"duck\"], orient=\"index\")\n",
    "dsnps_df['species'] = \"duck\"\n",
    "snps_df = pd.concat([dsnps_df, hsnps_df])\n",
    "snps_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 611,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>species</th>\n",
       "      <th>mean_synonymous</th>\n",
       "      <th>mean_nonsynonymous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NA</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NEP</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene  synonymous  nonsynonymous species  mean_synonymous  mean_nonsynonymous\n",
       "0   HA           2              4    duck         0.333333            0.666667\n",
       "1   M1           0              4    duck         0.000000            0.666667\n",
       "2   M2           0              1    duck         0.000000            0.166667\n",
       "3   NA           2              2    duck         0.333333            0.333333\n",
       "4  NEP           2              1    duck         0.333333            0.166667"
      ]
     },
     "execution_count": 611,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# take the mean number \n",
    "snps_df['mean_synonymous'] = snps_df['synonymous']/6\n",
    "snps_df['mean_nonsynonymous'] = snps_df['nonsynonymous']/6\n",
    "snps_df = snps_df.reset_index()\n",
    "snps_df = snps_df.rename(index=str, columns={'index':'gene'})\n",
    "snps_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 612,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>N_sites</th>\n",
       "      <th>S_sites</th>\n",
       "      <th>species</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>mean_synonymous</th>\n",
       "      <th>mean_nonsynonymous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NEP</td>\n",
       "      <td>286.101039</td>\n",
       "      <td>76.898961</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NEP</td>\n",
       "      <td>286.101039</td>\n",
       "      <td>76.898961</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NEP</td>\n",
       "      <td>285.992593</td>\n",
       "      <td>77.007407</td>\n",
       "      <td>duck</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NS1</td>\n",
       "      <td>520.724740</td>\n",
       "      <td>154.275260</td>\n",
       "      <td>human</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NS1</td>\n",
       "      <td>521.000000</td>\n",
       "      <td>154.000000</td>\n",
       "      <td>duck</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene     N_sites     S_sites species  synonymous  nonsynonymous  \\\n",
       "0  NEP  286.101039   76.898961   human           0              1   \n",
       "1  NEP  286.101039   76.898961   human           0              1   \n",
       "2  NEP  285.992593   77.007407    duck           2              1   \n",
       "3  NS1  520.724740  154.275260   human           3              0   \n",
       "4  NS1  521.000000  154.000000    duck           1              3   \n",
       "\n",
       "   mean_synonymous  mean_nonsynonymous  \n",
       "0         0.000000            0.166667  \n",
       "1         0.000000            0.166667  \n",
       "2         0.333333            0.166667  \n",
       "3         0.500000            0.000000  \n",
       "4         0.166667            0.500000  "
      ]
     },
     "execution_count": 612,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = pd.merge(mean_sites_df, snps_df, on=['gene','species'], how=\"outer\")\n",
    "df2 = df2.dropna(axis=0, how='any')\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 613,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>N_sites</th>\n",
       "      <th>S_sites</th>\n",
       "      <th>species</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>mean_synonymous</th>\n",
       "      <th>mean_nonsynonymous</th>\n",
       "      <th>s_snps/s_site</th>\n",
       "      <th>ns_snps/ns_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NEP</td>\n",
       "      <td>286.101039</td>\n",
       "      <td>76.898961</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NEP</td>\n",
       "      <td>286.101039</td>\n",
       "      <td>76.898961</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NEP</td>\n",
       "      <td>285.992593</td>\n",
       "      <td>77.007407</td>\n",
       "      <td>duck</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.004329</td>\n",
       "      <td>0.000583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NS1</td>\n",
       "      <td>520.724740</td>\n",
       "      <td>154.275260</td>\n",
       "      <td>human</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003241</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NS1</td>\n",
       "      <td>521.000000</td>\n",
       "      <td>154.000000</td>\n",
       "      <td>duck</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.001082</td>\n",
       "      <td>0.000960</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene     N_sites     S_sites species  synonymous  nonsynonymous  \\\n",
       "0  NEP  286.101039   76.898961   human           0              1   \n",
       "1  NEP  286.101039   76.898961   human           0              1   \n",
       "2  NEP  285.992593   77.007407    duck           2              1   \n",
       "3  NS1  520.724740  154.275260   human           3              0   \n",
       "4  NS1  521.000000  154.000000    duck           1              3   \n",
       "\n",
       "   mean_synonymous  mean_nonsynonymous  s_snps/s_site  ns_snps/ns_site  \n",
       "0         0.000000            0.166667       0.000000         0.000583  \n",
       "1         0.000000            0.166667       0.000000         0.000583  \n",
       "2         0.333333            0.166667       0.004329         0.000583  \n",
       "3         0.500000            0.000000       0.003241         0.000000  \n",
       "4         0.166667            0.500000       0.001082         0.000960  "
      ]
     },
     "execution_count": 613,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate the mean number of nonsynonymous snps per nonsynonymous site and synonymous snps per synonymous site\n",
    "df2['s_snps/s_site'] = df2['mean_synonymous']/df2['S_sites']\n",
    "df2['ns_snps/ns_site'] = df2['mean_nonsynonymous']/df2['N_sites']\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 614,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>species</th>\n",
       "      <th>variable</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NEP</td>\n",
       "      <td>human</td>\n",
       "      <td>s_snps/s_site</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NEP</td>\n",
       "      <td>human</td>\n",
       "      <td>s_snps/s_site</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NEP</td>\n",
       "      <td>duck</td>\n",
       "      <td>s_snps/s_site</td>\n",
       "      <td>0.004329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NS1</td>\n",
       "      <td>human</td>\n",
       "      <td>s_snps/s_site</td>\n",
       "      <td>0.003241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NS1</td>\n",
       "      <td>duck</td>\n",
       "      <td>s_snps/s_site</td>\n",
       "      <td>0.001082</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene species       variable     value\n",
       "0  NEP   human  s_snps/s_site  0.000000\n",
       "1  NEP   human  s_snps/s_site  0.000000\n",
       "2  NEP    duck  s_snps/s_site  0.004329\n",
       "3  NS1   human  s_snps/s_site  0.003241\n",
       "4  NS1    duck  s_snps/s_site  0.001082"
      ]
     },
     "execution_count": 614,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# melt to plot \n",
    "df2 = pd.melt(df2, id_vars = [\"gene\",\"species\"], value_vars = [\"s_snps/s_site\",\"ns_snps/ns_site\"])\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 615,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/rpy2/rinterface/__init__.py:145: RRuntimeWarning: Warning:\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/rpy2/rinterface/__init__.py:145: RRuntimeWarning:  Ignoring unknown parameters: stat\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "%%R -w 800 -h 500 -u px -i df2  # this sets the size of the plot...otherwise, it will go off the page\n",
    "require(ggplot2)\n",
    "library(ggplot2)\n",
    "\n",
    "df2$gene = factor(df2$gene, levels=c(\"PB2\",\"PB1\",\"PA\",\"HA\",\"NP\",\"NA\",\"M1\",\"M2\",\"NS1\",\"NEP\"))\n",
    "\n",
    "p <- ggplot(data=df2, aes(x=gene, y=value, fill=species)) + \n",
    "    geom_col(position=\"dodge\", stat=\"identity\")+\n",
    "    #geom_errorbar(data=df1, aes(x=gene, ymin=mean-sd, ymax=mean+sd, color=species),position=position_dodge(0.9), stat=\"identity\")+\n",
    "    facet_wrap(~variable, scales=\"free\")+\n",
    "    scale_fill_manual(values=c(\"#99bfaa\",\"#5c3d46\"))+\n",
    "    scale_color_manual(values=c(\"#99bfaa\",\"#5c3d46\"))+\n",
    "    labs(x=\"gene\",y=\"SNPs/site\")+\n",
    "    #ggtitle(\"number of shared sites in random simulations\") + \n",
    "    scale_y_continuous(limits=c(0,0.008))+\n",
    "    #scale_x_continuous(limits=c(0,500))+\n",
    "    theme(plot.title = element_text(size=20, hjust=0.5))+\n",
    "    theme(panel.grid.major.y=element_line(colour=NA))+\n",
    "    theme(panel.grid.minor=element_line(colour=NA,size=NA))+    \n",
    "    theme(strip.background = element_rect(colour=NA, fill=NA))+\n",
    "    theme(axis.line.x=element_line(colour=\"black\"))+\n",
    "    theme(axis.line.y=element_line(colour=\"black\"))+theme(strip.text.x=element_text(size=11))+\n",
    "    theme(axis.title.y=element_text(size=16, vjust=8))+\n",
    "    theme(axis.title.x=element_text(size=16, vjust=-8))+\n",
    "    theme(axis.text=element_text(size=16, colour=\"black\"))+\n",
    "    theme(axis.text.x=element_text(size=16, angle=90))+\n",
    "    theme(legend.text=element_text(size=16))+theme(legend.title=element_text(size=16, face=\"plain\"))+\n",
    "    theme(panel.margin=unit(1, \"lines\"))+theme(plot.margin=unit(c(1,1,1,1),\"cm\"))+\n",
    "    theme(legend.key.size=unit(0.7, \"cm\"))+\n",
    "    theme(panel.background=element_rect(fill=NA))+\n",
    "    theme(legend.key=element_rect(fill=NA))\n",
    "\n",
    "ggsave(\"snps_per_site-2018-05-11.pdf\", p, width = 12, height = 4, device=pdf, path=\"/Users/lmoncla/Documents/H5N1_Cambodian_outbreak_study/paper-and-figure-drafts/figures-2018-05-08\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
